diff --git a/make.h b/make.h
index 37c64c2..5c91663 100644
--- a/make.h
+++ b/make.h
@@ -269,6 +269,6 @@ extern char *program;
 extern unsigned int makelevel;
 
 
-#define DEBUGPR(msg)							\
-  if (debug_flag) { print_spaces (depth); printf (msg, file->name);	\
-		    fflush (stdout);  } else
+#define DEBUGPR(msg) \
+  do if (debug_flag) { print_spaces (depth); printf (msg, file->name); \
+		       fflush (stdout); } while (0)
-- 
cgit v1.0-41-gc330

