diff --git a/configure.in b/configure.in
index 8e3afd5..bf63d7f 100644
--- a/configure.in
+++ b/configure.in
@@ -29,9 +29,15 @@ AC_STAT_MACROS_BROKEN
 
 AC_SUBST(LIBOBJS)
 
-AC_HAVE_FUNCS(getdtablesize sys_siglist _sys_siglist psignal \
+AC_HAVE_FUNCS(getdtablesize psignal \
 	      dup2 getcwd sigsetmask getgroups setlinebuf \
 	      seteuid setegid setreuid setregid strerror)
+AC_COMPILE_CHECK(sys_siglist, ,
+   [extern char *sys_siglist[]; puts(*sys_siglist);],
+   AC_DEFINE(HAVE_SYS_SIGLIST))
+AC_COMPILE_CHECK(_sys_siglist, ,
+   [extern char *_sys_siglist[]; puts(*_sys_siglist);],
+   AC_DEFINE(HAVE__SYS_SIGLIST))
 AC_ALLOCA
 AC_VFORK
 AC_SETVBUF_REVERSED
-- 
cgit v1.0-41-gc330

