diff --git a/ChangeLog b/ChangeLog
index a8ea9d3..6436dbd 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,8 @@
+2000-09-21  Paul D. Smith  <paul@paulandlesley.org>
+
+	* main.c (log_working_directory): Rework the text to use complete
+	sentences, to make life simpler for the translators.
+
 2000-08-21  Paul D. Smith  <paul@paulandlesley.org>
 
 	* variable.c (try_variable_definition): Change how we handle
diff --git a/i18n/ja.po b/i18n/ja.po
index cb83b24..2e72d95 100644
--- a/i18n/ja.po
+++ b/i18n/ja.po
@@ -1049,15 +1049,15 @@ msgstr "\n# Make 
 
 #: main.c:2749
 msgid "Entering"
-msgstr "入ります"
+msgstr "ここに入ります: "
 
 #: main.c:2749
 msgid "Leaving"
-msgstr "出ます"
+msgstr "ここから出ます: "
 
 #: main.c:2768
 msgid "an unknown directory"
-msgstr "ディレクトリ不明"
+msgstr "不明なディレクトリ"
 
 #: main.c:2770
 #, c-format
diff --git a/main.c b/main.c
index c82591b..66369a4 100644
--- a/main.c
+++ b/main.c
@@ -2795,7 +2795,6 @@ log_working_directory (entering)
      int entering;
 {
   static int entered = 0;
-  char *msg = entering ? _("Entering") : _("Leaving");
 
   /* Print nothing without the flag.  Don't print the entering message
      again if we already have.  Don't print the leaving message if we
@@ -2809,12 +2808,20 @@ log_working_directory (entering)
     fputs ("# ", stdout);
 
   if (makelevel == 0)
-    printf ("%s: %s ", program, msg);
+    printf ("%s: ", program);
   else
-    printf ("%s[%u]: %s ", program, makelevel, msg);
+    printf ("%s[%u]: ", program, makelevel);
+
+  /* Use entire sentences to give the translators a fighting chance.  */
 
   if (starting_directory == 0)
-    puts (_("an unknown directory"));
+    if (entering)
+      puts (_("Entering an unknown directory"));
+    else
+      puts (_("Leaving an unknown directory"));
   else
-    printf (_("directory `%s'\n"), starting_directory);
+    if (entering)
+      printf (_("Entering directory `%s'\n"), starting_directory);
+    else
+      printf (_("Leaving directory `%s'\n"), starting_directory);
 }
-- 
cgit v1.0-41-gc330

