diff --git a/job.c b/job.c
index 0309e2e..eb5fab6 100644
--- a/job.c
+++ b/job.c
@@ -161,10 +161,10 @@ child_error (target_name, exit_code, exit_sig, coredump, ignored)
      int exit_code, exit_sig, coredump;
      int ignored;
 {
-  char *ignore_string = ignored ? " (ignored)" : "";
-
   if (exit_sig == 0)
-    error ("*** [%s] Error %d%s", target_name, exit_code, ignore_string);
+    error (ignored ? "[%s] Error %d (ignored)" :
+	   "*** [%s] Error %d%s",
+	   target_name, exit_code);
   else
     {
       char *coredump_string = coredump ? " (core dumped)" : "";
-- 
cgit v1.0-41-gc330

