diff --git a/file.c b/file.c
index 367c9aa..56b9912 100644
--- a/file.c
+++ b/file.c
@@ -425,10 +425,10 @@ print_file_data_base ()
 	      switch (f->command_state)
 		{
 		case cs_running:
-		  puts ("#  Commands currently running (?!).");
+		  puts ("#  Commands currently running (THIS IS A BUG).");
 		  break;
 		case cs_deps_running:
-		  puts ("#  Dependencies currently being made (?!).");
+		  puts ("#  Dependencies commands running (THIS IS A BUG).");
 		  break;
 		case cs_not_started:
 		case cs_finished:
@@ -450,7 +450,8 @@ print_file_data_base ()
 		    }
 		  break;
 		default:
-		  puts ("#  Invalid value in `command_state' member!");
+		  puts ("#  Invalid value in `command_state
+' member!");
 		  fflush (stdout);
 		  fflush (stderr);
 		  abort ();
-- 
cgit v1.0-41-gc330

