diff --git a/ChangeLog b/ChangeLog
index e4f9130..ff5bec5 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -30,6 +30,13 @@
 	already does.  Otherwise, the linker will look for, and fail to
 	find, gmk_* functions in some external dynamic library.
 
+2013-05-17  Benno Schulenberg  <bensberg@justemail.net>
+
+	* main.c (decode_output_sync_flags): Fix output message.
+	* read.c (EXTRANEOUS): Ditto.
+	(record_files): Ditto.
+	* remake.c (update_file_1): Ditto.
+
 2013-05-17  Eli Zaretskii  <eliz@gnu.org>
 
 	* main.c (prepare_mutex_handle_string): Define conditioned on
diff --git a/main.c b/main.c
index 2b30260..16ed318 100644
--- a/main.c
+++ b/main.c
@@ -2793,7 +2793,7 @@ decode_switches (int argc, char **argv, int env)
 
                       if (i < 1 || cp[0] != '\0')
                         {
-                          error (NILF, _("the '-%c' option requires a positive integral argument"),
+                          error (NILF, _("the '-%c' option requires a positive integer argument"),
                                  cs->c);
                           bad = 1;
                         }
diff --git a/read.c b/read.c
index 87cd554..36d3a8f 100644
--- a/read.c
+++ b/read.c
@@ -1566,7 +1566,7 @@ conditional_line (char *line, int len, const gmk_floc *flocp)
   /* Found one: skip past it and any whitespace after it.  */
   line = next_token (line + len);
 
-#define EXTRANEOUS() error (flocp, _("Extraneous text after '%s' directive"), cmdname)
+#define EXTRANEOUS() error (flocp, _("extraneous text after '%s' directive"), cmdname)
 
   /* An 'endif' cannot contain extra text, and reduces the if-depth by 1  */
   if (cmdtype == c_endif)
@@ -2082,7 +2082,7 @@ record_files (struct nameseq *filenames, const char *pattern,
              more than once.  Just give a warning since this is harmless.  */
           if (cmds != 0 && cmds == f->cmds)
             error (flocp,
-                   _("target '%s' given more than once in the same rule."),
+                   _("target '%s' given more than once in the same rule"),
                    f->name);
 
           /* Check for two single-colon entries both with commands.
diff --git a/remake.c b/remake.c
index 95c97cb..f56a856 100644
--- a/remake.c
+++ b/remake.c
@@ -827,7 +827,7 @@ update_file_1 (struct file *file, unsigned int depth)
       DBF (DB_BASIC, _("Successfully remade target file '%s'.\n"));
       break;
     case 1:
-      DBF (DB_BASIC, _("Target file '%s' needs remade under -q.\n"));
+      DBF (DB_BASIC, _("Target file '%s' needs to be remade under -q.\n"));
       break;
     default:
       assert (file->update_status >= 0 && file->update_status <= 2);
-- 
cgit v1.0-41-gc330

