diff --git a/read.c b/read.c
index 15d1b13..83224ac 100644
--- a/read.c
+++ b/read.c
@@ -1120,9 +1120,10 @@ eval (struct ebuffer *ebuf, int set_default)
             if (*p2 != '\0')
               /* There's no need to be ivory-tower about this: check for
                  one of the most common bugs found in makefiles...  */
-              fatal (fstart, _("missing separator%s"),
-                     (cmd_prefix == '\t' && !strneq (line, "        ", 8))
-                     ? "" : _(" (did you mean TAB instead of 8 spaces?)"));
+              if (cmd_prefix == '\t' && !strneq (line, "        ", 8))
+                fatal (fstart, _("missing separator (did you mean TAB instead of 8 spaces?)"));
+              else
+                fatal (fstart, _("missing separator"));
             continue;
           }
 
-- 
cgit v1.0-41-gc330

