diff --git a/ChangeLog b/ChangeLog
index 5a0e820..37b22fc 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,5 +1,13 @@
 2003-03-24  Paul D. Smith  <psmith@gnu.org>
 
+	* configure.in, Makefile.am, glob/Makefile.am, doc/Makefile.am:
+	Upgrade to autoconf 2.57 and automake 1.7.3.
+
+	* job.c: More OS/2 changes from Andreas Buening.
+
+	* file.c (print_file): Fix variable initialization.
+	Fixes bug #2892.
+
 	* remake.c (notice_finished_file):
 
 	* make.h (ENULLLOOP): Set errno = 0 before invoking the command;
diff --git a/Makefile.am b/Makefile.am
index 11eeb73..fea1f4d 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,6 +1,6 @@
 # This is a -*-Makefile-*-, or close enough
 
-AUTOMAKE_OPTIONS = 1.7.1 dist-bzip2 check-news ansi2knr
+AUTOMAKE_OPTIONS = 1.7.3 dist-bzip2 check-news ansi2knr
 ACLOCAL_AMFLAGS	=  -I config
 
 SUBDIRS =	glob config po doc
diff --git a/README.cvs b/README.cvs
index 0b416e8..9cf91b7 100644
--- a/README.cvs
+++ b/README.cvs
@@ -23,8 +23,8 @@ There is no password; just hit the ENTER key if you are asked for one.
 Building From CVS
 -----------------
 
-To build GNU make from CVS, you will need Autoconf 2.56 (or better),
-Automake 1.7.1 (or better), and Gettext 0.11.5 (or better), and any
+To build GNU make from CVS, you will need Autoconf 2.57 (or better),
+Automake 1.7.3 (or better), and Gettext 0.11.5 (or better), and any
 tools that those utilities require (GNU m4, Perl, etc.).  You will also
 need a copy of wget.
 
diff --git a/config/Makefile.am b/config/Makefile.am
index 1acf844..0a4e3c7 100644
--- a/config/Makefile.am
+++ b/config/Makefile.am
@@ -1,3 +1,7 @@
+# -*-Makefile-*-, or close enough
+
+AUTOMAKE_OPTIONS =	1.7.3
+
 EXTRA_DIST =	codeset.m4 gettext.m4 glibc21.m4 iconv.m4 isc-posix.m4 \
 		intdiv0.m4 inttypes-pri.m4 inttypes.m4 inttypes_h.m4 \
 		isc-posix.m4 lcmessage.m4 lib-ld.m4 lib-link.m4 lib-prefix.m4 \
diff --git a/configure.in b/configure.in
index 38a84ee..c2f065e 100644
--- a/configure.in
+++ b/configure.in
@@ -1,8 +1,8 @@
 # Process this file with autoconf to produce a configure script.
 
-AC_INIT([GNU make],[3.81a1],[bug-make@gnu.org])
+AC_INIT([GNU make],[3.81a2],[bug-make@gnu.org])
 
-AC_PREREQ(2.56)
+AC_PREREQ(2.57)
 
 AC_REVISION([[$Id$]])
 
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 4aa626c..0c130b2 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -1,7 +1,6 @@
-## Process this file with automake to create Makefile.in.
+# -*-Makefile-*-, or close enough
 
-## Makefile for GNU make documentation.
-## Copyright 2002 Free Software Foundation, Inc.
+AUTOMAKE_OPTIONS = 1.7.3
 
 TEXI2HTML = texi2html
 TEXI2HTML_FLAGS = -split_chapter
diff --git a/file.c b/file.c
index b2767a2..77e3e6a 100644
--- a/file.c
+++ b/file.c
@@ -633,7 +633,7 @@ file_timestamp_sprintf (char *p, FILE_TIMESTAMP ts)
 static void
 print_file (const void *item)
 {
-  struct file *f = (struct file *)f;
+  struct file *f = (struct file *) item;
   struct dep *d;
   struct dep *ood = 0;
 
diff --git a/glob/Makefile.am b/glob/Makefile.am
index 7623ffe..b777015 100644
--- a/glob/Makefile.am
+++ b/glob/Makefile.am
@@ -1,6 +1,6 @@
 # -*-Makefile-*-, or close enough
 
-AUTOMAKE_OPTIONS =	1.7 foreign
+AUTOMAKE_OPTIONS =	1.7.3 foreign
 
 # Only build the library when the system doesn't already have GNU glob.
 if USE_LOCAL_GLOB
diff --git a/job.c b/job.c
index 25988a9..b083214 100644
--- a/job.c
+++ b/job.c
@@ -2328,9 +2328,9 @@ child_execute_job (int stdin_fd, int stdout_fd, char **argv, char **envp)
 
   /* < 0 only if dup() failed */
   if (save_stdin < 0)
-    fatal (NILF, _("could not duplicate stdin\n"));
+    fatal (NILF, _("no more file handles: could not duplicate stdin\n"));
   if (save_stdout < 0)
-    fatal (NILF, _("could not duplicate stdout\n"));
+    fatal (NILF, _("no more file handles: could not duplicate stdout\n"));
 
   /* Close unnecessary file handles for the child.  */
   if (save_stdin != 0)
@@ -2510,6 +2510,8 @@ exec_command (char **argv, char **envp)
 	shell = lookup_variable ("SHELL", 5);
 	if (shell)
 	  shell = shell->value;
+        else
+          shell = 0;
 # else
 	shell = getenv ("SHELL");
 # endif
@@ -2655,7 +2657,7 @@ construct_command_argv_internal (char *line, char **restp, char *shell,
 			     "set", "setlocal", "shift", "start", "time",
                              "type", "ver", "verify", "vol", ":", 0 };
 
-  static char sh_chars_sh[]  = "#;\"*?[]&|<>(){}$`^";
+  static char sh_chars_sh[]  = "#;\"*?[]&|<>(){}$`^~'";
   static char *sh_cmds_sh[]  = { "echo", "cd", "eval", "exec", "exit", "login",
 				 "logout", "set", "umask", "wait", "while",
 				 "for", "case", "if", ":", ".", "break",
@@ -2751,7 +2753,8 @@ construct_command_argv_internal (char *line, char **restp, char *shell,
     {
       extern int _is_unixy_shell (const char *_path);
 
-      message (1, _("$SHELL changed (was `%s', now `%s')"), default_shell, shell);
+      DB (DB_BASIC, (_("$SHELL changed (was `%s', now `%s')\n"),
+                     default_shell, shell));
       unixy_shell = _is_unixy_shell (shell);
       default_shell = shell;
       /* we must allocate a copy of shell: construct_command_argv() will free
@@ -3236,8 +3239,7 @@ construct_command_argv_internal (char *line, char **restp, char *shell,
 		if (*q == quote)
 		  {
 		    /* remove the quote */
-		    while(*q == quote) /* do not ask */
-		      q++;
+                    q++;
 		    quote = 0;
 		  }
 		else /* normal character: copy it */
@@ -3253,8 +3255,7 @@ construct_command_argv_internal (char *line, char **restp, char *shell,
 	      {
 		/* remove opening quote */
 		quote = *q;
-		while(*q == quote) /* do not ask */
-		  q++;
+                q++;
 	      }
 
 	    /* spaces outside of a quoted string: remove them
-- 
cgit v1.0-41-gc330

