diff --git a/implicit.c b/implicit.c
index 15b0b3a..3cf309f 100644
--- a/implicit.c
+++ b/implicit.c
@@ -1,5 +1,5 @@
 /* Implicit rule searching for GNU Make.
-Copyright (C) 1988, 1989, 1990, 1991 Free Software Foundation, Inc.
+Copyright (C) 1988, 1989, 1990, 1991, 1992 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify
@@ -50,9 +50,11 @@ try_implicit_rule (file, depth)
   return pattern_search (file, 0, depth, 0);
 }
 
-#define DEBUGP2(msg, a1, a2) \
-  if (debug_flag) \
-    { print_spaces (depth); printf (msg, a1, a2); fflush (stdout);  } else
+#define DEBUGP2(msg, a1, a2)						      \
+  do {									      \
+    if (debug_flag)							      \
+      { print_spaces (depth); printf (msg, a1, a2); fflush (stdout); }	      \
+  } while (0)
 
 /* Search the pattern rules for a rule with an existing dependency to make
    FILE.  If a rule is found, the appropriate commands and deps are put in FILE
-- 
cgit v1.0-41-gc330

