diff --git a/AUTHORS b/AUTHORS
index 9b6212f..a503509 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -72,7 +72,7 @@ With suggestions/comments/bug reports from a cast of ... well ...
 hundreds, anyway :)
 
 -------------------------------------------------------------------------------
-Copyright (C) 1997-2016 Free Software Foundation, Inc.
+Copyright (C) 1997-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/Makefile.DOS.template b/Makefile.DOS.template
index 47d83eb..35dc038 100644
--- a/Makefile.DOS.template
+++ b/Makefile.DOS.template
@@ -1,7 +1,7 @@
 # -*-Makefile-*- template for DJGPP
 # Makefile.in generated automatically by automake 1.2 from Makefile.am
 #
-# Copyright (C) 1994-2016 Free Software Foundation, Inc.
+# Copyright (C) 1994-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/Makefile.am b/Makefile.am
index c88c465..d1bd50e 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,6 +1,6 @@
 # This is a -*-Makefile-*-, or close enough
 #
-# Copyright (C) 1997-2016 Free Software Foundation, Inc.
+# Copyright (C) 1997-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/Makefile.ami b/Makefile.ami
index bc4ebf4..0a62e6d 100644
--- a/Makefile.ami
+++ b/Makefile.ami
@@ -3,7 +3,7 @@
 # NOTE: If you have no 'make' program at all to process this makefile, run
 # 'build.sh' instead.
 #
-# Copyright (C) 1995-2016 Free Software Foundation, Inc.
+# Copyright (C) 1995-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/NEWS b/NEWS
index 6e2c5c6..3cd0cb9 100644
--- a/NEWS
+++ b/NEWS
@@ -1560,7 +1560,7 @@ Version 3.05
 (Changes from versions 1 through 3.05 were never recorded.  Sorry.)
 
 -------------------------------------------------------------------------------
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/NMakefile.template b/NMakefile.template
index 2007a3b..fdc6857 100644
--- a/NMakefile.template
+++ b/NMakefile.template
@@ -3,7 +3,7 @@
 # NOTE: If you have no 'make' program at all to process this makefile,
 # run 'build_w32.bat' instead.
 #
-# Copyright (C) 1996-2016 Free Software Foundation, Inc.
+# Copyright (C) 1996-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/README.Amiga b/README.Amiga
index 68d3ea7..bca913f 100644
--- a/README.Amiga
+++ b/README.Amiga
@@ -61,7 +61,7 @@ If you plan to use recursive makes, install make resident:
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1995-2016 Free Software Foundation, Inc.
+Copyright (C) 1995-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.DOS.template b/README.DOS.template
index f9a6d92..1535e6c 100644
--- a/README.DOS.template
+++ b/README.DOS.template
@@ -318,7 +318,7 @@ Bug reports:
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.OS2.template b/README.OS2.template
index 0ce81b4..2dc6a01 100644
--- a/README.OS2.template
+++ b/README.OS2.template
@@ -160,7 +160,7 @@ from the make source tree.
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 2003-2016 Free Software Foundation, Inc.
+Copyright (C) 2003-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.W32.template b/README.W32.template
index 1575766..048f019 100644
--- a/README.W32.template
+++ b/README.W32.template
@@ -319,7 +319,7 @@ Bug reports:
         is described in the GNU make manual and the base README.
 
 -------------------------------------------------------------------------------
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.customs b/README.customs
index 67e1252..5056a70 100644
--- a/README.customs
+++ b/README.customs
@@ -96,7 +96,7 @@ SunOS 4.1.x:
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1998-2016 Free Software Foundation, Inc.
+Copyright (C) 1998-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.git b/README.git
index 41a7675..cb73874 100644
--- a/README.git
+++ b/README.git
@@ -1,7 +1,7 @@
                                                                      -*-text-*-
 
 -------------------------------------------------------------------------------
-Copyright (C) 2002-2016 Free Software Foundation, Inc.
+Copyright (C) 2002-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.template b/README.template
index 8e26003..0883284 100644
--- a/README.template
+++ b/README.template
@@ -172,7 +172,7 @@ at the right README!
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/SMakefile.template b/SMakefile.template
index ce2fb63..2b60c5c 100644
--- a/SMakefile.template
+++ b/SMakefile.template
@@ -3,7 +3,7 @@
 # NOTE: If you have no 'make' program at all to process this makefile,
 # run 'build.sh' instead.
 #
-# Copyright (C) 1995-2016 Free Software Foundation, Inc.
+# Copyright (C) 1995-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/TODO.private b/TODO.private
index a56827a..f1ad1af 100644
--- a/TODO.private
+++ b/TODO.private
@@ -101,7 +101,7 @@ The Rest of the List
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1997-2016 Free Software Foundation, Inc.
+Copyright (C) 1997-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/acinclude.m4 b/acinclude.m4
index 0ac68aa..544fd86 100644
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -1,7 +1,7 @@
 dnl acinclude.m4 -- Extra macros needed for GNU make.
 dnl
 dnl Automake will incorporate this into its generated aclocal.m4.
-dnl Copyright (C) 1998-2016 Free Software Foundation, Inc.
+dnl Copyright (C) 1998-2017 Free Software Foundation, Inc.
 dnl This file is part of GNU Make.
 dnl
 dnl GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/amiga.c b/amiga.c
index cfd0d08..c1b31db 100644
--- a/amiga.c
+++ b/amiga.c
@@ -1,5 +1,5 @@
 /* Running commands on Amiga
-Copyright (C) 1995-2016 Free Software Foundation, Inc.
+Copyright (C) 1995-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/amiga.h b/amiga.h
index afc910a..940bc3a 100644
--- a/amiga.h
+++ b/amiga.h
@@ -1,5 +1,5 @@
 /* Definitions for amiga specific things
-Copyright (C) 1995-2016 Free Software Foundation, Inc.
+Copyright (C) 1995-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/ar.c b/ar.c
index b9c1cf7..65ea01c 100644
--- a/ar.c
+++ b/ar.c
@@ -1,5 +1,5 @@
 /* Interface to 'ar' archives for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 
 This file is part of GNU Make.
 
diff --git a/arscan.c b/arscan.c
index 3db252a..af3c96f 100644
--- a/arscan.c
+++ b/arscan.c
@@ -1,5 +1,5 @@
 /* Library function for scanning an archive file.
-Copyright (C) 1987-2016 Free Software Foundation, Inc.
+Copyright (C) 1987-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/build.template b/build.template
index 4b01b46..4b14693 100644
--- a/build.template
+++ b/build.template
@@ -2,7 +2,7 @@
 # Shell script to build GNU Make in the absence of any 'make' program.
 # @configure_input@
 
-# Copyright (C) 1993-2016 Free Software Foundation, Inc.
+# Copyright (C) 1993-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/build_w32.bat b/build_w32.bat
index 07b422a..4e0f51e 100755
--- a/build_w32.bat
+++ b/build_w32.bat
@@ -1,5 +1,5 @@
 @echo off
-rem Copyright (C) 1996-2016 Free Software Foundation, Inc.
+rem Copyright (C) 1996-2017 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/commands.c b/commands.c
index 124b93e..75a9d2d 100644
--- a/commands.c
+++ b/commands.c
@@ -1,5 +1,5 @@
 /* Command processing for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/commands.h b/commands.h
index 18d8c28..5983829 100644
--- a/commands.h
+++ b/commands.h
@@ -1,5 +1,5 @@
 /* Definition of data structures describing shell commands for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config.ami.template b/config.ami.template
index 8235f06..8141f0d 100644
--- a/config.ami.template
+++ b/config.ami.template
@@ -1,5 +1,5 @@
 /* config.h -- hand-massaged for Amiga                                  -*-C-*-
-Copyright (C) 1995-2016 Free Software Foundation, Inc.
+Copyright (C) 1995-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config.h-vms.template b/config.h-vms.template
index 2a4a943..780ea0b 100644
--- a/config.h-vms.template
+++ b/config.h-vms.template
@@ -1,6 +1,6 @@
 /* config.h-vms. Generated by hand by Klaus Kämpf <kkaempf@rmi.de>      -*-C-*-
 
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config.h.W32.template b/config.h.W32.template
index ce07067..c36935c 100644
--- a/config.h.W32.template
+++ b/config.h.W32.template
@@ -1,6 +1,6 @@
 /* config.h.W32 -- hand-massaged config.h file for Windows builds       -*-C-*-
 
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config/ChangeLog.1 b/config/ChangeLog.1
index 8549501..d46fef6 100644
--- a/config/ChangeLog.1
+++ b/config/ChangeLog.1
@@ -33,7 +33,7 @@
 	* Makefile.am: New file.
 
 
-Copyright (C) 2002-2016 Free Software Foundation, Inc.
+Copyright (C) 2002-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config/Makefile.am b/config/Makefile.am
index 7bce036..f83e17b 100644
--- a/config/Makefile.am
+++ b/config/Makefile.am
@@ -1,5 +1,5 @@
 # -*-Makefile-*-, or close enough
-# Copyright (C) 2002-2016 Free Software Foundation, Inc.
+# Copyright (C) 2002-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/config/dospaths.m4 b/config/dospaths.m4
index 9aa9814..8656525 100644
--- a/config/dospaths.m4
+++ b/config/dospaths.m4
@@ -1,7 +1,7 @@
 # Test if the system uses DOS-style pathnames (drive specs and backslashes)
 # By Paul Smith <psmith@gnu.org>.  Based on dos.m4 by Jim Meyering.
 #
-# Copyright (C) 1993-2016 Free Software Foundation, Inc.
+# Copyright (C) 1993-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/configh.dos.template b/configh.dos.template
index 50d86cc..72c501c 100644
--- a/configh.dos.template
+++ b/configh.dos.template
@@ -1,6 +1,6 @@
 /* configh.dos -- hand-massaged config.h file for MS-DOS builds         -*-C-*-
 
-Copyright (C) 1994-2016 Free Software Foundation, Inc.
+Copyright (C) 1994-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/configure.ac b/configure.ac
index 153a716..1b03135 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,6 +1,6 @@
 # Process this file with autoconf to produce a configure script.
 #
-# Copyright (C) 1993-2016 Free Software Foundation, Inc.
+# Copyright (C) 1993-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/configure.bat b/configure.bat
index 3c41f38..fa5ae46 100644
--- a/configure.bat
+++ b/configure.bat
@@ -1,5 +1,5 @@
 @echo off
-rem Copyright (C) 1994-2016 Free Software Foundation, Inc.
+rem Copyright (C) 1994-2017 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/debug.h b/debug.h
index 2915975..2213c95 100644
--- a/debug.h
+++ b/debug.h
@@ -1,5 +1,5 @@
 /* Debugging macros and interface.
-Copyright (C) 1999-2016 Free Software Foundation, Inc.
+Copyright (C) 1999-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/default.c b/default.c
index 3d865c7..483339f 100644
--- a/default.c
+++ b/default.c
@@ -1,5 +1,5 @@
 /* Data base of default implicit rules for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/dep.h b/dep.h
index 7f5076e..99d3f88 100644
--- a/dep.h
+++ b/dep.h
@@ -1,5 +1,5 @@
 /* Definitions of dependency data structures for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/dir.c b/dir.c
index f34bbf5..adbb8a9 100644
--- a/dir.c
+++ b/dir.c
@@ -1,5 +1,5 @@
 /* Directory hashing for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 11aa4d4..9a24a47 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -1,5 +1,5 @@
 # -*-Makefile-*-, or close enough
-# Copyright (C) 2000-2016 Free Software Foundation, Inc.
+# Copyright (C) 2000-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/doc/make.texi b/doc/make.texi
index dfa4454..8810f2b 100644
--- a/doc/make.texi
+++ b/doc/make.texi
@@ -26,7 +26,7 @@ of @cite{The GNU Make Manual}, for GNU @code{make} version @value{VERSION}.
 
 Copyright @copyright{} 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995,
 1996, 1997, 1998, 1999, 2000, 2002, 2003, 2004, 2005, 2006, 2007,
-2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016 Free Software
+2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017 Free Software
 Foundation, Inc.
 
 @quotation
diff --git a/dosbuild.bat b/dosbuild.bat
index 71e71e1..91d7804 100644
--- a/dosbuild.bat
+++ b/dosbuild.bat
@@ -1,5 +1,5 @@
 @echo off
-rem Copyright (C) 1998-2016 Free Software Foundation, Inc.
+rem Copyright (C) 1998-2017 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/expand.c b/expand.c
index 20c526f..3341f9e 100644
--- a/expand.c
+++ b/expand.c
@@ -1,5 +1,5 @@
 /* Variable expansion functions for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/file.c b/file.c
index ae1c285..fe58eba 100644
--- a/file.c
+++ b/file.c
@@ -1,5 +1,5 @@
 /* Target file management for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/filedef.h b/filedef.h
index 14b4187..7de6ac0 100644
--- a/filedef.h
+++ b/filedef.h
@@ -1,5 +1,5 @@
 /* Definition of target file data structures for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/function.c b/function.c
index 92a750d..c70c155 100644
--- a/function.c
+++ b/function.c
@@ -1,5 +1,5 @@
 /* Builtin function expansion for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/getloadavg.c b/getloadavg.c
index 10ae56a..663a7de 100644
--- a/getloadavg.c
+++ b/getloadavg.c
@@ -1,5 +1,5 @@
 /* Get the system load averages.
-Copyright (C) 1985-2016 Free Software Foundation, Inc.
+Copyright (C) 1985-2017 Free Software Foundation, Inc.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
 terms of the GNU General Public License as published by the Free Software
diff --git a/getopt.c b/getopt.c
index e3538d4..5d4b89e 100644
--- a/getopt.c
+++ b/getopt.c
@@ -3,7 +3,7 @@ NOTE: getopt is now part of the C library, so if you don't know what
 "Keep this file name-space clean" means, talk to drepper@gnu.org
 before changing it!
 
-Copyright (C) 1987-2016 Free Software Foundation, Inc.
+Copyright (C) 1987-2017 Free Software Foundation, Inc.
 
 NOTE: The canonical source of this file is maintained with the GNU C Library.
 Bugs can be reported to bug-glibc@gnu.org.
diff --git a/getopt.h b/getopt.h
index 4acd4ee..bb756fa 100644
--- a/getopt.h
+++ b/getopt.h
@@ -1,5 +1,5 @@
 /* Declarations for getopt.
-Copyright (C) 1989-2016 Free Software Foundation, Inc.
+Copyright (C) 1989-2017 Free Software Foundation, Inc.
 
 NOTE: The canonical source of this file is maintained with the GNU C Library.
 Bugs can be reported to bug-glibc@gnu.org.
diff --git a/getopt1.c b/getopt1.c
index 0e38b2f..5d0240a 100644
--- a/getopt1.c
+++ b/getopt1.c
@@ -1,5 +1,5 @@
 /* getopt_long and getopt_long_only entry points for GNU getopt.
-Copyright (C) 1987-1994, 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1987-1994, 1996-2017 Free Software Foundation, Inc.
 
 NOTE: The canonical source of this file is maintained with the GNU C Library.
 Bugs can be reported to bug-glibc@gnu.org.
diff --git a/gettext.h b/gettext.h
index c48ffa0..10992fb 100644
--- a/gettext.h
+++ b/gettext.h
@@ -1,5 +1,5 @@
 /* Convenience header for conditional use of GNU <libintl.h>.
-Copyright (C) 1995-2016 Free Software Foundation, Inc.
+Copyright (C) 1995-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/gmk-default.scm b/gmk-default.scm
index e7353f9..a827cae 100644
--- a/gmk-default.scm
+++ b/gmk-default.scm
@@ -1,5 +1,5 @@
 ;; Contents of the (gnu make) Guile module
-;; Copyright (C) 2011-2016 Free Software Foundation, Inc.
+;; Copyright (C) 2011-2017 Free Software Foundation, Inc.
 ;; This file is part of GNU Make.
 ;;
 ;; GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/gnumake.h b/gnumake.h
index b508562..c1a44ec 100644
--- a/gnumake.h
+++ b/gnumake.h
@@ -1,7 +1,7 @@
 /* External interfaces usable by dynamic objects loaded into GNU Make.
    --THIS API IS A "TECHNOLOGY PREVIEW" ONLY.  IT IS NOT A STABLE INTERFACE--
 
-Copyright (C) 2013-2016 Free Software Foundation, Inc.
+Copyright (C) 2013-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/guile.c b/guile.c
index 1b055c3..a593666 100644
--- a/guile.c
+++ b/guile.c
@@ -1,5 +1,5 @@
 /* GNU Guile interface for GNU Make.
-Copyright (C) 2011-2016 Free Software Foundation, Inc.
+Copyright (C) 2011-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/implicit.c b/implicit.c
index ed49bd1..e5046a4 100644
--- a/implicit.c
+++ b/implicit.c
@@ -1,5 +1,5 @@
 /* Implicit rule searching for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/job.c b/job.c
index 98e7925..5f24597 100644
--- a/job.c
+++ b/job.c
@@ -1,5 +1,5 @@
 /* Job execution and handling for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/job.h b/job.h
index a8677bf..0b7730b 100644
--- a/job.h
+++ b/job.h
@@ -1,5 +1,5 @@
 /* Definitions for managing subprocesses in GNU Make.
-Copyright (C) 1992-2016 Free Software Foundation, Inc.
+Copyright (C) 1992-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/load.c b/load.c
index 37e7b8e..814764d 100644
--- a/load.c
+++ b/load.c
@@ -1,5 +1,5 @@
 /* Loading dynamic objects for GNU Make.
-Copyright (C) 2012-2016 Free Software Foundation, Inc.
+Copyright (C) 2012-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/loadapi.c b/loadapi.c
index 14b75f8..db61e77 100644
--- a/loadapi.c
+++ b/loadapi.c
@@ -1,5 +1,5 @@
 /* API for GNU Make dynamic objects.
-Copyright (C) 2013-2016 Free Software Foundation, Inc.
+Copyright (C) 2013-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/main.c b/main.c
index b2c4ac7..226b26b 100644
--- a/main.c
+++ b/main.c
@@ -1,5 +1,5 @@
 /* Argument parsing and main program of GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
@@ -3318,7 +3318,7 @@ print_version (void)
      year, and none of the rest of it should be translated (including the
      word "Copyright"), so it hardly seems worth it.  */
 
-  printf ("%sCopyright (C) 1988-2016 Free Software Foundation, Inc.\n",
+  printf ("%sCopyright (C) 1988-2017 Free Software Foundation, Inc.\n",
           precede);
 
   printf (_("%sLicense GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\
diff --git a/maintMakefile b/maintMakefile
index cab8dd6..4a2d62c 100644
--- a/maintMakefile
+++ b/maintMakefile
@@ -449,7 +449,7 @@ $(UPLOADS): $(DIST_ARCHIVES) $(DIST_ARCHIVES_SIG) $(DIST_ARCHIVES_DIRECTIVE)
 # Rebuild Makefile.in if this file is modifed.
 Makefile.in: maintMakefile
 
-# Copyright (C) 1997-2016 Free Software Foundation, Inc.
+# Copyright (C) 1997-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/make.1 b/make.1
index e0faeae..031321e 100644
--- a/make.1
+++ b/make.1
@@ -361,7 +361,7 @@ This manual page contributed by Dennis Morse of Stanford University.
 Further updates contributed by Mike Frysinger.  It has been reworked by Roland
 McGrath.  Maintained by Paul Smith.
 .SH "COPYRIGHT"
-Copyright \(co 1992-1993, 1996-2016 Free Software Foundation, Inc.
+Copyright \(co 1992-1993, 1996-2017 Free Software Foundation, Inc.
 This file is part of
 .IR "GNU make" .
 .LP
diff --git a/makefile.com b/makefile.com
index f579695..b525fff 100644
--- a/makefile.com
+++ b/makefile.com
@@ -149,7 +149,7 @@ $ exit
 $ endsubroutine : compileit
 $!
 $!-----------------------------------------------------------------------------
-$!Copyright (C) 1996-2016 Free Software Foundation, Inc.
+$!Copyright (C) 1996-2017 Free Software Foundation, Inc.
 $!This file is part of GNU Make.
 $!
 $!GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/makefile.vms b/makefile.vms
index 37702d5..6670a0b 100644
--- a/makefile.vms
+++ b/makefile.vms
@@ -1,6 +1,6 @@
 # -*-Makefile-*- to build GNU make on VMS
 #
-# Copyright (C) 1996-2016 Free Software Foundation, Inc.
+# Copyright (C) 1996-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/makeint.h b/makeint.h
index f7af77b..c79d0b4 100644
--- a/makeint.h
+++ b/makeint.h
@@ -1,5 +1,5 @@
 /* Miscellaneous global declarations and portability cruft for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/misc.c b/misc.c
index 5dc1881..92641b5 100644
--- a/misc.c
+++ b/misc.c
@@ -1,5 +1,5 @@
 /* Miscellaneous generic support functions for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/os.h b/os.h
index 111fa9c..b68b005 100644
--- a/os.h
+++ b/os.h
@@ -1,5 +1,5 @@
 /* Declarations for operating system interfaces for GNU Make.
-Copyright (C) 2016 Free Software Foundation, Inc.
+Copyright (C) 2016-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/output.c b/output.c
index 142a964..93b3dc7 100644
--- a/output.c
+++ b/output.c
@@ -1,5 +1,5 @@
 /* Output to stdout / stderr for GNU make
-Copyright (C) 2013-2016 Free Software Foundation, Inc.
+Copyright (C) 2013-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/output.h b/output.h
index 083cd24..8adec4a 100644
--- a/output.h
+++ b/output.h
@@ -1,5 +1,5 @@
 /* Output to stdout / stderr for GNU make
-Copyright (C) 2013-2016 Free Software Foundation, Inc.
+Copyright (C) 2013-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/po/Makevars b/po/Makevars
index e32ea34..60d008e 100644
--- a/po/Makevars
+++ b/po/Makevars
@@ -1,5 +1,5 @@
 # This is a -*-Makefile-*-
-# Copyright (C) 2002-2016 Free Software Foundation, Inc.
+# Copyright (C) 2002-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/po/POTFILES.in b/po/POTFILES.in
index 78b8c85..214fdf1 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -1,5 +1,5 @@
 # List of source files containing translatable strings.
-# Copyright (C) 2000-2016 Free Software Foundation, Inc.
+# Copyright (C) 2000-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/posixos.c b/posixos.c
index dbafa51..1c18daf 100644
--- a/posixos.c
+++ b/posixos.c
@@ -1,5 +1,5 @@
 /* POSIX-based operating system interface for GNU Make.
-Copyright (C) 2016 Free Software Foundation, Inc.
+Copyright (C) 2016-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/read.c b/read.c
index 29adcd4..764486f 100644
--- a/read.c
+++ b/read.c
@@ -1,5 +1,5 @@
 /* Reading and parsing of makefiles for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/remake.c b/remake.c
index bac5c2b..037ee73 100644
--- a/remake.c
+++ b/remake.c
@@ -1,5 +1,5 @@
 /* Basic dependency engine for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/remote-cstms.c b/remote-cstms.c
index 894b179..dea0276 100644
--- a/remote-cstms.c
+++ b/remote-cstms.c
@@ -3,7 +3,7 @@
    Please do not send bug reports or questions about it to
    the Make maintainers.
 
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/remote-stub.c b/remote-stub.c
index 8e31a20..02c7a88 100644
--- a/remote-stub.c
+++ b/remote-stub.c
@@ -1,5 +1,5 @@
 /* Template for the remote job exportation interface to GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/rule.c b/rule.c
index a173fdd..1fe77e1 100644
--- a/rule.c
+++ b/rule.c
@@ -1,5 +1,5 @@
 /* Pattern and suffix rule internals for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/rule.h b/rule.h
index 9156b8e..b6adc9a 100644
--- a/rule.h
+++ b/rule.h
@@ -1,5 +1,5 @@
 /* Definitions for using pattern rules in GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/signame.c b/signame.c
index 55646e9..9ee00e2 100644
--- a/signame.c
+++ b/signame.c
@@ -1,5 +1,5 @@
 /* Convert between signal names and numbers.
-Copyright (C) 1990-2016 Free Software Foundation, Inc.
+Copyright (C) 1990-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/strcache.c b/strcache.c
index 6dcf2bc..19f43c9 100644
--- a/strcache.c
+++ b/strcache.c
@@ -1,5 +1,5 @@
 /* Constant string caching for GNU Make.
-Copyright (C) 2006-2016 Free Software Foundation, Inc.
+Copyright (C) 2006-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/subproc.bat b/subproc.bat
index 685ed9d..1154be0 100644
--- a/subproc.bat
+++ b/subproc.bat
@@ -1,5 +1,5 @@
 @echo off
-rem Copyright (C) 1996-2016 Free Software Foundation, Inc.
+rem Copyright (C) 1996-2017 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/tests/ChangeLog.1 b/tests/ChangeLog.1
index 684af03..13514a5 100644
--- a/tests/ChangeLog.1
+++ b/tests/ChangeLog.1
@@ -1413,7 +1413,7 @@
 	ChangeLog file for the test suite created.
 
 
-Copyright (C) 1992-2016 Free Software Foundation, Inc.
+Copyright (C) 1992-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/tests/NEWS b/tests/NEWS
index f55e4ba..81ab138 100644
--- a/tests/NEWS
+++ b/tests/NEWS
@@ -162,7 +162,7 @@ Changes from 0.1 to 0.2 (5-4-92):
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1992-2016 Free Software Foundation, Inc.
+Copyright (C) 1992-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/tests/README b/tests/README
index 0213159..aae50d3 100644
--- a/tests/README
+++ b/tests/README
@@ -7,7 +7,7 @@ This entire test suite, including all test files, are copyright and
 distributed under the following terms:
 
  -----------------------------------------------------------------------------
- Copyright (C) 1992-2016 Free Software Foundation, Inc.
+ Copyright (C) 1992-2017 Free Software Foundation, Inc.
  This file is part of GNU Make.
 
  GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/tests/config_flags_pm.com b/tests/config_flags_pm.com
index a4271b6..c5cf0e1 100755
--- a/tests/config_flags_pm.com
+++ b/tests/config_flags_pm.com
@@ -3,7 +3,7 @@ $! config_flags_pm.com  - Build config-flags.pm on VMS.
 $!
 $! Just good enough to run the self tests for now.
 $!
-$! Copyright (C) 2014-2016 Free Software Foundation, Inc.
+$! Copyright (C) 2014-2017 Free Software Foundation, Inc.
 $! This file is part of GNU Make.
 $!
 $! GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/tests/mkshadow b/tests/mkshadow
index 23c7ab0..abdf2fa 100755
--- a/tests/mkshadow
+++ b/tests/mkshadow
@@ -3,7 +3,7 @@
 # Simple script to make a "shadow" test directory, using symbolic links.
 # Typically you'd put the shadow in /tmp or another local disk
 #
-# Copyright (C) 1992-2016 Free Software Foundation, Inc.
+# Copyright (C) 1992-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/tests/run_make_tests.com b/tests/run_make_tests.com
index de79a91..e40e4c4 100755
--- a/tests/run_make_tests.com
+++ b/tests/run_make_tests.com
@@ -4,7 +4,7 @@ $! This is a wrapper for the GNU make perl test programs on VMS.
 $!
 $! Parameter "-help" for description on how to use described below.
 $!
-$! Copyright (C) 2014-2016 Free Software Foundation, Inc.
+$! Copyright (C) 2014-2017 Free Software Foundation, Inc.
 $! This file is part of GNU Make.
 $!
 $! GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/tests/run_make_tests.pl b/tests/run_make_tests.pl
index d0e886b..90c7c58 100644
--- a/tests/run_make_tests.pl
+++ b/tests/run_make_tests.pl
@@ -11,7 +11,7 @@
 #                         [-make <make prog>]
 #                        (and others)
 
-# Copyright (C) 1992-2016 Free Software Foundation, Inc.
+# Copyright (C) 1992-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/tests/test_driver.pl b/tests/test_driver.pl
index f7dbc8d..0e87908 100644
--- a/tests/test_driver.pl
+++ b/tests/test_driver.pl
@@ -5,7 +5,7 @@
 # Written 91-12-02 through 92-01-01 by Stephen McGee.
 # Modified 92-02-11 through 92-02-22 by Chris Arthur to further generalize.
 #
-# Copyright (C) 1991-2016 Free Software Foundation, Inc.
+# Copyright (C) 1991-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/variable.c b/variable.c
index 14d527f..2c68cea 100644
--- a/variable.c
+++ b/variable.c
@@ -1,5 +1,5 @@
 /* Internals of variables for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/variable.h b/variable.h
index a36e147..12aea39 100644
--- a/variable.h
+++ b/variable.h
@@ -1,5 +1,5 @@
 /* Definitions for using variables in GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/version.c b/version.c
index e6d822d..039d5ab 100644
--- a/version.c
+++ b/version.c
@@ -1,5 +1,5 @@
 /* Record version and build host architecture for GNU make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/vms_exit.c b/vms_exit.c
index b08d84d..988ac8a 100644
--- a/vms_exit.c
+++ b/vms_exit.c
@@ -7,7 +7,7 @@
  * Author: John E. Malmberg
  */
 
-/* Copyright (C) 2014-2016 Free Software Foundation, Inc.
+/* Copyright (C) 2014-2017 Free Software Foundation, Inc.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
 terms of the GNU General Public License as published by the Free Software
diff --git a/vms_export_symbol.c b/vms_export_symbol.c
index 954f3f4..38daf12 100644
--- a/vms_export_symbol.c
+++ b/vms_export_symbol.c
@@ -4,7 +4,7 @@
  * DCL symbols.
  */
 
-/* Copyright (C) 2014-2016 Free Software Foundation, Inc.
+/* Copyright (C) 2014-2017 Free Software Foundation, Inc.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
 terms of the GNU General Public License as published by the Free Software
diff --git a/vms_progname.c b/vms_progname.c
index 0665a54..47d12a7 100644
--- a/vms_progname.c
+++ b/vms_progname.c
@@ -36,7 +36,7 @@
  *
  */
 
-/* Copyright (C) 2014-2016 Free Software Foundation, Inc.
+/* Copyright (C) 2014-2017 Free Software Foundation, Inc.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
 terms of the GNU General Public License as published by the Free Software
diff --git a/vmsdir.h b/vmsdir.h
index 814b89d..f1d52d6 100644
--- a/vmsdir.h
+++ b/vmsdir.h
@@ -1,5 +1,5 @@
 /* dirent.h for vms
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/vmsfunctions.c b/vmsfunctions.c
index e422d48..356524d 100644
--- a/vmsfunctions.c
+++ b/vmsfunctions.c
@@ -1,5 +1,5 @@
 /* VMS functions
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/vmsify.c b/vmsify.c
index e504a09..249aa06 100644
--- a/vmsify.c
+++ b/vmsify.c
@@ -1,5 +1,5 @@
 /* vmsify.c -- Module for vms <-> unix file name conversion
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/vmsjobs.c b/vmsjobs.c
index b6c01cf..717861d 100644
--- a/vmsjobs.c
+++ b/vmsjobs.c
@@ -1,7 +1,7 @@
 /* --------------- Moved here from job.c ---------------
    This file must be #included in job.c, as it accesses static functions.
 
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/vpath.c b/vpath.c
index 7ad83c9..1649664 100644
--- a/vpath.c
+++ b/vpath.c
@@ -1,5 +1,5 @@
 /* Implementation of pattern-matching file search paths for GNU Make.
-Copyright (C) 1988-2016 Free Software Foundation, Inc.
+Copyright (C) 1988-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/Makefile.am b/w32/Makefile.am
index 5527f77..eea9d69 100644
--- a/w32/Makefile.am
+++ b/w32/Makefile.am
@@ -1,5 +1,5 @@
 # Makefile.am to create libw32.a for mingw32 host.
-# Copyright (C) 1997-2016 Free Software Foundation, Inc.
+# Copyright (C) 1997-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/w32/compat/dirent.c b/w32/compat/dirent.c
index 8675b09..8c9fd48 100644
--- a/w32/compat/dirent.c
+++ b/w32/compat/dirent.c
@@ -1,5 +1,5 @@
 /* Directory entry code for Window platforms.
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/compat/posixfcn.c b/w32/compat/posixfcn.c
index c760cc8..7b9a577 100644
--- a/w32/compat/posixfcn.c
+++ b/w32/compat/posixfcn.c
@@ -1,6 +1,6 @@
 /* Replacements for Posix functions and Posix functionality for MS-Windows.
 
-Copyright (C) 2013-2016 Free Software Foundation, Inc.
+Copyright (C) 2013-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/dirent.h b/w32/include/dirent.h
index a34fabf..ea817f1 100644
--- a/w32/include/dirent.h
+++ b/w32/include/dirent.h
@@ -1,5 +1,5 @@
 /* Windows version of dirent.h
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/dlfcn.h b/w32/include/dlfcn.h
index 5a2ae28..55c5f20 100644
--- a/w32/include/dlfcn.h
+++ b/w32/include/dlfcn.h
@@ -1,5 +1,5 @@
 /* dlfcn.h replacement for MS-Windows build.
-Copyright (C) 2013-2016 Free Software Foundation, Inc.
+Copyright (C) 2013-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/pathstuff.h b/w32/include/pathstuff.h
index 923dc00..6f3615d 100644
--- a/w32/include/pathstuff.h
+++ b/w32/include/pathstuff.h
@@ -1,5 +1,5 @@
 /* Definitions for Windows path manipulation.
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/sub_proc.h b/w32/include/sub_proc.h
index f2be0be..b3147bd 100644
--- a/w32/include/sub_proc.h
+++ b/w32/include/sub_proc.h
@@ -1,5 +1,5 @@
 /* Definitions for Windows process invocation.
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/w32err.h b/w32/include/w32err.h
index b4292f2..7a40607 100644
--- a/w32/include/w32err.h
+++ b/w32/include/w32err.h
@@ -1,5 +1,5 @@
 /* Definitions for Windows error handling.
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/pathstuff.c b/w32/pathstuff.c
index 9bd55e6..82450de 100644
--- a/w32/pathstuff.c
+++ b/w32/pathstuff.c
@@ -1,5 +1,5 @@
 /* Path conversion for Windows pathnames.
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/subproc/NMakefile b/w32/subproc/NMakefile
index 325e55c..4a8b403 100644
--- a/w32/subproc/NMakefile
+++ b/w32/subproc/NMakefile
@@ -1,7 +1,7 @@
 # NOTE: If you have no 'make' program at all to process this makefile, run
 # 'build.bat' instead.
 #
-# Copyright (C) 1996-2016 Free Software Foundation, Inc.
+# Copyright (C) 1996-2017 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/w32/subproc/misc.c b/w32/subproc/misc.c
index 8b17413..6e7ac9a 100644
--- a/w32/subproc/misc.c
+++ b/w32/subproc/misc.c
@@ -1,5 +1,5 @@
 /* Process handling for Windows
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/subproc/proc.h b/w32/subproc/proc.h
index 7ccb5ea..ef2972d 100644
--- a/w32/subproc/proc.h
+++ b/w32/subproc/proc.h
@@ -1,5 +1,5 @@
 /* Definitions for Windows
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/subproc/sub_proc.c b/w32/subproc/sub_proc.c
index 2a6eaae..e8552be 100644
--- a/w32/subproc/sub_proc.c
+++ b/w32/subproc/sub_proc.c
@@ -1,5 +1,5 @@
 /* Process handling for Windows.
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/subproc/w32err.c b/w32/subproc/w32err.c
index 14eebed..16537d2 100644
--- a/w32/subproc/w32err.c
+++ b/w32/subproc/w32err.c
@@ -1,5 +1,5 @@
 /* Error handling for Windows
-Copyright (C) 1996-2016 Free Software Foundation, Inc.
+Copyright (C) 1996-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/w32os.c b/w32/w32os.c
index 4f35051..0023e79 100644
--- a/w32/w32os.c
+++ b/w32/w32os.c
@@ -1,5 +1,5 @@
 /* Windows32-based operating system interface for GNU Make.
-Copyright (C) 2016 Free Software Foundation, Inc.
+Copyright (C) 2016-2017 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
-- 
cgit v1.0-41-gc330

