diff --git a/dir.c b/dir.c
index 63c82a0..f34bbf5 100644
--- a/dir.c
+++ b/dir.c
@@ -1082,7 +1082,7 @@ print_dir_data_base (void)
           else if (dir->contents->dirfiles.ht_vec == 0)
             {
 #ifdef WINDOWS32
-              printf (_("# %s (key %s, mtime %ull): could not be opened.\n"),
+              printf (_("# %s (key %s, mtime %I64u): could not be opened.\n"),
                       dir->name, dir->contents->path_key,
                       (unsigned long long)dir->contents->mtime);
 #else  /* WINDOWS32 */
@@ -1119,7 +1119,7 @@ print_dir_data_base (void)
                     }
                 }
 #ifdef WINDOWS32
-              printf (_("# %s (key %s, mtime %ull): "),
+              printf (_("# %s (key %s, mtime %I64u): "),
                       dir->name, dir->contents->path_key,
                       (unsigned long long)dir->contents->mtime);
 #else  /* WINDOWS32 */
-- 
cgit v1.0-41-gc330

