diff --git a/dir.c b/dir.c
index d997762..11343dd 100644
--- a/dir.c
+++ b/dir.c
@@ -785,8 +785,9 @@ print_dir_data_base ()
 		  dir->contents->ino[0], dir->contents->ino[1],
 		  dir->contents->ino[2]);
 #else
-	  printf ("# %s (device %d, inode %d): could not be opened.\n",
-		  dir->name, dir->contents->dev, dir->contents->ino);
+	  printf ("# %s (device %ld, inode %ld): could not be opened.\n",
+		  dir->name, (long int) dir->contents->dev,
+		  (long int) dir->contents->ino);
 #endif
 #endif /* WIN32 */
 	else
-- 
cgit v1.0-41-gc330

