diff --git a/AUTHORS b/AUTHORS
index 14ae479..6539881 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -72,7 +72,7 @@ With suggestions/comments/bug reports from a cast of ... well ...
 hundreds, anyway :)
 
 -------------------------------------------------------------------------------
-Copyright (C) 1997-2013 Free Software Foundation, Inc.
+Copyright (C) 1997-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/Makefile.DOS.template b/Makefile.DOS.template
index c8e7043..a608eeb 100644
--- a/Makefile.DOS.template
+++ b/Makefile.DOS.template
@@ -1,7 +1,7 @@
 # -*-Makefile-*- template for DJGPP
 # Makefile.in generated automatically by automake 1.2 from Makefile.am
 #
-# Copyright (C) 1994-2013 Free Software Foundation, Inc.
+# Copyright (C) 1994-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/Makefile.am b/Makefile.am
index e2c0fe9..204ab32 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,6 +1,6 @@
 # This is a -*-Makefile-*-, or close enough
 #
-# Copyright (C) 1997-2013 Free Software Foundation, Inc.
+# Copyright (C) 1997-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/Makefile.ami b/Makefile.ami
index efbab01..7c561a3 100644
--- a/Makefile.ami
+++ b/Makefile.ami
@@ -3,7 +3,7 @@
 # NOTE: If you have no 'make' program at all to process this makefile, run
 # 'build.sh' instead.
 #
-# Copyright (C) 1995-2013 Free Software Foundation, Inc.
+# Copyright (C) 1995-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/NEWS b/NEWS
index e826623..4b482df 100644
--- a/NEWS
+++ b/NEWS
@@ -1,6 +1,6 @@
 GNU make NEWS                                               -*-indented-text-*-
   History of user-visible changes.
-  13 Oct 2013
+  30 Sep 2014
 
 See the end of this file for copyrights and conditions.
 
@@ -9,15 +9,15 @@ manual, which is contained in this distribution as the file doc/make.texi.
 See the README file and the GNU make manual for instructions for
 reporting bugs.
 
-Version 4.0.90
+Version 4.0.90 (30 Sep 2014)
 
 A complete list of bugs fixed in this version is available here:
 
 http://sv.gnu.org/bugs/index.php?group=make&report_id=111&fix_release_id=105&set=custom
 
-* New variables: $(MAKE_TERMOUT) and $(MAKE_TERMERR) are set if stdout or
-  stderr, respectively, are believed to be writing to a terminal.  These
-  variables are exported by default.
+* New variables: $(MAKE_TERMOUT) and $(MAKE_TERMERR) are set to non-empty
+  values if stdout or stderr, respectively, are believed to be writing to a
+  terminal.  These variables are exported by default.
 
 * Allow a no-text-argument form of the $(file ...) function.  Without a text
   argument nothing is written to the file: it is simply opened in the
@@ -1219,7 +1219,7 @@ Version 3.05
 (Changes from versions 1 through 3.05 were never recorded.  Sorry.)
 
 -------------------------------------------------------------------------------
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/NMakefile.template b/NMakefile.template
index f2e038f..99bf906 100644
--- a/NMakefile.template
+++ b/NMakefile.template
@@ -3,7 +3,7 @@
 # NOTE: If you have no 'make' program at all to process this makefile,
 # run 'build_w32.bat' instead.
 #
-# Copyright (C) 1996-2013 Free Software Foundation, Inc.
+# Copyright (C) 1996-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/README.Amiga b/README.Amiga
index 1fabe07..f9be348 100644
--- a/README.Amiga
+++ b/README.Amiga
@@ -61,7 +61,7 @@ If you plan to use recursive makes, install make resident:
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1995-2013 Free Software Foundation, Inc.
+Copyright (C) 1995-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.DOS.template b/README.DOS.template
index ac087f6..70bada0 100644
--- a/README.DOS.template
+++ b/README.DOS.template
@@ -324,7 +324,7 @@ Bug reports:
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.OS2.template b/README.OS2.template
index 792c4bb..9e2045e 100644
--- a/README.OS2.template
+++ b/README.OS2.template
@@ -160,7 +160,7 @@ from the make source tree.
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 2003-2013 Free Software Foundation, Inc.
+Copyright (C) 2003-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.VMS b/README.VMS
index f7b064d..2ff553c 100644
--- a/README.VMS
+++ b/README.VMS
@@ -268,7 +268,7 @@ Long command lines are now converted to command files.
 Comma (',') as a separator is now allowed. See makefile.vms for an example.
 
 -------------------------------------------------------------------------------
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.W32.template b/README.W32.template
index 12de42c..7648e63 100644
--- a/README.W32.template
+++ b/README.W32.template
@@ -293,7 +293,7 @@ Bug reports:
         is described in the GNU make manual and the base README.
 
 -------------------------------------------------------------------------------
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.customs b/README.customs
index 8b3fa74..30d98ac 100644
--- a/README.customs
+++ b/README.customs
@@ -96,7 +96,7 @@ SunOS 4.1.x:
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1998-2013 Free Software Foundation, Inc.
+Copyright (C) 1998-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.git b/README.git
index 59d3353..8089153 100644
--- a/README.git
+++ b/README.git
@@ -1,7 +1,7 @@
                                                                      -*-text-*-
 
 -------------------------------------------------------------------------------
-Copyright (C) 2002-2013 Free Software Foundation, Inc.
+Copyright (C) 2002-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/README.template b/README.template
index 8458406..ed8aa6f 100644
--- a/README.template
+++ b/README.template
@@ -162,7 +162,7 @@ at the right README!
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/SMakefile.template b/SMakefile.template
index 4dfe691..4af89ae 100644
--- a/SMakefile.template
+++ b/SMakefile.template
@@ -3,7 +3,7 @@
 # NOTE: If you have no 'make' program at all to process this makefile,
 # run 'build.sh' instead.
 #
-# Copyright (C) 1995-2013 Free Software Foundation, Inc.
+# Copyright (C) 1995-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/TODO.private b/TODO.private
index 8faa63e..214ef1b 100644
--- a/TODO.private
+++ b/TODO.private
@@ -101,7 +101,7 @@ The Rest of the List
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1997-2013 Free Software Foundation, Inc.
+Copyright (C) 1997-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/acinclude.m4 b/acinclude.m4
index b7d8efd..53d7ef1 100644
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -1,7 +1,7 @@
 dnl acinclude.m4 -- Extra macros needed for GNU make.
 dnl
 dnl Automake will incorporate this into its generated aclocal.m4.
-dnl Copyright (C) 1998-2013 Free Software Foundation, Inc.
+dnl Copyright (C) 1998-2014 Free Software Foundation, Inc.
 dnl This file is part of GNU Make.
 dnl
 dnl GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/amiga.c b/amiga.c
index 8025483..1ef949c 100644
--- a/amiga.c
+++ b/amiga.c
@@ -1,5 +1,5 @@
 /* Running commands on Amiga
-Copyright (C) 1995-2013 Free Software Foundation, Inc.
+Copyright (C) 1995-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/amiga.h b/amiga.h
index a11a128..00bca4f 100644
--- a/amiga.h
+++ b/amiga.h
@@ -1,5 +1,5 @@
 /* Definitions for amiga specific things
-Copyright (C) 1995-2013 Free Software Foundation, Inc.
+Copyright (C) 1995-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/build.template b/build.template
index 1565b72..369ea32 100644
--- a/build.template
+++ b/build.template
@@ -2,7 +2,7 @@
 # Shell script to build GNU Make in the absence of any 'make' program.
 # @configure_input@
 
-# Copyright (C) 1993-2013 Free Software Foundation, Inc.
+# Copyright (C) 1993-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/build_w32.bat b/build_w32.bat
index b9b8235..00632a7 100644
--- a/build_w32.bat
+++ b/build_w32.bat
@@ -1,5 +1,5 @@
 @echo off
-rem Copyright (C) 1996-2013 Free Software Foundation, Inc.
+rem Copyright (C) 1996-2014 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/commands.c b/commands.c
index 74bf9f4..7123021 100644
--- a/commands.c
+++ b/commands.c
@@ -1,5 +1,5 @@
 /* Command processing for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/commands.h b/commands.h
index 7829e40..0d58f22 100644
--- a/commands.h
+++ b/commands.h
@@ -1,5 +1,5 @@
 /* Definition of data structures describing shell commands for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config.ami.template b/config.ami.template
index 419824f..200e033 100644
--- a/config.ami.template
+++ b/config.ami.template
@@ -1,5 +1,5 @@
 /* config.h -- hand-massaged for Amiga                                  -*-C-*-
-Copyright (C) 1995-2013 Free Software Foundation, Inc.
+Copyright (C) 1995-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config.h.W32.template b/config.h.W32.template
index 07bb626..4cac22a 100644
--- a/config.h.W32.template
+++ b/config.h.W32.template
@@ -1,6 +1,6 @@
 /* config.h.W32 -- hand-massaged config.h file for Windows builds       -*-C-*-
 
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config/ChangeLog.1 b/config/ChangeLog.1
index 4e3155b..7803316 100644
--- a/config/ChangeLog.1
+++ b/config/ChangeLog.1
@@ -33,7 +33,7 @@
 	* Makefile.am: New file.
 
 
-Copyright (C) 2002-2013 Free Software Foundation, Inc.
+Copyright (C) 2002-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/config/Makefile.am b/config/Makefile.am
index f770e64..396c234 100644
--- a/config/Makefile.am
+++ b/config/Makefile.am
@@ -1,5 +1,5 @@
 # -*-Makefile-*-, or close enough
-# Copyright (C) 2002-2013 Free Software Foundation, Inc.
+# Copyright (C) 2002-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/config/dospaths.m4 b/config/dospaths.m4
index 3686377..92fb540 100644
--- a/config/dospaths.m4
+++ b/config/dospaths.m4
@@ -1,7 +1,7 @@
 # Test if the system uses DOS-style pathnames (drive specs and backslashes)
 # By Paul Smith <psmith@gnu.org>.  Based on dos.m4 by Jim Meyering.
 #
-# Copyright (C) 1993-2013 Free Software Foundation, Inc.
+# Copyright (C) 1993-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/configh.dos.template b/configh.dos.template
index 797e234..4aecd04 100644
--- a/configh.dos.template
+++ b/configh.dos.template
@@ -1,6 +1,6 @@
 /* configh.dos -- hand-massaged config.h file for MS-DOS builds         -*-C-*-
 
-Copyright (C) 1994-2013 Free Software Foundation, Inc.
+Copyright (C) 1994-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/configure.ac b/configure.ac
index 8afa51d..a6ebf3e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,6 +1,6 @@
 # Process this file with autoconf to produce a configure script.
 #
-# Copyright (C) 1993-2013 Free Software Foundation, Inc.
+# Copyright (C) 1993-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/configure.bat b/configure.bat
index aba3a3b..e54ba84 100644
--- a/configure.bat
+++ b/configure.bat
@@ -1,5 +1,5 @@
 @echo off
-rem Copyright (C) 1994-2013 Free Software Foundation, Inc.
+rem Copyright (C) 1994-2014 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/debug.h b/debug.h
index 9f630d6..9d2ec04 100644
--- a/debug.h
+++ b/debug.h
@@ -1,5 +1,5 @@
 /* Debugging macros and interface.
-Copyright (C) 1999-2013 Free Software Foundation, Inc.
+Copyright (C) 1999-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/dep.h b/dep.h
index 13cefdc..b8c0d29 100644
--- a/dep.h
+++ b/dep.h
@@ -1,5 +1,5 @@
 /* Definitions of dependency data structures for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/dir.c b/dir.c
index 01373b3..7e00b8f 100644
--- a/dir.c
+++ b/dir.c
@@ -1,5 +1,5 @@
 /* Directory hashing for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 2242259..03977ba 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -1,5 +1,5 @@
 # -*-Makefile-*-, or close enough
-# Copyright (C) 2000-2013 Free Software Foundation, Inc.
+# Copyright (C) 2000-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/doc/make.texi b/doc/make.texi
index 15da9ba..b0f5af7 100644
--- a/doc/make.texi
+++ b/doc/make.texi
@@ -3,7 +3,7 @@
 @setfilename make.info
 
 @include version.texi
-@set EDITION 0.72
+@set EDITION 0.73
 
 @settitle GNU @code{make}
 @setchapternewpage odd
@@ -26,7 +26,7 @@ of @cite{The GNU Make Manual}, for GNU @code{make} version @value{VERSION}.
 
 Copyright @copyright{} 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995,
 1996, 1997, 1998, 1999, 2000, 2002, 2003, 2004, 2005, 2006, 2007,
-2008, 2009, 2010, 2011, 2012, 2013 Free Software Foundation, Inc.
+2008, 2009, 2010, 2011, 2012, 2013, 2014 Free Software Foundation, Inc.
 
 @quotation
 Permission is granted to copy, distribute and/or modify this document
diff --git a/dosbuild.bat b/dosbuild.bat
index 4091463..fac2e88 100644
--- a/dosbuild.bat
+++ b/dosbuild.bat
@@ -1,5 +1,5 @@
 @echo off
-rem Copyright (C) 1998-2013 Free Software Foundation, Inc.
+rem Copyright (C) 1998-2014 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/expand.c b/expand.c
index e4b08f1..1c87db1 100644
--- a/expand.c
+++ b/expand.c
@@ -1,5 +1,5 @@
 /* Variable expansion functions for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/file.c b/file.c
index 92c4052..e1a8e80 100644
--- a/file.c
+++ b/file.c
@@ -1,5 +1,5 @@
 /* Target file management for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/filedef.h b/filedef.h
index f4dfa87..b8973db 100644
--- a/filedef.h
+++ b/filedef.h
@@ -1,5 +1,5 @@
 /* Definition of target file data structures for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/getloadavg.c b/getloadavg.c
index 0f5d400..a755b6d 100644
--- a/getloadavg.c
+++ b/getloadavg.c
@@ -1,5 +1,5 @@
 /* Get the system load averages.
-Copyright (C) 1985-2013 Free Software Foundation, Inc.
+Copyright (C) 1985-2014 Free Software Foundation, Inc.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
 terms of the GNU General Public License as published by the Free Software
diff --git a/getopt.c b/getopt.c
index 5a40b6c..a8051fc 100644
--- a/getopt.c
+++ b/getopt.c
@@ -3,7 +3,7 @@ NOTE: getopt is now part of the C library, so if you don't know what
 "Keep this file name-space clean" means, talk to drepper@gnu.org
 before changing it!
 
-Copyright (C) 1987-2013 Free Software Foundation, Inc.
+Copyright (C) 1987-2014 Free Software Foundation, Inc.
 
 NOTE: The canonical source of this file is maintained with the GNU C Library.
 Bugs can be reported to bug-glibc@gnu.org.
diff --git a/getopt.h b/getopt.h
index cb8865e..adef9a5 100644
--- a/getopt.h
+++ b/getopt.h
@@ -1,5 +1,5 @@
 /* Declarations for getopt.
-Copyright (C) 1989-2013 Free Software Foundation, Inc.
+Copyright (C) 1989-2014 Free Software Foundation, Inc.
 
 NOTE: The canonical source of this file is maintained with the GNU C Library.
 Bugs can be reported to bug-glibc@gnu.org.
diff --git a/getopt1.c b/getopt1.c
index 51bea51..42dea22 100644
--- a/getopt1.c
+++ b/getopt1.c
@@ -1,5 +1,5 @@
 /* getopt_long and getopt_long_only entry points for GNU getopt.
-Copyright (C) 1987-1994, 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1987-1994, 1996-2014 Free Software Foundation, Inc.
 
 NOTE: The canonical source of this file is maintained with the GNU C Library.
 Bugs can be reported to bug-glibc@gnu.org.
diff --git a/gettext.h b/gettext.h
index f1156b9..c2030ff 100644
--- a/gettext.h
+++ b/gettext.h
@@ -1,5 +1,5 @@
 /* Convenience header for conditional use of GNU <libintl.h>.
-Copyright (C) 1995-2013 Free Software Foundation, Inc.
+Copyright (C) 1995-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/gmk-default.scm b/gmk-default.scm
index 89db652..fa19d3c 100644
--- a/gmk-default.scm
+++ b/gmk-default.scm
@@ -1,5 +1,5 @@
 ;; Contents of the (gnu make) Guile module
-;; Copyright (C) 2011-2013 Free Software Foundation, Inc.
+;; Copyright (C) 2011-2014 Free Software Foundation, Inc.
 ;; This file is part of GNU Make.
 ;;
 ;; GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/gnumake.h b/gnumake.h
index 94d725e..fd0b3a4 100644
--- a/gnumake.h
+++ b/gnumake.h
@@ -1,7 +1,7 @@
 /* External interfaces usable by dynamic objects loaded into GNU Make.
    --THIS API IS A "TECHNOLOGY PREVIEW" ONLY.  IT IS NOT A STABLE INTERFACE--
 
-Copyright (C) 2013 Free Software Foundation, Inc.
+Copyright (C) 2013-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/guile.c b/guile.c
index 40d99aa..7c9a015 100644
--- a/guile.c
+++ b/guile.c
@@ -1,5 +1,5 @@
 /* GNU Guile interface for GNU Make.
-Copyright (C) 2011-2013 Free Software Foundation, Inc.
+Copyright (C) 2011-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/implicit.c b/implicit.c
index 5fedec0..8e1d541 100644
--- a/implicit.c
+++ b/implicit.c
@@ -1,5 +1,5 @@
 /* Implicit rule searching for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/job.h b/job.h
index 207fe89..3c921ba 100644
--- a/job.h
+++ b/job.h
@@ -1,5 +1,5 @@
 /* Definitions for managing subprocesses in GNU Make.
-Copyright (C) 1992-2013 Free Software Foundation, Inc.
+Copyright (C) 1992-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/load.c b/load.c
index b937c9f..79294b3 100644
--- a/load.c
+++ b/load.c
@@ -1,5 +1,5 @@
 /* Loading dynamic objects for GNU Make.
-Copyright (C) 2012-2013 Free Software Foundation, Inc.
+Copyright (C) 2012-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/loadapi.c b/loadapi.c
index 298ebf8..fb275dd 100644
--- a/loadapi.c
+++ b/loadapi.c
@@ -1,5 +1,5 @@
 /* API for GNU Make dynamic objects.
-Copyright (C) 2013 Free Software Foundation, Inc.
+Copyright (C) 2013-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/main.c b/main.c
index 09fb1fb..0d43575 100644
--- a/main.c
+++ b/main.c
@@ -3276,7 +3276,7 @@ print_version (void)
      year, and none of the rest of it should be translated (including the
      word "Copyright"), so it hardly seems worth it.  */
 
-  printf ("%sCopyright (C) 1988-2013 Free Software Foundation, Inc.\n",
+  printf ("%sCopyright (C) 1988-2014 Free Software Foundation, Inc.\n",
           precede);
 
   printf (_("%sLicense GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\
diff --git a/maintMakefile b/maintMakefile
index 7f34860..463106e 100644
--- a/maintMakefile
+++ b/maintMakefile
@@ -382,7 +382,7 @@ $(UPLOADS): $(DIST_ARCHIVES) $(DIST_ARCHIVES_SIG) $(DIST_ARCHIVES_DIRECTIVE)
 	$(FTPPUT) "$(gnu-upload-host)" "$(gnu-upload-dir)/$(@:upload-%=%)" $^
 
 
-# Copyright (C) 1997-2013 Free Software Foundation, Inc.
+# Copyright (C) 1997-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/make.1 b/make.1
index 94536fe..ed04db9 100644
--- a/make.1
+++ b/make.1
@@ -363,7 +363,7 @@ This manual page contributed by Dennis Morse of Stanford University.
 Further updates contributed by Mike Frysinger.  It has been reworked by Roland
 McGrath.  Maintained by Paul Smith.
 .SH "COPYRIGHT"
-Copyright \(co 1992-1993, 1996-2013 Free Software Foundation, Inc.
+Copyright \(co 1992-1993, 1996-2014 Free Software Foundation, Inc.
 This file is part of
 .IR "GNU make" .
 .LP
diff --git a/makefile.com b/makefile.com
index 8e6b39b..fe37c05 100644
--- a/makefile.com
+++ b/makefile.com
@@ -146,7 +146,7 @@ $ exit
 $ endsubroutine : compileit
 $!
 $!-----------------------------------------------------------------------------
-$!Copyright (C) 1996-2013 Free Software Foundation, Inc.
+$!Copyright (C) 1996-2014 Free Software Foundation, Inc.
 $!This file is part of GNU Make.
 $!
 $!GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/makefile.vms b/makefile.vms
index d7e67c3..ad5ded7 100644
--- a/makefile.vms
+++ b/makefile.vms
@@ -1,6 +1,6 @@
 # -*-Makefile-*- to build GNU make on VMS
 #
-# Copyright (C) 1996-2013 Free Software Foundation, Inc.
+# Copyright (C) 1996-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/misc.c b/misc.c
index 8e1ad4d..00a3210 100644
--- a/misc.c
+++ b/misc.c
@@ -1,5 +1,5 @@
 /* Miscellaneous generic support functions for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/output.c b/output.c
index a15d563..e537f3a 100644
--- a/output.c
+++ b/output.c
@@ -1,5 +1,5 @@
 /* Output to stdout / stderr for GNU make
-Copyright (C) 2013 Free Software Foundation, Inc.
+Copyright (C) 2013-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/output.h b/output.h
index e0d5daa..7c5332c 100644
--- a/output.h
+++ b/output.h
@@ -1,5 +1,5 @@
 /* Output to stdout / stderr for GNU make
-Copyright (C) 2013 Free Software Foundation, Inc.
+Copyright (C) 2013-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/po/Makevars b/po/Makevars
index 900e794..b69192e 100644
--- a/po/Makevars
+++ b/po/Makevars
@@ -1,5 +1,5 @@
 # This is a -*-Makefile-*-
-# Copyright (C) 2002-2013 Free Software Foundation, Inc.
+# Copyright (C) 2002-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/po/POTFILES.in b/po/POTFILES.in
index ba156ee..4287c5d 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -1,5 +1,5 @@
 # List of source files containing translatable strings.
-# Copyright (C) 2000-2013 Free Software Foundation, Inc.
+# Copyright (C) 2000-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/read.c b/read.c
index 915fa53..6ff4bcc 100644
--- a/read.c
+++ b/read.c
@@ -1,5 +1,5 @@
 /* Reading and parsing of makefiles for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/remake.c b/remake.c
index acf7ca5..299a2aa 100644
--- a/remake.c
+++ b/remake.c
@@ -1,5 +1,5 @@
 /* Basic dependency engine for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/remote-cstms.c b/remote-cstms.c
index a5ef99b..867ee8d 100644
--- a/remote-cstms.c
+++ b/remote-cstms.c
@@ -3,7 +3,7 @@
    Please do not send bug reports or questions about it to
    the Make maintainers.
 
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/remote-stub.c b/remote-stub.c
index 45c3a7c..aeb335c 100644
--- a/remote-stub.c
+++ b/remote-stub.c
@@ -1,5 +1,5 @@
 /* Template for the remote job exportation interface to GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/rule.c b/rule.c
index 2535bd6..9abf5b0 100644
--- a/rule.c
+++ b/rule.c
@@ -1,5 +1,5 @@
 /* Pattern and suffix rule internals for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/rule.h b/rule.h
index 3c9b21f..df9fba2 100644
--- a/rule.h
+++ b/rule.h
@@ -1,5 +1,5 @@
 /* Definitions for using pattern rules in GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/signame.c b/signame.c
index 59dc949..7db2439 100644
--- a/signame.c
+++ b/signame.c
@@ -1,5 +1,5 @@
 /* Convert between signal names and numbers.
-Copyright (C) 1990-2013 Free Software Foundation, Inc.
+Copyright (C) 1990-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/strcache.c b/strcache.c
index ff6a2d1..1ade5e7 100644
--- a/strcache.c
+++ b/strcache.c
@@ -1,5 +1,5 @@
 /* Constant string caching for GNU Make.
-Copyright (C) 2006-2013 Free Software Foundation, Inc.
+Copyright (C) 2006-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/subproc.bat b/subproc.bat
index 9a90487..e3b8d3c 100644
--- a/subproc.bat
+++ b/subproc.bat
@@ -1,5 +1,5 @@
 @echo off
-rem Copyright (C) 1996-2013 Free Software Foundation, Inc.
+rem Copyright (C) 1996-2014 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/tests/ChangeLog.1 b/tests/ChangeLog.1
index 653c5a7..9abe731 100644
--- a/tests/ChangeLog.1
+++ b/tests/ChangeLog.1
@@ -1413,7 +1413,7 @@
 	ChangeLog file for the test suite created.
 
 
-Copyright (C) 1992-2013 Free Software Foundation, Inc.
+Copyright (C) 1992-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/tests/NEWS b/tests/NEWS
index 7fb864c..584f4a4 100644
--- a/tests/NEWS
+++ b/tests/NEWS
@@ -162,7 +162,7 @@ Changes from 0.1 to 0.2 (5-4-92):
 
 
 -------------------------------------------------------------------------------
-Copyright (C) 1992-2013 Free Software Foundation, Inc.
+Copyright (C) 1992-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/tests/README b/tests/README
index 0663082..3b699ee 100644
--- a/tests/README
+++ b/tests/README
@@ -7,7 +7,7 @@ This entire test suite, including all test files, are copyright and
 distributed under the following terms:
 
  -----------------------------------------------------------------------------
- Copyright (C) 1992-2013 Free Software Foundation, Inc.
+ Copyright (C) 1992-2014 Free Software Foundation, Inc.
  This file is part of GNU Make.
 
  GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/tests/mkshadow b/tests/mkshadow
index 5e0a402..aa9b3bc 100755
--- a/tests/mkshadow
+++ b/tests/mkshadow
@@ -3,7 +3,7 @@
 # Simple script to make a "shadow" test directory, using symbolic links.
 # Typically you'd put the shadow in /tmp or another local disk
 #
-# Copyright (C) 1992-2013 Free Software Foundation, Inc.
+# Copyright (C) 1992-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/tests/run_make_tests.pl b/tests/run_make_tests.pl
index 75924e0..9468fab 100644
--- a/tests/run_make_tests.pl
+++ b/tests/run_make_tests.pl
@@ -11,7 +11,7 @@
 #                         [-make <make prog>]
 #                        (and others)
 
-# Copyright (C) 1992-2013 Free Software Foundation, Inc.
+# Copyright (C) 1992-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/tests/test_driver.pl b/tests/test_driver.pl
index 68982f4..91141eb 100644
--- a/tests/test_driver.pl
+++ b/tests/test_driver.pl
@@ -5,7 +5,7 @@
 # Written 91-12-02 through 92-01-01 by Stephen McGee.
 # Modified 92-02-11 through 92-02-22 by Chris Arthur to further generalize.
 #
-# Copyright (C) 1991-2013 Free Software Foundation, Inc.
+# Copyright (C) 1991-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/variable.c b/variable.c
index 3f57e7d..01b6d89 100644
--- a/variable.c
+++ b/variable.c
@@ -1,5 +1,5 @@
 /* Internals of variables for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/variable.h b/variable.h
index 3e2328a..fa2dbbd 100644
--- a/variable.h
+++ b/variable.h
@@ -1,5 +1,5 @@
 /* Definitions for using variables in GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/version.c b/version.c
index f71cfca..d94ecec 100644
--- a/version.c
+++ b/version.c
@@ -1,5 +1,5 @@
 /* Record version and build host architecture for GNU make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/vmsdir.h b/vmsdir.h
index ef53acc..3be3f0c 100644
--- a/vmsdir.h
+++ b/vmsdir.h
@@ -1,5 +1,5 @@
 /* dirent.h for vms
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/vmsify.c b/vmsify.c
index cc90485..02bdbf5 100644
--- a/vmsify.c
+++ b/vmsify.c
@@ -1,5 +1,5 @@
 /* vmsify.c -- Module for vms <-> unix file name conversion
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/vpath.c b/vpath.c
index 94956fa..1bcba04 100644
--- a/vpath.c
+++ b/vpath.c
@@ -1,5 +1,5 @@
 /* Implementation of pattern-matching file search paths for GNU Make.
-Copyright (C) 1988-2013 Free Software Foundation, Inc.
+Copyright (C) 1988-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/Makefile.am b/w32/Makefile.am
index 88ca826..a2c4063 100644
--- a/w32/Makefile.am
+++ b/w32/Makefile.am
@@ -1,5 +1,5 @@
 # Makefile.am to create libw32.a for mingw32 host.
-# Copyright (C) 1997-2013 Free Software Foundation, Inc.
+# Copyright (C) 1997-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/w32/compat/dirent.c b/w32/compat/dirent.c
index 40c559a..56407dd 100644
--- a/w32/compat/dirent.c
+++ b/w32/compat/dirent.c
@@ -1,5 +1,5 @@
 /* Directory entry code for Window platforms.
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/compat/posixfcn.c b/w32/compat/posixfcn.c
index 946f16b..b366399 100644
--- a/w32/compat/posixfcn.c
+++ b/w32/compat/posixfcn.c
@@ -1,6 +1,6 @@
 /* Replacements for Posix functions and Posix functionality for MS-Windows.
 
-Copyright (C) 2013 Free Software Foundation, Inc.
+Copyright (C) 2013-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/dirent.h b/w32/include/dirent.h
index 16f81f7..13308de 100644
--- a/w32/include/dirent.h
+++ b/w32/include/dirent.h
@@ -1,5 +1,5 @@
 /* Windows version of dirent.h
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/dlfcn.h b/w32/include/dlfcn.h
index e920703..c64ac7b 100644
--- a/w32/include/dlfcn.h
+++ b/w32/include/dlfcn.h
@@ -1,5 +1,5 @@
 /* dlfcn.h replacement for MS-Windows build.
-Copyright (C) 2013 Free Software Foundation, Inc.
+Copyright (C) 2013-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/pathstuff.h b/w32/include/pathstuff.h
index 4e34a13..fb03f12 100644
--- a/w32/include/pathstuff.h
+++ b/w32/include/pathstuff.h
@@ -1,5 +1,5 @@
 /* Definitions for Windows path manipulation.
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/sub_proc.h b/w32/include/sub_proc.h
index ae93a8a..1e1b213 100644
--- a/w32/include/sub_proc.h
+++ b/w32/include/sub_proc.h
@@ -1,5 +1,5 @@
 /* Definitions for Windows process invocation.
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/include/w32err.h b/w32/include/w32err.h
index 938a119..7d72226 100644
--- a/w32/include/w32err.h
+++ b/w32/include/w32err.h
@@ -1,5 +1,5 @@
 /* Definitions for Windows error handling.
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/pathstuff.c b/w32/pathstuff.c
index 4b550cb..55332ab 100644
--- a/w32/pathstuff.c
+++ b/w32/pathstuff.c
@@ -1,5 +1,5 @@
 /* Path conversion for Windows pathnames.
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/subproc/NMakefile b/w32/subproc/NMakefile
index 5258fa3..ebf516d 100644
--- a/w32/subproc/NMakefile
+++ b/w32/subproc/NMakefile
@@ -1,7 +1,7 @@
 # NOTE: If you have no 'make' program at all to process this makefile, run
 # 'build.bat' instead.
 #
-# Copyright (C) 1996-2013 Free Software Foundation, Inc.
+# Copyright (C) 1996-2014 Free Software Foundation, Inc.
 # This file is part of GNU Make.
 #
 # GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/w32/subproc/build.bat b/w32/subproc/build.bat
index ecd2ab7..f8604ce 100644
--- a/w32/subproc/build.bat
+++ b/w32/subproc/build.bat
@@ -17,7 +17,7 @@ gcc -mthreads -Wall -gdwarf-2 -g3 %OPT% -I.. -I. -I../include -I../.. -DWINDOWS3
 :BuildEnd
 
 @echo off
-rem Copyright (C) 1996-2013 Free Software Foundation, Inc.
+rem Copyright (C) 1996-2014 Free Software Foundation, Inc.
 rem This file is part of GNU Make.
 rem
 rem GNU Make is free software; you can redistribute it and/or modify it under
diff --git a/w32/subproc/misc.c b/w32/subproc/misc.c
index 5273fa0..3f3bdf8 100644
--- a/w32/subproc/misc.c
+++ b/w32/subproc/misc.c
@@ -1,5 +1,5 @@
 /* Process handling for Windows
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/subproc/proc.h b/w32/subproc/proc.h
index 2dad395..1e4054c 100644
--- a/w32/subproc/proc.h
+++ b/w32/subproc/proc.h
@@ -1,5 +1,5 @@
 /* Definitions for Windows
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/subproc/sub_proc.c b/w32/subproc/sub_proc.c
index 88749da..5496c7e 100644
--- a/w32/subproc/sub_proc.c
+++ b/w32/subproc/sub_proc.c
@@ -1,5 +1,5 @@
 /* Process handling for Windows.
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
diff --git a/w32/subproc/w32err.c b/w32/subproc/w32err.c
index b4fe9c1..9ff4f4c 100644
--- a/w32/subproc/w32err.c
+++ b/w32/subproc/w32err.c
@@ -1,5 +1,5 @@
 /* Error handling for Windows
-Copyright (C) 1996-2013 Free Software Foundation, Inc.
+Copyright (C) 1996-2014 Free Software Foundation, Inc.
 This file is part of GNU Make.
 
 GNU Make is free software; you can redistribute it and/or modify it under the
-- 
cgit v1.0-41-gc330

