diff --git a/variable.h b/variable.h
index bc535a5..68350b9 100644
--- a/variable.h
+++ b/variable.h
@@ -73,6 +73,7 @@ extern void push_new_variable_scope (), pop_variable_scope ();
 
 extern int handle_function ();
 
+extern char *variable_buffer_output ();
 extern char *variable_expand (), *variable_expand_for_file ();
 extern char *allocated_variable_expand_for_file ();
 #define	allocated_variable_expand(line) \
-- 
cgit v1.0-41-gc330

