diff --git a/INSTALL b/INSTALL
index a2c8722..50dbe43 100644
--- a/INSTALL
+++ b/INSTALL
@@ -167,7 +167,9 @@ operates.
 `--quiet'
 `--silent'
 `-q'
-     Do not print messages saying which checks are being made.
+     Do not print messages saying which checks are being made.  To
+     suppress all normal output, redirect it to `/dev/null' (any error
+     messages will still be shown).
 
 `--srcdir=DIR'
      Look for the package's source code in directory DIR.  Usually
-- 
cgit v1.0-41-gc330

