diff --git a/build_w32.bat b/build_w32.bat
index bef80f9..9d75258 100755
--- a/build_w32.bat
+++ b/build_w32.bat
@@ -157,7 +157,7 @@ goto :EOF
 :GccCompile
 :: GCC Compile
 echo on
-gcc -mthreads -Wall -gdwarf-2 -g3 %OPTS% -I%OUTDIR% -I. -I./glob -I./w32/include -DWINDOWS32 -DHAVE_CONFIG_H %EXTRAS% -o %OUTDIR%\%1.o -c %1.c
+gcc -mthreads -Wall -std=gnu99 -gdwarf-2 -g3 %OPTS% -I%OUTDIR% -I. -I./glob -I./w32/include -DWINDOWS32 -DHAVE_CONFIG_H %EXTRAS% -o %OUTDIR%\%1.o -c %1.c
 @echo off
 goto :EOF
 
diff --git a/dir.c b/dir.c
index de85a3c..203a146 100644
--- a/dir.c
+++ b/dir.c
@@ -1082,8 +1082,9 @@ print_dir_data_base (void)
           else if (dir->contents->dirfiles.ht_vec == 0)
             {
 #ifdef WINDOWS32
-              printf (_("# %s (key %s, mtime %d): could not be opened.\n"),
-                      dir->name, dir->contents->path_key,dir->contents->mtime);
+              printf (_("# %s (key %s, mtime %I64d): could not be opened.\n"),
+                      dir->name, dir->contents->path_key,
+		      (int64_t)dir->contents->mtime);
 #else  /* WINDOWS32 */
 #ifdef VMS_INO_T
               printf (_("# %s (device %d, inode [%d,%d,%d]): could not be opened.\n"),
@@ -1118,8 +1119,9 @@ print_dir_data_base (void)
                     }
                 }
 #ifdef WINDOWS32
-              printf (_("# %s (key %s, mtime %d): "),
-                      dir->name, dir->contents->path_key, dir->contents->mtime);
+              printf (_("# %s (key %s, mtime %I64d): "),
+                      dir->name, dir->contents->path_key,
+		      (int64_t)dir->contents->mtime);
 #else  /* WINDOWS32 */
 #ifdef VMS_INO_T
               printf (_("# %s (device %d, inode [%d,%d,%d]): "),
-- 
cgit v1.0-41-gc330

