diff --git a/job.c b/job.c
index be539c7..96f30ae 100644
--- a/job.c
+++ b/job.c
@@ -2322,7 +2322,13 @@ exec_command (char **argv, char **envp)
   switch (errno)
     {
     case ENOENT:
-      OS (error, NILF, _("%s: Command not found"), argv[0]);
+      /* We are in the child: don't use the output buffer.
+         It's not right to run fprintf() here!  */
+      if (makelevel == 0)
+        fprintf (stderr, _("%s: %s: Command not found\n"), program, argv[0]);
+      else
+        fprintf (stderr, _("%s[%u]: %s: Command not found\n"),
+                 program, makelevel, argv[0]);
       break;
     case ENOEXEC:
       {
diff --git a/tests/scripts/features/output-sync b/tests/scripts/features/output-sync
index a7534cb..ff3aff1 100644
--- a/tests/scripts/features/output-sync
+++ b/tests/scripts/features/output-sync
@@ -337,5 +337,13 @@ foo: $(OBJS) ; echo $(or $(filter %.o,$^),$(error fail))
 !,
               '-O', "#MAKEFILE#:2: *** fail.  Stop.\n", 512);
 
+# SV 47365: Make sure exec failure error messages are shown
+# Is "127" not always the same everywhere?  We may have to detect it?
+
+run_make_test(q!
+all:: ; @./foo bar baz
+!,
+              '-O', "#MAKE#: ./foo: Command not found\n#MAKEFILE#:2: recipe for target 'all' failed\n#MAKE#: *** [all] Error 127\n", 512);
+
 # This tells the test driver that the perl test script executed properly.
 1;
-- 
cgit v1.0-41-gc330

