diff --git a/ChangeLog b/ChangeLog
index eb0a46a..bfd80eb 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,17 @@
+1999-09-10  Paul Eggert  <eggert@twinsun.com>
+
+        * acinclude.m4 (AC_SYS_LARGEFILE_FLAGS): If on HP-UX 10.20 or
+        later, and using GCC, define __STDC_EXT__; this works around a
+        bug in GCC 2.95.1.
+
+1999-09-08  Paul D. Smith  <psmith@gnu.org>
+
+	* main.c (print_version): Ugh.  GLIBC's configure tries to check
+	make version strings and is too aggressive with their matching
+	expressions.  I've struck a deal with them to leave the version
+	output as-is for 3.78, and they'll change their configure checks
+	so that I can change this back in the future.
+
 1999-09-07  Eli Zaretskii  <eliz@is.elta.co.il>
 
 	* job.c (construct_command_argv_internal) [__MSDOS__]: Add "echo"
diff --git a/README.template b/README.template
index a8ea5a8..1822174 100644
--- a/README.template
+++ b/README.template
@@ -3,6 +3,9 @@ This directory contains the %VERSION% release of GNU Make.
 See the file NEWS for the user-visible changes from previous releases.
 In addition, there have been bugs fixed.
 
+Please check the system-specific notes below for any caveats related to
+your operating system.
+
 For general building and installation instructions, see the file INSTALL.
 
 If you need to build GNU Make and have no other `make' program to use,
@@ -27,9 +30,9 @@ site.  There is information there about ordering hardcopy documentation.
 
 You can also find the latest versions of GNU Make from there.
 
-Please send GNU make bug reports to bug-make@gnu.org.  Please see the
-section of the manual entitles `Problems and Bugs' for information on
-submitting bug reports.
+You can send GNU make bug reports to bug-make@gnu.org.  Please see the
+section of the GNU make manual entitled `Problems and Bugs' for
+information on submitting useful and complete bug reports.
 
 If you need help using GNU make, try these forums:
 
@@ -43,7 +46,7 @@ Also:
   - See README.customs for details on integrating GNU make with the
     Customs distributed build environment from the Pmake distribution.
 
-  - See README.W32 for details about GNU Make on Windows NT or 95.
+  - See README.W32 for details about GNU Make on Windows NT, 95, or 98.
 
   - See README.Amiga for details about GNU Make on AmigaDOS.
 
@@ -54,6 +57,13 @@ Also:
     of DJGPP; see the WWW page http://www.delorie.com/djgpp/ for more
     information.
 
+
+GNU Make is free software.  See the file COPYING for copying conditions.
+
+
+System-specific Notes
+---------------------
+
 It has been reported that the XLC 1.2 compiler on AIX 3.2 is buggy such
 that if you compile make with `cc -O' on AIX 3.2, it will not work correctly.
 It is said that using `cc' without `-O' does work.
@@ -64,5 +74,9 @@ debug this code, you can do `make check-loadavg' to see if it works
 properly on your system.  (You must run `configure' beforehand, but you
 need not build Make itself to run this test.)
 
-
-GNU Make is free software.  See the file COPYING for copying conditions.
+Another potential source of porting problems is the configure support
+enabling GNU make to access large files (LFS) on those operating systems
+that support them.  Please report any bugs that you find in this area.
+If you run into difficulties, then as a workaround you should be able to
+disable LFS by adding the `--disable-largefile' option to the
+`configure' script.
diff --git a/acinclude.m4 b/acinclude.m4
index 574ea37..3f4f9ad 100644
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -105,6 +105,14 @@ AC_DEFUN(AC_SYS_LARGEFILE_FLAGS,
 	ac_cv_sys_largefile_$1=no
 	ifelse($1, CFLAGS,
 	  [case "$host_os" in
+	   # HP-UX 10.20 requires -D__STDC_EXT__ with gcc 2.95.1.
+changequote(, )dnl
+	   hpux10.[2-9][0-9]* | hpux1[1-9]* | hpux[2-9][0-9]*)
+changequote([, ])dnl
+	     if test "$GCC" = yes; then
+	       ac_cv_sys_largefile_CFLAGS=-D__STDC_EXT__
+	     fi
+	     ;;
 	   # IRIX 6.2 and later require cc -n32.
 changequote(, )dnl
 	   irix6.[2-9]* | irix6.1[0-9]* | irix[7-9].* | irix[1-9][0-9]*)
@@ -112,11 +120,13 @@ changequote([, ])dnl
 	     if test "$GCC" != yes; then
 	       ac_cv_sys_largefile_CFLAGS=-n32
 	     fi
+	   esac
+	   if test "$ac_cv_sys_largefile_CFLAGS" != no; then
 	     ac_save_CC="$CC"
 	     CC="$CC $ac_cv_sys_largefile_CFLAGS"
 	     AC_TRY_LINK(, , , ac_cv_sys_largefile_CFLAGS=no)
 	     CC="$ac_save_CC"
-	   esac])
+	   fi])
       }])])
 
 dnl Internal subroutine of AC_SYS_LARGEFILE.
diff --git a/configure.in b/configure.in
index 366a81e..3002e6a 100644
--- a/configure.in
+++ b/configure.in
@@ -3,7 +3,7 @@ AC_REVISION([$Id$])
 AC_PREREQ(2.13)dnl		dnl Minimum Autoconf version required.
 AC_INIT(vpath.c)dnl		dnl A distinctive file to look for in srcdir.
 
-AM_INIT_AUTOMAKE(make, 3.77.95)
+AM_INIT_AUTOMAKE(make, 3.77.96)
 AM_CONFIG_HEADER(config.h)
 
 dnl Regular configure stuff
diff --git a/glob/ChangeLog b/glob/ChangeLog
index 592e7a0..0cb2fcb 100644
--- a/glob/ChangeLog
+++ b/glob/ChangeLog
@@ -1,3 +1,8 @@
+1999-09-08  Eli Zaretskii  <eliz@is.elta.co.il>
+
+	* glob.c (prefix_array) [__MSDOS__,WINDOWS32]: Keep the trailing
+	slash unless DIRNAME is just "x:/".
+
 1999-09-06  Paul D. Smith  <psmith@gnu.org>
 
 	* fnmatch.c: Update to latest version from GLIBC.
diff --git a/glob/glob.c b/glob/glob.c
index 1ab5d8b..65055fb 100644
--- a/glob/glob.c
+++ b/glob/glob.c
@@ -1105,7 +1105,7 @@ prefix_array (dirname, array, n)
 #if defined __MSDOS__ || defined WINDOWS32
   else if (dirlen > 1)
     {
-      if (dirname[dirlen - 1] == '/')
+      if (dirname[dirlen - 1] == '/' && dirname[dirlen - 2] == ':')
 	/* DIRNAME is "d:/".  Don't prepend the slash from DIRNAME.  */
 	--dirlen;
       else if (dirname[dirlen - 1] == ':')
diff --git a/main.c b/main.c
index 226fa91..f3e1135 100644
--- a/main.c
+++ b/main.c
@@ -2542,20 +2542,20 @@ print_version ()
     /* Do it only once.  */
     return;
 
-  printf ("%sGNU Make %s", precede, version_string);
+  printf ("%sGNU Make version %s", precede, version_string);
   if (remote_description != 0 && *remote_description != '\0')
     printf ("-%s", remote_description);
 
-  printf (_(" (%s)\n\
-%s  by Richard Stallman and Roland McGrath.\n\
+  printf (_(", by Richard Stallman and Roland McGrath.\n\
+%sBuilt for %s\n\
 %sCopyright (C) 1988, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99\n\
 %s\tFree Software Foundation, Inc.\n\
 %sThis is free software; see the source for copying conditions.\n\
 %sThere is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A\n\
 %sPARTICULAR PURPOSE.\n\n\
 %sReport bugs to <bug-make@gnu.org>.\n\n"),
-          make_host,
-	  precede, precede, precede, precede, precede, precede, precede);
+          precede, make_host,
+	  precede, precede, precede, precede, precede, precede);
 
   printed_version = 1;
 
-- 
cgit v1.0-41-gc330

