diff --git a/ChangeLog b/ChangeLog
index 58ce528..7dbfdb2 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,8 @@
+2012-01-25  Eli Zaretskii  <eliz@gnu.org>
+
+	* function.c (define_new_function): Fix format strings in calls to
+	`fatal'.
+
 2012-01-17  Paul Smith  <psmith@gnu.org>
 
 	* guile.c (func_guile): Handle NULL returns from Guile.
diff --git a/function.c b/function.c
index c387b35..52235db 100644
--- a/function.c
+++ b/function.c
@@ -2424,10 +2424,10 @@ define_new_function(const struct floc *flocp,
   if (len > 255)
     fatal (flocp, _("Function name too long: %s\n"), name);
   if (min < 0 || min > 255)
-    fatal (flocp, _("Invalid minimum argument count (%d) for function %s%s\n"),
+    fatal (flocp, _("Invalid minimum argument count (%d) for function %s\n"),
            min, name);
   if (max < 0 || max > 255 || max < min)
-    fatal (flocp, _("Invalid maximum argument count (%d) for function %s%s\n"),
+    fatal (flocp, _("Invalid maximum argument count (%d) for function %s\n"),
            max, name);
 
   ent->name = name;
-- 
cgit v1.0-41-gc330

