diff --git a/remake.c b/remake.c
index 064fab1..5fdeae3 100644
--- a/remake.c
+++ b/remake.c
@@ -719,11 +719,11 @@ remake_file (file)
       else
 	{
 	  /* This is a dependency file we cannot remake.  Fail.  */
-	  static char noway[] = "*** No rule to make target";
+	  static char noway[] = "No rule to make target";
 	  if (keep_going_flag || file->dontcare)
 	    {
 	      if (!file->dontcare)
-		error ("%s `%s'.", noway, file->name);
+		error ("*** %s `%s'.", noway, file->name);
  	      file->update_status = 1;
 	    }
 	  else
-- 
cgit v1.0-41-gc330

