diff --git a/src/ChangeLog b/src/ChangeLog
index cc0a28d..c96e503 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,8 @@
+2014-06-21  Giuseppe Scrivano  <gscrivan@redhat.com>
+
+	* mswindows.c (fake_fork_child): Fix build error.
+	Reported by: Gisle Vanem <gvanem@yahoo.no>.
+
 2014-06-21  Nikita Vetrov  <admin@kanaria.ru> (tiny change)
 
 	* openssl.c: Add one-time call OPENSSL_config in ssl_init.
diff --git a/src/mswindows.c b/src/mswindows.c
index 0a0a868..9cbc94b 100644
--- a/src/mswindows.c
+++ b/src/mswindows.c
@@ -165,7 +165,8 @@ fake_fork_child (void)
       if (new_log_fp)
         {
           info->logfile_changed = true;
-          snprintf (info->filename, sizeof (info->lfilename), "%s", opt.lfilename);
+          snprintf (info->lfilename, sizeof (info->lfilename), "%s",
+                    opt.lfilename);
           fclose (new_log_fp);
         }
     }
-- 
cgit v1.0-41-gc330

