diff --git a/src/ChangeLog b/src/ChangeLog
index d4ea3d0..61f178f 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,10 @@
 2005-04-21  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* gen_sslfunc.c (ssl_init): Fix warning message text; mark the
+	message as translatable.
+
+2005-04-21  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* main.c (print_help): Print the EGD option outside the cluster of
 	SSL options.
 
diff --git a/src/gen_sslfunc.c b/src/gen_sslfunc.c
index 5b13f99..44b5414 100644
--- a/src/gen_sslfunc.c
+++ b/src/gen_sslfunc.c
@@ -219,12 +219,14 @@ ssl_init ()
     {
       if (!can_validate)
 	{
-	  logprintf (LOG_NOTQUIET, "Warrining validation of Server Cert not possible!\n");
+	  logputs (LOG_NOTQUIET,
+		   _("Warning: validation of server certificate not possible!\n"));
 	  verify = SSL_VERIFY_NONE;
 	}
      else
 	{
-	  /* break handshake if server cert is not valid but allow NO-Cert mode */
+	  /* break handshake if server cert is not valid but allow
+	     NO-Cert mode */
 	  verify = SSL_VERIFY_PEER;
 	}
     }
-- 
cgit v1.0-41-gc330

