diff --git a/src/ChangeLog b/src/ChangeLog
index 4e974b0..0420ec6 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,8 @@
+2014-11-19  Tim Ruehsen  <tim.ruehsen@gmx.de>
+ 
+	* gnutls.c (ssl_init): Report load failure of cert files from
+        cert directory only with --debug. Fixes #41331.
+
 2014-11-18  Tim Ruehsen  <tim.ruehsen@gmx.de>
 
 	* Fix warnings from clang-analyzer 3.6
diff --git a/src/gnutls.c b/src/gnutls.c
index 87d1d0b..3b6b1f0 100644
--- a/src/gnutls.c
+++ b/src/gnutls.c
@@ -138,8 +138,7 @@ ssl_init (void)
               hash_table_put (inode_map, (void *)(intptr_t) st.st_ino, NULL);
               if ((rc = gnutls_certificate_set_x509_trust_file (credentials, ca_file,
                                                                 GNUTLS_X509_FMT_PEM)) <= 0)
-                logprintf (LOG_NOTQUIET, _("ERROR: Failed to open cert %s: (%d).\n"),
-                           ca_file, rc);
+                DEBUGP (("WARNING: Failed to open cert %s: (%d).\n", ca_file, rc));
               else
                 ncerts += rc;
             }
-- 
cgit v1.0-41-gc330

