diff --git a/src/ChangeLog b/src/ChangeLog
index 4fd1e27..04ef990 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,10 @@
 2001-11-24  Hrvoje Niksic  <hniksic@arsdigita.com>
 
+	* retr.c (retrieve_url): When the redirection URL doesn't parse,
+	print the correct error message rather than "UNKNOWN".
+
+2001-11-24  Hrvoje Niksic  <hniksic@arsdigita.com>
+
 	* progress.c (bar_finish): If the timer didn't record any time
 	since the download beginning, fake 1ms.
 
diff --git a/src/retr.c b/src/retr.c
index 9b8ddc2..32c58a7 100644
--- a/src/retr.c
+++ b/src/retr.c
@@ -326,10 +326,11 @@ retrieve_url (const char *origurl, char **file, char **newloc,
       mynewloc = construced_newloc;
 
       /* Now, see if this new location makes sense. */
-      newloc_struct = url_parse (mynewloc, NULL);
+      newloc_struct = url_parse (mynewloc, &up_error_code);
       if (!newloc_struct)
 	{
-	  logprintf (LOG_NOTQUIET, "%s: %s.\n", mynewloc, "UNKNOWN");
+	  logprintf (LOG_NOTQUIET, "%s: %s.\n", mynewloc,
+		     url_error (up_error_code));
 	  url_free (newloc_struct);
 	  url_free (u);
 	  if (redirections)
-- 
cgit v1.0-41-gc330

