diff --git a/ChangeLog b/ChangeLog
index 60426f4..e990d33 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,7 @@
+2010-07-09  Giuseppe Scrivano  <gscrivano@gnu.org>
+
+	* bootstrap.conf (gnulib_modules): Add `vasprintf'.  Remove `asprintf'.
+
 2010-07-05  Giuseppe Scrivano  <gscrivano@gnu.org>
 
 	* bootstrap.conf (gnulib_modules): Add `asprintf'.
diff --git a/bootstrap.conf b/bootstrap.conf
index 1692f5c..e8ed80f 100644
--- a/bootstrap.conf
+++ b/bootstrap.conf
@@ -27,7 +27,6 @@ gnulib_modules="
 accept
 alloca
 announce-gen
-asprintf
 bind
 c-ctype
 close
@@ -53,6 +52,7 @@ socket
 stdbool
 strcasestr
 update-copyright
+vasprintf
 write
 "
 
-- 
cgit v1.0-41-gc330

