diff --git a/src/ChangeLog b/src/ChangeLog
index ab42113..4a356d6 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,9 @@
 2010-06-10  Giuseppe Scrivano  <gscrivano@gnu.org>
 
+	* main.c (print_version): Drop the current maintainer line.
+
+2010-06-10  Giuseppe Scrivano  <gscrivano@gnu.org>
+
 	* Makefile.am (build_info.c): Generate the `build_info.c' file under
 	the build directory.
 	Patch by: Daniel Stenberg <daniel@haxx.se>.
diff --git a/src/main.c b/src/main.c
index fbf6a33..fa46eb5 100644
--- a/src/main.c
+++ b/src/main.c
@@ -853,8 +853,6 @@ There is NO WARRANTY, to the extent permitted by law.\n"), stdout);
      names such as this one. See en_US.po for reference. */
   fputs (_("\nOriginally written by Hrvoje Niksic <hniksic@xemacs.org>.\n"),
          stdout);
-  fputs (_("Currently maintained by Micah Cowan <micah@cowan.name>.\n"),
-         stdout);
   fputs (_("Please send bug reports and questions to <bug-wget@gnu.org>.\n"),
          stdout);
   exit (0);
-- 
cgit v1.0-41-gc330

