diff --git a/src/ChangeLog b/src/ChangeLog
index 6976327..6e7df87 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,10 @@
 2005-06-27  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* http.c (gethttp): Improve "POST data file missing" error
+	message.
+
+2005-06-27  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* progress.c (set_progress_implementation): Type COLON as const
 	char *.
 
diff --git a/src/http.c b/src/http.c
index 1a276e6..02519cf 100644
--- a/src/http.c
+++ b/src/http.c
@@ -1353,7 +1353,7 @@ gethttp (struct url *u, struct http_stat *hs, int *dt, struct url *proxy)
 	  post_data_size = file_size (opt.post_file_name);
 	  if (post_data_size == -1)
 	    {
-	      logprintf (LOG_NOTQUIET, _("POST data file missing: %s (%s)\n"),
+	      logprintf (LOG_NOTQUIET, _("POST data file `%s' missing: %s\n"),
 			 opt.post_file_name, strerror (errno));
 	      post_data_size = 0;
 	    }
-- 
cgit v1.0-41-gc330

