diff --git a/src/ChangeLog b/src/ChangeLog
index 6f34167..0a5ed0a 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,9 @@
 2005-06-26  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* progress.c (create_image): Mark the "eta" string for translation.
+
+2005-06-26  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* html-url.c (get_urls_file): Don't explicitly set entry->next to
 	NULL since entry is already zeroed out.
 
diff --git a/src/progress.c b/src/progress.c
index b675bbf..4563ea4 100644
--- a/src/progress.c
+++ b/src/progress.c
@@ -880,7 +880,7 @@ create_image (struct bar_progress *bp, double dl_total_time, bool done)
 
 	  /* Translation note: "ETA" is English-centric, but this must
 	     be short, ideally 3 chars.  Abbreviate if necessary.  */
-	  sprintf (p, "  eta %s", eta_to_human_short (eta));
+	  sprintf (p, _("  eta %s"), eta_to_human_short (eta));
 	  move_to_end (p);
 	}
       else if (bp->total_length > 0)
-- 
cgit v1.0-41-gc330

