diff --git a/src/ChangeLog b/src/ChangeLog
index 632cc70..435e324 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,7 @@
 2001-11-27  Hrvoje Niksic  <hniksic@arsdigita.com>
 
+	* url.c (convert_links): Don't translate %d-%d.
+
 	* main.c (print_help): Remove stray HAVE_RANDOM code.
 
 2001-11-27  Hrvoje Niksic  <hniksic@arsdigita.com>
diff --git a/src/url.c b/src/url.c
index 1ce7222..28e41c8 100644
--- a/src/url.c
+++ b/src/url.c
@@ -1833,8 +1833,7 @@ convert_links (const char *file, struct urlpos *links)
   fclose (fp);
   read_file_free (fm);
 
-  logprintf (LOG_VERBOSE,
-	     _("%d-%d\n"), to_file_count, to_url_count);
+  logprintf (LOG_VERBOSE, "%d-%d\n", to_file_count, to_url_count);
 }
 
 /* Construct and return a malloced copy of the relative link from two
-- 
cgit v1.0-41-gc330

