diff --git a/src/ChangeLog b/src/ChangeLog
index a3ca569..0b36ed4 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,12 @@
+2007-02-02  Hrvoje Niksic  <hniksic@xemacs.org>
+
+	* netrc.c: Don't make netrc_list static, as it prevents
+	compilation with DEBUG_MALLOC.
+
+	* utils.c (aprintf): Don't use vasprintf when DEBUG_MALLOC is
+	requested because, in that case, we want the calls to malloc to be
+	coming from us.
+
 2007-01-23  Hrvoje Niksic  <hniksic@xemacs.org>
 
 	* cookies.c (parse_set_cookie): Would erroneously discard cookies
diff --git a/src/netrc.c b/src/netrc.c
index 48ac0c7..dabcf92 100644
--- a/src/netrc.c
+++ b/src/netrc.c
@@ -46,7 +46,7 @@ so, delete this exception statement from your version.  */
 
 #define NETRC_FILE_NAME ".netrc"
 
-static acc_t *netrc_list;
+acc_t *netrc_list;
 
 static acc_t *parse_netrc (const char *);
 
diff --git a/src/utils.c b/src/utils.c
index 7a90c06..60b040b 100644
--- a/src/utils.c
+++ b/src/utils.c
@@ -159,7 +159,7 @@ sepstring (const char *s)
 char *
 aprintf (const char *fmt, ...)
 {
-#ifdef HAVE_VASPRINTF
+#if defined HAVE_VASPRINTF && !defined DEBUG_MALLOC
   /* Use vasprintf. */
   int ret;
   va_list args;
-- 
cgit v1.0-41-gc330

