diff --git a/src/ChangeLog b/src/ChangeLog
index e2b3c30..9d1c466 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,9 @@
 2014-12-04  Darshit Shah  <darnir@gmail.com>
 
+	* ftp.c (getftp): Remove a call to assert(1) and replace with error message
+
+2014-12-04  Darshit Shah  <darnir@gmail.com>
+
 	* wget.h: Define MIX and MAX macros globally for all files
 	* gnutls.c, retr.c: Delete declaration of MIN macro
 	* http.c, progress.c: Delete declaration of MIN and MAX macros
diff --git a/src/ftp.c b/src/ftp.c
index 9ea8819..c159b9b 100644
--- a/src/ftp.c
+++ b/src/ftp.c
@@ -724,8 +724,10 @@ Error in server response, closing control connection.\n"));
                 break;
 
               default:
-                /* Can't happen. */
-                assert (1);
+                logprintf (LOG_ALWAYS, _("Logically impossible section reached in getftp()"));
+                logprintf (LOG_ALWAYS, _("cwd_count: %d\ncwd_start: %d\ncwd_end: %d\n"),
+                                         cwd_count, cwd_start, cwd_end);
+                abort ();
             }
 
           if (!opt.server_response)
-- 
cgit v1.0-41-gc330

