diff --git a/src/ChangeLog b/src/ChangeLog
index c9e45a4..9e0706c 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,9 @@
+2000-11-02  Hrvoje Niksic  <hniksic@arsdigita.com>
+
+	* ftp.c (ftp_loop_internal): Hide the password from the URL when
+	printing non-verbose.  Problem spotted by Dariusz Mlynarczyk
+	<darekm@bydg.lomac.com.pl>.
+
 2000-11-02  Junio Hamano  <junio@twinsun.com>
 
 	* ftp-basic.c (ftp_login): Make comparison case-insensitive.
diff --git a/src/ftp.c b/src/ftp.c
index d239931..3a19d01 100644
--- a/src/ftp.c
+++ b/src/ftp.c
@@ -996,8 +996,16 @@ ftp_loop_internal (struct urlinfo *u, struct fileinfo *f, ccon *con)
 	}
       logprintf (LOG_VERBOSE, _("%s (%s) - `%s' saved [%ld]\n\n"),
 		 tms, tmrate, locf, len);
-      logprintf (LOG_NONVERBOSE, "%s URL: %s [%ld] -> \"%s\" [%d]\n",
-		 tms, u->url, len, locf, count);
+      if (!opt.verbose && !opt.quiet)
+	{
+	  /* Need to hide the password from the URL.  The `if' is here
+             so that we don't do the needless allocation every
+             time. */
+	  char *hurl = str_url (u->proxy ? u->proxy : u, 1);
+	  logprintf (LOG_NONVERBOSE, "%s URL: %s [%ld] -> \"%s\" [%d]\n",
+		     tms, hurl, len, locf, count);
+	  free (hurl);
+	}
 
       if ((con->cmd & DO_LIST))
 	/* This is a directory listing file. */
-- 
cgit v1.0-41-gc330

