diff --git a/src/log.c b/src/log.c
index 48cf2c3..caa8a4e 100644
--- a/src/log.c
+++ b/src/log.c
@@ -553,7 +553,7 @@ log_init (const char *file, int appendp)
       logfp = fopen (file, appendp ? "a" : "w");
       if (!logfp)
 	{
-	  perror (opt.lfilename);
+	  fprintf (stderr, "%s: %s: %s\n", exec_name, file, strerror (errno));
 	  exit (1);
 	}
     }
@@ -567,16 +567,16 @@ log_init (const char *file, int appendp)
          easier on the user.  */
       logfp = stderr;
 
-      /* If the output is a TTY, enable storing, which will make Wget
-         remember the last several printed messages, to be able to
-         dump them to a log file in case SIGHUP or SIGUSR1 is received
-         (or Ctrl+Break is pressed under Windows).  */
       if (1
 #ifdef HAVE_ISATTY
 	  && isatty (fileno (logfp))
 #endif
 	  )
 	{
+	  /* If the output is a TTY, enable save context, i.e. store
+	     the most recent several messages ("context") and dump
+	     them to a log file in case SIGHUP or SIGUSR1 is received
+	     (or Ctrl+Break is pressed under Windows).  */
 	  save_context_p = 1;
 	}
     }
-- 
cgit v1.0-41-gc330

