diff --git a/src/ChangeLog b/src/ChangeLog
index e066561..127d5df 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,10 @@
 2005-04-11  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* xmalloc.c (register_ptr): Warn the user to increase SZ prior to
+	aborting.
+
+2005-04-11  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* convert.c (convert_cleanup): Free converted_files.
 
 2005-04-10  Hrvoje Niksic  <hniksic@xemacs.org>
diff --git a/src/xmalloc.c b/src/xmalloc.c
index ab5ac81..75112c6 100644
--- a/src/xmalloc.c
+++ b/src/xmalloc.c
@@ -218,9 +218,8 @@ static int malloc_count, free_count;
 
 /* Home-grown hash table of mallocs: */
 
-#define SZ 100003		/* Prime number a little over 100,000.
-				   Increase the table size if you need
-				   to debug larger Wget runs.  */
+#define SZ 100003		/* Prime just over 100,000.  Increase
+				   it to debug larger Wget runs.  */
 
 static struct {
   const void *ptr;
@@ -250,7 +249,11 @@ register_ptr (const void *ptr, const char *file, int line)
 {
   int i;
   if (malloc_count - free_count > SZ)
-    abort ();
+    {
+      fprintf (stderr, "Increase SZ to a larger value and recompile.\n");
+      fflush (stderr);
+      abort ();
+    }
 
   i = ptr_position (ptr);
   malloc_table[i].ptr = ptr;
-- 
cgit v1.0-41-gc330

