diff --git a/src/ChangeLog b/src/ChangeLog
index 3df8dda..d0e26e4 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,8 @@
+2008-06-25  Steven Schubiger  <stsc@members.fsf.org>
+
+	* ftp.c (getftp): When spidering a FTP URL, emit a diagnostic
+	message if the remote file exists.
+
 2008-06-24  Steven Schubiger  <stsc@members.fsf.org>
 
 	* http.c (http_loop): Replace escnonprint() occurence with
diff --git a/src/ftp.c b/src/ftp.c
index 7256faa..482651b 100644
--- a/src/ftp.c
+++ b/src/ftp.c
@@ -805,8 +805,14 @@ Error in server response, closing control connection.\n"));
 		    }
 		  f = f->next;
 		}
-	      if (!exists)
-		{
+              if (exists)
+                {
+                  logputs (LOG_VERBOSE, "\n");
+                  logprintf (LOG_NOTQUIET, _("File %s exists.\n"),
+                             quote (u->file));
+                }
+	      else
+                {
 		  logputs (LOG_VERBOSE, "\n");
 		  logprintf (LOG_NOTQUIET, _("No such file %s.\n"),
 			     quote (u->file));
-- 
cgit v1.0-41-gc330

