diff --git a/src/host.c b/src/host.c
index 5a53022..fdb35b1 100644
--- a/src/host.c
+++ b/src/host.c
@@ -712,7 +712,8 @@ lookup_host (const char *host, int flags)
   /* No luck with the cache; resolve HOST. */
 
   if (!silent && !numeric_address)
-    logprintf (LOG_VERBOSE, _("Resolving %s... "), escnonprint (host));
+    logprintf (LOG_VERBOSE, _("Resolving %s... "), 
+               quotearg_style (escape_quoting_style, host));
 
 #ifdef ENABLE_IPV6
   {
-- 
cgit v1.0-41-gc330

