diff --git a/src/ChangeLog b/src/ChangeLog
index 1ca8b34..5f0a55e 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,10 @@
 2001-12-13  Hrvoje Niksic  <hniksic@arsdigita.com>
 
+	* init.c (wgetrc_file_name): Print correct message when loading
+	getenv("WGETRC") fails.
+
+2001-12-13  Hrvoje Niksic  <hniksic@arsdigita.com>
+
 	* recur.c (register_download): Don't abort when one URL references
 	two different files.
 
diff --git a/src/init.c b/src/init.c
index 030abe9..4a49da4 100644
--- a/src/init.c
+++ b/src/init.c
@@ -301,7 +301,7 @@ wgetrc_file_name (void)
     {
       if (!file_exists_p (env))
 	{
-	  fprintf (stderr, "%s: %s: %s.\n", exec_name, file, strerror (errno));
+	  fprintf (stderr, "%s: %s: %s.\n", exec_name, env, strerror (errno));
 	  exit (1);
 	}
       return xstrdup (env);
-- 
cgit v1.0-41-gc330

