diff --git a/NEWS b/NEWS
index c6f09a4..625ae97 100644
--- a/NEWS
+++ b/NEWS
@@ -5,6 +5,13 @@ See the end for copying conditions.
 
 Please send GNU Wget bug reports to <bug-wget@gnu.org>.
 
+* Changes in Wget 1.19.1
+
+* Fix bugs, a regression, portability/build issues
+
+* Add new option --retry-on-http-error
+
+
 * Changes in Wget 1.19
 
 * New option --use-askpass=COMMAND. Fetch user/password by calling
-- 
cgit v1.0-41-gc330

