diff --git a/src/ChangeLog b/src/ChangeLog
index f83a97b..dd99fd9 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,11 @@
 2003-11-05  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* connect.c (resolve_bind_address): Use a more accurate error
+	message -- we're not binding to ANY, we're disabling bind
+	altogether.
+
+2003-11-05  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* cookies.c (save_cookies_mapper): Respect the setting of
 	keep-session-cookies.
 	(cookie_jar_load): Import session cookies.
diff --git a/src/connect.c b/src/connect.c
index 52dc0b0..80df361 100644
--- a/src/connect.c
+++ b/src/connect.c
@@ -175,9 +175,10 @@ resolve_bind_address (const char *host, struct sockaddr *sa, int flags)
   al = lookup_host (host, flags | LH_SILENT | LH_PASSIVE);
   if (al == NULL)
     {
+      /* #### We should print the error message here. */
       logprintf (LOG_NOTQUIET,
-		 _("Unable to convert `%s' to a bind address.  Reverting to ANY.\n"),
-		 opt.bind_address);
+		 _("%s: unable to resolve bind address `%s'; disabling bind.\n"),
+		 exec_name, opt.bind_address);
       return 0;
     }
 
-- 
cgit v1.0-41-gc330

