diff --git a/ChangeLog b/ChangeLog
index 8ad5c4a..1eeb82c 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,9 @@
+2011-04-17  Giuseppe Scrivano  <gscrivano@gnu.org>
+
+	* configure.ac: Do not check the host_os twice and if windres is
+	available.
+	Reported by: Ray Satiro <raysatiro@yahoo.com>
+
 2011-04-16  Ray Satiro <raysatiro@yahoo.com>
 2011-04-16  Giuseppe Scrivano  <gscrivano@gnu.org>
 
diff --git a/configure.ac b/configure.ac
index a755a9d..88d498e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -220,8 +220,6 @@ WGET_NANOSLEEP
 WGET_POSIX_CLOCK
 WGET_NSL_SOCKET
 
-AM_CONDITIONAL(HAVE_WINDRES, env windres > /dev/null)
-
 dnl Deal with specific hosts
 case $host_os in
   *mingw32* )
@@ -256,25 +254,21 @@ AS_IF([test x"$with_ssl" = xopenssl], [
         then
           AC_CHECK_LIB(ssl32, SSL_connect,,
                     AC_MSG_ERROR([openssl not found: shared lib eay32 found but ssl32 not found]))
-		  AC_MSG_NOTICE([Enabling support for SSL via OpenSSL (shared)])
-		else
-
-        case $host_os in
-            *mingw32* )
-                LIBS+=' -lgdi32'
-            ;;
-        esac
-
-		    dnl fallback and test static libs
-        AC_CHECK_LIB(crypto, EVP_MD_CTX_init,,,)
-		    if test x"$ac_cv_lib_crypto_EVP_MD_CTX_init" != xno
-		    then
-            AC_CHECK_LIB(ssl, SSL_connect,, AC_MSG_ERROR([openssl not found]))
-            AC_MSG_NOTICE([Enabling support for SSL via OpenSSL])
-        else
-		        AC_MSG_ERROR([openssl not found: shared (eay32/ssl32) or static (crypto/ssl) libs needed])
-		    fi
-    fi
+		      AC_MSG_NOTICE([Enabling support for SSL via OpenSSL (shared)])
+		    else
+
+          LIBS+=' -lgdi32'
+
+	  	    dnl fallback and test static libs
+          AC_CHECK_LIB(crypto, EVP_MD_CTX_init,,,)
+		      if test x"$ac_cv_lib_crypto_EVP_MD_CTX_init" != xno
+		      then
+              AC_CHECK_LIB(ssl, SSL_connect,, AC_MSG_ERROR([openssl not found]))
+              AC_MSG_NOTICE([Enabling support for SSL via OpenSSL])
+          else
+		          AC_MSG_ERROR([openssl not found: shared (eay32/ssl32) or static (crypto/ssl) libs needed])
+		      fi
+      fi
 		
         dnl add zdll lib as dep for above tests?
       ;;
-- 
cgit v1.0-41-gc330

