diff --git a/src/ChangeLog b/src/ChangeLog
index c99218f..0dfd14d 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -3,6 +3,7 @@
 	* ftp.c: Use Gnulib's quote function for printing filenames and
 	such.
 	* connect.c: Likewise.
+	* convert.c: Likewise.
 
 2008-04-16  Steven Schubiger  <schubiger@gmail.com>
 
diff --git a/src/convert.c b/src/convert.c
index 2811bff..69d713f 100644
--- a/src/convert.c
+++ b/src/convert.c
@@ -230,8 +230,8 @@ convert_links (const char *file, struct urlpos *links)
      zeroes from the mmaped region.  */
   if (unlink (file) < 0 && errno != ENOENT)
     {
-      logprintf (LOG_NOTQUIET, _("Unable to delete `%s': %s\n"),
-                 file, strerror (errno));
+      logprintf (LOG_NOTQUIET, _("Unable to delete %s: %s\n"),
+                 quote (file), strerror (errno));
       read_file_free (fm);
       return;
     }
-- 
cgit v1.0-41-gc330

