diff --git a/src/iri.c b/src/iri.c
index 783aa33..44a262b 100644
--- a/src/iri.c
+++ b/src/iri.c
@@ -331,7 +331,7 @@ iri_free (struct iri *i)
 void
 set_uri_encoding (struct iri *i, char *charset)
 {
-  logprintf (LOG_VERBOSE, "[ uri = `%s'\n", charset);
+  DEBUGP (("[IRI uri = `%s'\n", quote (charset)));
   if (opt.encoding_remote)
     return;
   if (i->uri_encoding)
@@ -347,7 +347,7 @@ set_uri_encoding (struct iri *i, char *charset)
 void
 set_content_encoding (struct iri *i, char *charset)
 {
-  logprintf (LOG_VERBOSE, "[ content = `%s'\n", charset);
+  DEBUGP (("[IRI content = %s\n", quote (charset)));
   if (opt.encoding_remote)
     return;
   if (i->content_encoding)
diff --git a/src/recur.c b/src/recur.c
index e2f58d1..aa83e9a 100644
--- a/src/recur.c
+++ b/src/recur.c
@@ -116,7 +116,8 @@ url_enqueue (struct url_queue *queue, struct iri *i,
   DEBUGP (("Queue count %d, maxcount %d.\n", queue->count, queue->maxcount));
 
   if (i)
-    printf ("[Enqueuing %s with %s\n", url, i->uri_encoding);
+    DEBUGP (("[IRI Enqueuing %s with %s\n", quote (url),
+               quote (i->uri_encoding)));
 
   if (queue->tail)
     queue->tail->next = qel;
diff --git a/src/retr.c b/src/retr.c
index ae8ef3e..691b8f5 100644
--- a/src/retr.c
+++ b/src/retr.c
@@ -635,7 +635,8 @@ retrieve_url (const char *origurl, char **file, char **newloc,
       return URLERROR;
     }
 
-  printf ("[Retrieving %s with %s (UTF-8=%d)\n", url, iri->uri_encoding, iri->utf8_encode);
+  DEBUGP (("[IRI Retrieving %s with %s (UTF-8=%d)\n", quote (url),
+             quote (iri->uri_encoding), iri->utf8_encode));
 
   if (!refurl)
     refurl = opt.referer;
@@ -788,7 +789,7 @@ retrieve_url (const char *origurl, char **file, char **newloc,
   if (!(*dt & RETROKF) && iri->utf8_encode)
     {
       iri->utf8_encode = false;
-      printf ("[Fallbacking to non-utf8 for `%s'\n", url);
+      DEBUGP (("[IRI Fallbacking to non-utf8 for %s\n", quote (url)));
       goto second_try;
     }
 
-- 
cgit v1.0-41-gc330

