diff --git a/src/ChangeLog b/src/ChangeLog
index f028933..52ebd6f 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,10 @@
 2005-07-04  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* xmalloc.c (debugging_free): Prefix hex pointer value with "0x"
+	when printing.
+
+2005-07-04  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* utils.c (NEXT_BASE64_CHAR): Rename to NEXT_CHAR and simplify to
 	get the next non-whitespace character.
 
diff --git a/src/xmalloc.c b/src/xmalloc.c
index bf4b9a3..4368faa 100644
--- a/src/xmalloc.c
+++ b/src/xmalloc.c
@@ -357,7 +357,7 @@ debugging_free (void *ptr, const char *source_file, int source_line)
     }
   if (!unregister_ptr (ptr))
     {
-      fprintf (stderr, "%s: bad xfree(%0*lx) at %s:%d\n",
+      fprintf (stderr, "%s: bad xfree(0x%0*lx) at %s:%d\n",
 	       exec_name, PTR_FORMAT (ptr), source_file, source_line);
       abort ();
     }
-- 
cgit v1.0-41-gc330

