diff --git a/src/ChangeLog b/src/ChangeLog
index 74f3653..8fe0e34 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,11 +1,15 @@
+2008-09-09  Gisle Vanem  <gvanem@broadpark.no>
+
+	* url.c (url_error): Use aprintf, not asprintf.
+	
 2008-09-09  Micah Cowan  <micah@cowan.name>
 
 	* init.c (home_dir): Save the calculated value for home,
 	to avoid duplicated work on repeated calls.
 	(wgetrc_file_name) [WINDOWS]: Define and initialize home var.
 
-	* build_info.c: Remove unnecessary extern vars system_wgetrc and
-	locale_dir.
+	* build_info.c, main.c: Remove unnecessary extern vars
+	system_wgetrc and locale_dir.
 
 	* main.c: Define program_name for lib/error.c.
 
diff --git a/src/url.c b/src/url.c
index 3f4b899..fea9558 100644
--- a/src/url.c
+++ b/src/url.c
@@ -900,9 +900,9 @@ url_error (const char *url, int error_code)
       if ((p = strchr (scheme, ':')))
         *p = '\0';
       if (!strcasecmp (scheme, "https"))
-        asprintf (&error, _("HTTPS support not compiled in"));
+        error = aprintf (_("HTTPS support not compiled in"));
       else
-        asprintf (&error, _(parse_errors[error_code]), quote (scheme));
+        error = aprintf (_(parse_errors[error_code]), quote (scheme));
       xfree (scheme);
 
       return error;
-- 
cgit v1.0-41-gc330

