diff --git a/src/ChangeLog b/src/ChangeLog
index 3945b84..698fdb5 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -8,6 +8,7 @@
 	* ftp-opie.c: Likewise.
 	* gnutls.c: Likewise.
 	* init.c: Likewise.
+	* log.c: Likewise.
 
 2008-04-16  Steven Schubiger  <schubiger@gmail.com>
 
diff --git a/src/log.c b/src/log.c
index 078a7ac..f7b3dca 100644
--- a/src/log.c
+++ b/src/log.c
@@ -781,8 +781,8 @@ redirect_output (void)
   logfp = unique_create (DEFAULT_LOGFILE, false, &logfile);
   if (logfp)
     {
-      fprintf (stderr, _("\n%s received, redirecting output to `%s'.\n"),
-               redirect_request_signal_name, logfile);
+      fprintf (stderr, _("\n%s received, redirecting output to %s.\n"),
+               redirect_request_signal_name, quote (logfile));
       xfree (logfile);
       /* Dump the context output to the newly opened log.  */
       log_dump_context ();
-- 
cgit v1.0-41-gc330

