diff --git a/windows/config-compiler.h b/windows/config-compiler.h
index 806f9a4..648ac50 100644
--- a/windows/config-compiler.h
+++ b/windows/config-compiler.h
@@ -73,8 +73,7 @@ so, delete this exception statement from your version.  */
    from all import libraries. */
 #ifdef ENABLE_IPV6
 # undef gai_strerror
-#  define gai_strerror windows_strerror
-# endif
+# define gai_strerror windows_strerror
 #endif
 
 /* -------------------- */
-- 
cgit v1.0-41-gc330

