diff --git a/testenv/ChangeLog b/testenv/ChangeLog
index 814a374..fb34a97 100644
--- a/testenv/ChangeLog
+++ b/testenv/ChangeLog
@@ -1,3 +1,7 @@
+2014-11-21  Tim Ruehsen <tim.ruehsen@gmx.de>
+
+	* server/http/http_server.py: Fixed typo Blackisted to Blacklisted
+
 2014-11-19  Darshit Shah  <darnir@gmail.com>
 
 	* Test-504.py: Add new test case to show how Wget handles 504 Gateway
diff --git a/testenv/server/http/http_server.py b/testenv/server/http/http_server.py
index e28759a..52c4913 100644
--- a/testenv/server/http/http_server.py
+++ b/testenv/server/http/http_server.py
@@ -336,7 +336,7 @@ class _Handler (BaseHTTPRequestHandler):
         for header_line in rej_headers:
             header_recd = self.headers.get (header_line)
             if header_recd is not None and header_recd == rej_headers[header_line]:
-                self.send_error (400, 'Blackisted Header ' + header_line + ' received')
+                self.send_error (400, 'Blacklisted Header ' + header_line + ' received')
                 self.finish_headers ()
                 raise ServerError ("Header " + header_line + ' received')
 
-- 
cgit v1.0-41-gc330

