diff --git a/src/ChangeLog b/src/ChangeLog
index 0fa409a..c3863fd 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,10 @@
 2001-12-06  Hrvoje Niksic  <hniksic@arsdigita.com>
 
+	* gen_sslfunc.c (ssl_init_prng): Make the printed message
+	translatable.
+
+2001-12-06  Hrvoje Niksic  <hniksic@arsdigita.com>
+
 	* url.c (scheme_disable): New function.
 
 	* main.c (main): Call ssl_init_prng from here rather than from
diff --git a/src/gen_sslfunc.c b/src/gen_sslfunc.c
index a782727..9b3ebe0 100644
--- a/src/gen_sslfunc.c
+++ b/src/gen_sslfunc.c
@@ -105,7 +105,8 @@ ssl_init_prng (void)
 
   if (RAND_status () == 0)
     {
-      logprintf (LOG_NOTQUIET, "Could not seed OpenSSL PRNG; disabling SSL.\n");
+      logprintf (LOG_NOTQUIET,
+		 _("Could not seed OpenSSL PRNG; disabling SSL.\n"));
       scheme_disable (SCHEME_HTTPS);
     }
 #endif /* SSLEAY_VERSION_NUMBER >= 0x00905100 */
-- 
cgit v1.0-41-gc330

