diff --git a/src/convert.c b/src/convert.c
index b199956..d89c380 100644
--- a/src/convert.c
+++ b/src/convert.c
@@ -454,7 +454,7 @@ write_backup_file (const char *file, downloaded_file_t downloaded_file_return)
          because convert_file() is called from convert_all_links at
          the end of the retrieval with a freshly built new urlpos
          list.
-	 -- Hrvoje Niksic <hniksic@arsdigita.com>
+	 -- Hrvoje Niksic <hniksic@xemacs.org>
       */
       converted_file_ptr = xmalloc(sizeof(*converted_file_ptr));
       converted_file_ptr->string = xstrdup(file);  /* die on out-of-mem. */
diff --git a/src/main.c b/src/main.c
index 6b0c6d1..acbb992 100644
--- a/src/main.c
+++ b/src/main.c
@@ -515,13 +515,13 @@ hpVqvdkKsxmNWrHSLcFbEY:G:g:T:U:O:l:n:i:o:a:t:D:A:R:P:B:e:Q:X:I:w:C:",
 	case 'V':
 	  printf ("GNU Wget %s\n\n", version_string);
 	  printf ("%s", _("\
-Copyright (C) 1995, 1996, 1997, 1998, 2000, 2001 Free Software Foundation, Inc.\n"));
+Copyright (C) 2003 Free Software Foundation, Inc.\n"));
 	  printf ("%s", _("\
 This program is distributed in the hope that it will be useful,\n\
 but WITHOUT ANY WARRANTY; without even the implied warranty of\n\
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n\
 GNU General Public License for more details.\n"));
-	  printf (_("\nOriginally written by Hrvoje Niksic <hniksic@arsdigita.com>.\n"));
+	  printf (_("\nOriginally written by Hrvoje Niksic <hniksic@xemacs.org>.\n"));
 	  exit (0);
 	  break;
 	case 'v':
diff --git a/src/snprintf.c b/src/snprintf.c
index 4d1c3dd..f8dffa4 100644
--- a/src/snprintf.c
+++ b/src/snprintf.c
@@ -62,7 +62,7 @@
  *    fixed return value to comply with C99
  *    fixed handling of snprintf(NULL, ...)
  *
- *  Hrvoje Niksic <hniksic@arsdigita.com> 2000-11-04
+ *  Hrvoje Niksic <hniksic@xemacs.org> 2000-11-04
  *    include <config.h> instead of "config.h".
  *    moved TEST_SNPRINTF stuff out of HAVE_SNPRINTF ifdef.
  *    include <stdio.h> for NULL.
-- 
cgit v1.0-41-gc330

