diff --git a/ChangeLog b/ChangeLog
index 89f122b..5a8f750 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,5 +1,14 @@
 2003-10-26  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* windows/config.h.bor: DEBUG is now ENABLE_DEBUG.  Borland has
+	snprintf, but not u_int32_t.
+
+	* windows/Makefile.src.bor (OBJS): Use convert.c.
+
+	From Chin-yuan Kuo.
+
+2003-10-26  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* windows/config.h.mingw: Ditto.
 
 	* windows/Makefile.top.mingw: Ditto.
diff --git a/windows/Makefile.src.bor b/windows/Makefile.src.bor
index 6cca91c..60f7292 100644
--- a/windows/Makefile.src.bor
+++ b/windows/Makefile.src.bor
@@ -1,5 +1,5 @@
 ## Compiler, linker, and lib stuff
-## Makefile for use with watcom win95/winnt executable.
+## Makefile for use with Borland C++ for Win32 executable.
 
 CC=bcc32
 LINK=ilink32
@@ -8,7 +8,7 @@ LFLAGS=
 CFLAGS=-DWINDOWS -DHAVE_CONFIG_H -I. -w- -O2
 
 ## variables
-OBJS=cmpt.obj connect.obj ftp.obj ftp-basic.obj  \
+OBJS=cmpt.obj connect.obj convert.obj ftp.obj ftp-basic.obj  \
       ftp-ls.obj ftp-opie.obj getopt.obj headers.obj host.obj html-parse.obj html-url.obj \
       http.obj init.obj log.obj main.obj gnu-md5.obj netrc.obj rbuf.obj  \
       safe-ctype.obj hash.obj progress.obj gen-md5.obj cookies.obj \
@@ -20,6 +20,7 @@ wget.exe: $(OBJS)
   $(LINK) @&&|
 $(LFLAGS) -Tpe -ap -c +
 $(LIBDIR)\c0x32.obj+
+convert.obj+
 cookies.obj+
 hash.obj+
 safe-ctype.obj+
diff --git a/windows/config.h.bor b/windows/config.h.bor
index a51e25a..f5153dc 100644
--- a/windows/config.h.bor
+++ b/windows/config.h.bor
@@ -29,6 +29,8 @@
 #ifndef CONFIG_H
 #define CONFIG_H
 
+#define HAVE_SNPRINTF
+
 #define HAVE_MEMMOVE
 #define ftruncate chsize
 #define inline __inline
@@ -76,7 +78,7 @@
 #define USE_DIGEST 1
 
 /* Define if you want the debug output support compiled in.  */
-#define DEBUG
+#define ENABLE_DEBUG
 
 /* Define if you have sys/time.h header. */
 #undef HAVE_SYS_TIME_H
@@ -196,6 +198,6 @@
 #define socklen_t int
 
 /* Define if you have u_int32_t.  */
-#define HAVE_U_INT32_T 1
+/* #undef HAVE_U_INT32_T */
 
 #endif /* CONFIG_H */
-- 
cgit v1.0-41-gc330

