diff --git a/ChangeLog b/ChangeLog
index c142e0c..5f13bc1 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,5 +1,9 @@
 2014-12-05  Tim Ruehsen <tim.ruehsen@gmx.de>
 
+	* configure.ac: Fix message for when GnuTLS has not been found.
+
+2014-12-05  Tim Ruehsen <tim.ruehsen@gmx.de>
+
 	* bootstrap.conf: Add modules fnmatch, memrchr, strcase, strptime,
 	strtoll, timegm
 	Remove module strcasecmp
diff --git a/configure.ac b/configure.ac
index 43949a2..a366e83 100644
--- a/configure.ac
+++ b/configure.ac
@@ -422,7 +422,7 @@ AS_IF([test x"$with_ssl" = xopenssl], [
         AC_LIBOBJ([gnutls])
         LIBS="$LIBGNUTLS $LIBS"
       else
-        AC_MSG_ERROR([--with-ssl=gnutls was given, but GNUTLS is not available.])
+        AC_MSG_ERROR([GnuTLS has not been found. Use --with-ssl=openssl if you explicitly want OpenSSL.])
       fi
     ])
 
-- 
cgit v1.0-41-gc330

