diff --git a/doc/ChangeLog b/doc/ChangeLog
index 7157a76..44c5cd8 100644
--- a/doc/ChangeLog
+++ b/doc/ChangeLog
@@ -1,3 +1,8 @@
+2008-05-12  Micah Cowan  <micah@cowan.name>
+
+	* wget.texi (Download Options): -N with -O downgraded to a
+	warning.
+
 2008-04-27  Micah Cowan  <micah@cowan.name>
 
 	* wget.texi (Download Options) <-O>: Elaborate on why certain
diff --git a/doc/wget.texi b/doc/wget.texi
index 7a949bf..3c29f00 100644
--- a/doc/wget.texi
+++ b/doc/wget.texi
@@ -553,9 +553,8 @@ immediately, and @emph{all} downloaded content will be written there.
 
 For this reason, @samp{-N} (for timestamp-checking) is not supported
 in combination with @samp{-O}: since @var{file} is always newly
-created, it will always have a very new timestamp. Contrary to some
-users' expectations, the combination has never worked, and as of
-version 1.11, it results in an error.
+created, it will always have a very new timestamp. A warning will be
+issued if this combination is used.
 
 Similarly, using @samp{-r} or @samp{-p} with @samp{-O} may not work as
 you expect: Wget won't just download the first file to @var{file} and
diff --git a/src/ChangeLog b/src/ChangeLog
index d618697..7d8e5ec 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,7 @@
+2008-05-12  Micah Cowan  <micah@cowan.name>
+
+	* main.c (main): Downgrade "-N with -O" to a warning.
+
 2008-04-30  Micah Cowan  <micah@cowan.name>
 
 	* progress.c (create_image): Fix glitch where too many spaces are
diff --git a/src/main.c b/src/main.c
index 1ad0a48..fbdafd2 100644
--- a/src/main.c
+++ b/src/main.c
@@ -900,10 +900,9 @@ will be placed in the single file you specified.\n\n"));
         }
       if (opt.timestamping)
         {
-          fputs (_("\
-Cannot specify -N if -O is given. See the manual for details.\n\n"), stdout);
-          print_usage ();
-          exit (1);
+          logprintf (LOG_NOTQUIET, "%s", _("\
+WARNING: timestamping does nothing in combination with -O. See the manual\n\
+for details.\n\n"));
         }
     }
 
-- 
cgit v1.0-41-gc330

