diff --git a/src/ChangeLog b/src/ChangeLog
index 070aa95..0f7c069 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,9 @@
 2001-11-23  Hrvoje Niksic  <hniksic@arsdigita.com>
 
+	* progress.c (create_image): Don't translate "%ld ".
+
+2001-11-23  Hrvoje Niksic  <hniksic@arsdigita.com>
+
 	* progress.c (bar_set_params): Allow the user to force the use of
 	the bar.
 
diff --git a/src/progress.c b/src/progress.c
index 6c721cd..929aa56 100644
--- a/src/progress.c
+++ b/src/progress.c
@@ -527,7 +527,7 @@ create_image (struct bar_progress *bp, long dltime)
     }
 
   /* "12376 " */
-  sprintf (p, _("%ld "), size);
+  sprintf (p, "%ld ", size);
   p += strlen (p);
 
   /* "ETA: xx:xx:xx" */
-- 
cgit v1.0-41-gc330

