diff --git a/src/ChangeLog b/src/ChangeLog
index 73f91bb..eeff748 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,11 @@
 2005-05-12  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* ftp-ls.c (ftp_index): Use %d to print the port number, which is
+	now int.
+	From Steven M. Schweda's VMS patches.
+
+2005-05-12  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* url.c (rewrite_shorthand_url): Don't rewrite "https://host" to
 	"ftp://https//host" when SSL is not used.
 
diff --git a/src/ftp-ls.c b/src/ftp-ls.c
index dd703df..6f6eeac 100644
--- a/src/ftp-ls.c
+++ b/src/ftp-ls.c
@@ -926,7 +926,7 @@ ftp_index (const char *file, struct url *u, struct fileinfo *f)
 	  break;
 	}
       htclfile = html_quote_string (f->name);
-      fprintf (fp, "<a href=\"ftp://%s%s:%hu", upwd, u->host, u->port);
+      fprintf (fp, "<a href=\"ftp://%s%s:%d", upwd, u->host, u->port);
       if (*u->dir != '/')
 	putc ('/', fp);
       fprintf (fp, "%s", u->dir);
-- 
cgit v1.0-41-gc330

