diff --git a/src/ChangeLog b/src/ChangeLog
index 39390a5..07c99b9 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,7 @@
+2011-04-18  Giuseppe Scrivano  <gscrivano@gnu.org>
+
+	* openssl.c (openssl_read): Fix build error.
+
 2011-04-17  Giuseppe Scrivano  <gscrivano@gnu.org>
 
 	* wget.h [HAVE_LIBSSL32]: Define HAVE_SSL.
diff --git a/src/openssl.c b/src/openssl.c
index b42d698..a1b35b2 100644
--- a/src/openssl.c
+++ b/src/openssl.c
@@ -263,9 +263,8 @@ openssl_read (int fd, char *buf, int bufsize, void *arg)
   do
     ret = SSL_read (conn, buf, bufsize);
   while (ret == -1
-         && (SSL_get_error (conn, ret) == SSL_ERROR_WANT_READ
              || (SSL_get_error (conn, ret) == SSL_ERROR_SYSCALL
-                 && errno == EINTR)));
+                 && errno == EINTR));
 
   return ret;
 }
-- 
cgit v1.0-41-gc330

