diff --git a/src/log.c b/src/log.c
index 6f9b0c7..a3fe63c 100644
--- a/src/log.c
+++ b/src/log.c
@@ -344,7 +344,9 @@ get_warc_log_fp (void)
     return NULL;
   if (warclogfp)
     return warclogfp;
-  return NULL;
+  if (logfp)
+    return NULL;
+  return stderr;
 }
 
 /* Sets the file descriptor for the secondary log file.  */
diff --git a/src/main.c b/src/main.c
index f9759c3..ff0616b 100644
--- a/src/main.c
+++ b/src/main.c
@@ -37,7 +37,7 @@ as that of the covered work.  */
 #include <string.h>
 #include <signal.h>
 #include <spawn.h>
-#ifdef ENABLE_NLS
+#if defined(ENABLE_NLS) || defined(WINDOWS)
 # include <locale.h>
 #endif
 #include <assert.h>
@@ -74,6 +74,9 @@ as that of the covered work.  */
 #ifdef WINDOWS
 # include <io.h>
 # include <fcntl.h>
+#ifndef ENABLE_NLS
+# include <mbctype.h>
+#endif
 #endif
 
 #ifdef __VMS
@@ -148,6 +151,14 @@ i18n_initialize (void)
   /* Set the text message domain.  */
   bindtextdomain ("wget", LOCALEDIR);
   textdomain ("wget");
+#elif defined WINDOWS
+  char MBCP[16] = "";
+  int CP;
+
+  CP = _getmbcp(); /* Consider it's different from default. */
+  if (CP > 0)
+    snprintf(MBCP, sizeof(MBCP), ".%d", CP);
+  setlocale(LC_ALL, MBCP);
 #endif /* ENABLE_NLS */
 }
 
@@ -1529,6 +1540,9 @@ main (int argc, char **argv)
 
   nurl = argc - optind;
 
+  /* Initialize logging ASAP.  */
+  log_init (opt.lfilename, append_to_log);
+
   /* If we do not have Debug support compiled in AND Wget is invoked with the
    * --debug switch, instead of failing, we silently turn it into a no-op. For
    *  this no-op, we explicitly set opt.debug to false and hence none of the
@@ -1634,8 +1648,8 @@ for details.\n\n"));
            {
               /* Check if output file exists; if it does, exit. */
               logprintf (LOG_VERBOSE,
-                         _("File `%s' already there; not retrieving.\n"),
-                         opt.output_document);
+                         _("File %s already there; not retrieving.\n"),
+                         quote (opt.output_document));
               exit (WGET_EXIT_GENERIC_ERROR);
            }
     }
@@ -1892,9 +1906,6 @@ for details.\n\n"));
     }
   url[i] = NULL;
 
-  /* Initialize logging.  */
-  log_init (opt.lfilename, append_to_log);
-
   /* Open WARC file. */
   if (opt.warc_filename != 0)
     warc_init ();
-- 
cgit v1.0-41-gc330

