diff --git a/src/ChangeLog b/src/ChangeLog
index 4c6ab20..001b62e 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,6 +1,7 @@
 2011-04-04  Giuseppe Scrivano  <gscrivano@gnu.org>
 
 	* openssl.c [WINDOWS]: Include <w32sock.h>.
+	(openssl_read): Retry the read on SSL_ERROR_WANT_READ.
 
 	* host.c [WINDOWS]: Include <winsock2.h> and <ws2tcpip.h>.
 	Suggested by: Ray Satiro <raysatiro@yahoo.com>.
diff --git a/src/openssl.c b/src/openssl.c
index 60e794f..201d245 100644
--- a/src/openssl.c
+++ b/src/openssl.c
@@ -261,8 +261,10 @@ openssl_read (int fd, char *buf, int bufsize, void *arg)
   do
     ret = SSL_read (conn, buf, bufsize);
   while (ret == -1
-         && SSL_get_error (conn, ret) == SSL_ERROR_SYSCALL
-         && errno == EINTR);
+         && (SSL_get_error (conn, ret) == SSL_ERROR_WANT_READ
+             || (SSL_get_error (conn, ret) == SSL_ERROR_SYSCALL
+                 && errno == EINTR)));
+
   return ret;
 }
 
-- 
cgit v1.0-41-gc330

