diff --git a/src/ChangeLog b/src/ChangeLog
index 698fdb5..7292f47 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -9,6 +9,7 @@
 	* gnutls.c: Likewise.
 	* init.c: Likewise.
 	* log.c: Likewise.
+	* mswindows.c: Likewise.
 
 2008-04-16  Steven Schubiger  <schubiger@gmail.com>
 
diff --git a/src/mswindows.c b/src/mswindows.c
index 2a14696..cdfea2d 100644
--- a/src/mswindows.c
+++ b/src/mswindows.c
@@ -291,7 +291,7 @@ fake_fork (void)
 
   printf (_("Continuing in background, pid %lu.\n"), pi.dwProcessId);
   if (info->logfile_changed)
-    printf (_("Output will be written to `%s'.\n"), info->lfilename);
+    printf (_("Output will be written to %s.\n"), quote (info->lfilename));
 
   UnmapViewOfFile (info);
 
-- 
cgit v1.0-41-gc330

