diff --git a/src/metalink.c b/src/metalink.c
index 38613a7..fd6d0e2 100644
--- a/src/metalink.c
+++ b/src/metalink.c
@@ -566,7 +566,8 @@ badhash_suffix (char *name)
   bhash = concat_strings (name, ".badhash", (char *)0);
   uname = unique_name (bhash, false);
 
-  logprintf (LOG_VERBOSE, _("Renaming ‘%s’ to ‘%s’.\n"), name, uname);
+  logprintf (LOG_VERBOSE, _("Renaming %s to %s.\n"),
+             quote_n (0, name), quote_n (1, uname));
 
   if (link (name, uname))
     logprintf (LOG_NOTQUIET, "link: %s\n", strerror (errno));
-- 
cgit v1.0-41-gc330

