diff --git a/src/ChangeLog b/src/ChangeLog
index 435026e..32eebda 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,7 @@
+2014-11-12  Tim Ruehsen  <tim.ruehsen@gmx.de>
+
+	* openssl.c (ssl_init): Fix error handling for CRL loading
+
 2014-11-12  Witchakorn Kamolpornwijit <witchakk@mit.edu> (tiny change)
 
 	* html-parse.c (map_html_tags): Fix range check
diff --git a/src/openssl.c b/src/openssl.c
index 371913c..ae06edc 100644
--- a/src/openssl.c
+++ b/src/openssl.c
@@ -258,14 +258,10 @@ ssl_init (void)
     {
       X509_STORE *store = SSL_CTX_get_cert_store (ssl_ctx);
       X509_LOOKUP *lookup;
-      int rc;
 
       if (!(lookup = X509_STORE_add_lookup (store, X509_LOOKUP_file ()))
-          || (!(rc = X509_load_crl_file (lookup, opt.crl_file, X509_FILETYPE_PEM))))
-        {
-          logprintf (LOG_NOTQUIET, _("ERROR: Failed to load CRL file '%s': (%d)\n"), opt.crl_file, rc);
-          return false;
-        }
+          || (!X509_load_crl_file (lookup, opt.crl_file, X509_FILETYPE_PEM)))
+        goto error;
 
       X509_STORE_set_flags (store, X509_V_FLAG_CRL_CHECK | X509_V_FLAG_CRL_CHECK_ALL);
     }
-- 
cgit v1.0-41-gc330

