diff --git a/src/ChangeLog b/src/ChangeLog
index d949c29..459a3b8 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,9 @@
 2001-12-09  Hrvoje Niksic  <hniksic@arsdigita.com>
 
+	* main.c (main): Remove stray debugging message.
+
+2001-12-09  Hrvoje Niksic  <hniksic@arsdigita.com>
+
 	* progress.c (create_image): Fix ETA padding when hours are prined.
 
 2001-12-09  Hrvoje Niksic  <hniksic@arsdigita.com>
diff --git a/src/main.c b/src/main.c
index 49e979f..765ee73 100644
--- a/src/main.c
+++ b/src/main.c
@@ -738,10 +738,7 @@ Can't timestamp and not clobber old files at the same time.\n"));
     {
       char *rewritten = rewrite_shorthand_url (argv[optind]);
       if (rewritten)
-	{
-	  printf ("Converted %s to %s\n", argv[optind], rewritten);
-	  url[i] = rewritten;
-	}
+	url[i] = rewritten;
       else
 	url[i] = xstrdup (argv[optind]);
     }
-- 
cgit v1.0-41-gc330

