diff --git a/src/wget.h b/src/wget.h
index bfe60c8..7ab17b7 100644
--- a/src/wget.h
+++ b/src/wget.h
@@ -243,7 +243,7 @@ typedef double SUM_SIZE_INT;
 } while (0)
 
 /* Used to print pointers (usually for debugging).  Print pointers
-   using printf ("%0*lx", PTR_FORMAT (p)).  (%p is too unpredictable;
+   using printf ("0x%0*lx", PTR_FORMAT (p)).  (%p is too unpredictable;
    some implementations prepend 0x, while some don't, and most don't
    0-pad the address.)  */
 #define PTR_FORMAT(p) (int) (2 * sizeof (void *)), (unsigned long) (p)
-- 
cgit v1.0-41-gc330

