diff --git a/src/ChangeLog b/src/ChangeLog
index e6c060d..b7225ad 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,9 @@
 2001-11-27  Hrvoje Niksic  <hniksic@arsdigita.com>
 
+	* progress.c (bar_create): Print two newlines.
+
+2001-11-27  Hrvoje Niksic  <hniksic@arsdigita.com>
+
 	* cmpt.c (random): New function, a simple-minded replacement for
 	random() on systems that don't have it.
 
diff --git a/src/progress.c b/src/progress.c
index 7c9bba1..efc03d6 100644
--- a/src/progress.c
+++ b/src/progress.c
@@ -404,7 +404,7 @@ bar_create (long initial, long total)
   bp->width = screen_width;
   bp->buffer = xmalloc (bp->width + 1);
 
-  logputs (LOG_VERBOSE, "\n");
+  logputs (LOG_VERBOSE, "\n\n");
 
   create_image (bp, 0);
   display_image (bp->buffer);
-- 
cgit v1.0-41-gc330

