diff --git a/src/ChangeLog b/src/ChangeLog
index 270d937..a0127f1 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,15 @@
 2003-11-15  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* http.c (persistent_available_p): Correctly specify the endpoint
+	argument to socket_ip_address.
+	(gethttp): When printing the "reusing connection to..." message,
+	specify the host name of the reused connection, not the current
+	host name.  That makes more sense because it provides a useful
+	piece of information -- we know to which host we're supposed to
+	connect anyway!
+
+2003-11-15  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* init.c: Make sure the options are in alphabetic order!
 
 	* host.c (lookup_host): Merge lookup_host_passive and lookup_host
diff --git a/src/http.c b/src/http.c
index 71006d1..d594e85 100644
--- a/src/http.c
+++ b/src/http.c
@@ -460,7 +460,7 @@ persistent_available_p (const char *host, int port, int ssl,
 	 resolves to, pconn.socket is for all intents and purposes
 	 already talking to HOST.  */
 
-      if (!socket_ip_address (pconn.socket, &ip, 0))
+      if (!socket_ip_address (pconn.socket, &ip, ENDPOINT_PEER))
 	{
 	  /* Can't get the peer's address -- something must be very
 	     wrong with the connection.  */
@@ -739,10 +739,8 @@ gethttp (struct url *u, struct http_stat *hs, int *dt, struct url *proxy)
     }
   else
     {
-      logprintf (LOG_VERBOSE, _("Reusing connection to %s:%hu.\n"),
-		 conn->host, conn->port);
-      /* #### pc_last_fd should be accessed through an accessor
-         function.  */
+      logprintf (LOG_VERBOSE, _("Reusing existing connection to %s:%d.\n"),
+		 pconn.host, pconn.port);
       sock = pconn.socket;
       using_ssl = pconn.ssl;
       DEBUGP (("Reusing fd %d.\n", sock));
-- 
cgit v1.0-41-gc330

