diff --git a/ChangeLog b/ChangeLog
index 3f5f3c3..5b41094 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,9 @@
+2005-08-26  Stepan Kasal  <kasal@ucw.cz>
+
+	* configure.in: Abort configure if --with-ssl given but SSL
+	unavailable.  Use HAVE_LIBSSL and HAVE_LIBGNUTLS symbols provided
+	by AC_LIB_HAVE_LINKFLAGS instead of inventing new ones.
+
 2005-08-11  Hrvoje Niksic  <hniksic@xemacs.org>
 
 	* configure.in: Check for strtoll and strtoimax.
diff --git a/configure.in b/configure.in
index 2a6f416..b2075c7 100644
--- a/configure.in
+++ b/configure.in
@@ -248,15 +248,15 @@ then
   if test x"$LIBGNUTLS" != x
   then
     AC_MSG_NOTICE([compiling in support for SSL via GnuTLS])
-    AC_DEFINE([HAVE_GNUTLS], 1,
-	      [Define if support for the GnuTLS library is being compiled in.])
     SSL_OBJ='gnutls.o'
+  else
+    AC_MSG_ERROR([--with-ssl=gnutls was given, but GNUTLS is not available.])
   fi
-else
+elif test x"$with_ssl" != xno; then
   dnl As of this writing (OpenSSL 0.9.6), the libcrypto shared library
   dnl doesn't record its dependency on libdl, so we need to make sure
   dnl -ldl ends up in LIBS on systems that have it.  Most OSes use
-  dnl dlopen(), but HP-UX uses dnl shl_load().
+  dnl dlopen(), but HP-UX uses shl_load().
   AC_CHECK_LIB(dl, dlopen, [], [
     AC_CHECK_LIB(dl, shl_load)
   ])
@@ -274,9 +274,10 @@ else
   if test x"$LIBSSL" != x
   then
     AC_MSG_NOTICE([compiling in support for SSL via OpenSSL])
-    AC_DEFINE([HAVE_OPENSSL], 1,
-	      [Define if support for the OpenSSL library is being compiled in.])
     SSL_OBJ='openssl.o'
+  elif x"$with_ssl" != x
+  then
+    AC_MSG_ERROR([--with-ssl was given, but SSL is not available.])
   fi
 fi
 
diff --git a/src/wget.h b/src/wget.h
index 5abce27..bfcaf8d 100644
--- a/src/wget.h
+++ b/src/wget.h
@@ -40,7 +40,8 @@ so, delete this exception statement from your version.  */
 # define NDEBUG
 #endif
 
-#if defined HAVE_OPENSSL || defined HAVE_GNUTLS
+/* Is OpenSSL or GNUTLS available? */
+#if defined HAVE_LIBSSL || defined HAVE_LIBGNUTLS
 # define HAVE_SSL
 #endif
 
-- 
cgit v1.0-41-gc330

