diff --git a/src/ChangeLog b/src/ChangeLog
index 33bd319..56d6def 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,7 @@
+2011-05-23  Giuseppe Scrivano  <gscrivano@gnu.org>
+
+	* mswindows.h [NEED_GAI_STRERROR]: remove definition for gai_strerror.
+
 2011-05-15  Ray Satiro <raysatiro@yahoo.com>
 
 	* connect.c (select_fd) [WINDOWS]: Ensure the socket is blocking.
diff --git a/src/mswindows.h b/src/mswindows.h
index e0a20e4..9b91aaf 100644
--- a/src/mswindows.h
+++ b/src/mswindows.h
@@ -113,11 +113,6 @@ typedef __int64 wgint;
 const char *inet_ntop (int, const void *, char *, socklen_t);
 #endif
 
-#ifdef NEED_GAI_STRERROR
-# undef gai_strerror
-# define gai_strerror strerror
-#endif
-
 /* ioctl needed by set_windows_fd_as_blocking_socket() */
 #include <sys/ioctl.h>
 
-- 
cgit v1.0-41-gc330

