diff --git a/src/connect.c b/src/connect.c
index 661a6ba..a343f66 100644
--- a/src/connect.c
+++ b/src/connect.c
@@ -366,7 +366,7 @@ connect_to_ip (const ip_address *ip, int port, const char *print)
     if (sock >= 0)
       fd_close (sock);
     if (print)
-      logprintf (LOG_VERBOSE, _("failed: %s.\n"), strerror (errno));
+      logprintf (LOG_NOTQUIET, _("failed: %s.\n"), strerror (errno));
     errno = save_errno;
     return -1;
   }
-- 
cgit v1.0-41-gc330

