diff --git a/ChangeLog b/ChangeLog
index 0da0206..a95f837 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,29 @@
+2004-05-09  David Fritz  <zeroxdf@att.net>
+
+	* windows/Makefile.src.bor: Fix broken build rule.  Add clean target.
+
+	* windows/Makefile.top.bor: Use tabs instead of spaces.  Ignore
+	errors in clean rules.  Use lowercase filenames when building
+	distribution .zip archive.
+
+	* windows/config.h.bor:  Don't define HAVE_UINT32_T.
+
+	* windows/Makefile.doc: Fix remaining instance of build rules
+	indented with spaces instead of tabs.
+
+	* windows/Makefile.src: Update copyright year.
+
+	* windows/Makefile.top: Update copyright year.
+
+	* windows/config.h.mingw (WGET_USE_STDARG, HAVE_SIG_ATOMIC_T): Define.
+
+	* windows/config.h.ms (HAVE_STRPBRK, HAVE_LIMITS_H)
+	(HAVE_LOCALE_H): Define.
+
+	* windows/Makefile.watcom: Add /I. to CFLAGS.  Remove reference to
+	specific Wget version from linker flags.  Add missing
+	dependencies.
+
 2004-02-09  David Fritz  <zeroxdf@att.net>
 
 	* configure.bat.in: Don't clear the screen.
diff --git a/windows/Makefile.doc b/windows/Makefile.doc
index cff926f..0c6ded5 100644
--- a/windows/Makefile.doc
+++ b/windows/Makefile.doc
@@ -1,5 +1,5 @@
 # Makefile for `wget' utility
-# Copyright (C) 1995, 1996, 1997 Free Software Foundation, Inc.
+# Copyright (C) 1995, 1996, 1997, 2004 Free Software Foundation, Inc.
 
 # This program is free software; you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
@@ -71,9 +71,9 @@ $(MAN): $(WGETPOD)
 	$(POD2MAN) --center="GNU Wget" --release="GNU Wget @VERSION@" $? > $@
 
 $(WGETHLP): $(WGETTEXI) $(SAMPLERCTEXI)
-        $(MAKEINFO) --no-validate --no-warn --force \
+	$(MAKEINFO) --no-validate --no-warn --force \
 --hpj wget.hpj --output wget.rtf $(WGETTEXI)
-        hcrtf -xn wget.hpj
+	hcrtf -xn wget.hpj
 
 clean:
 	$(RM) *.bak
diff --git a/windows/Makefile.src b/windows/Makefile.src
index 6a6b5af..c4946eb 100644
--- a/windows/Makefile.src
+++ b/windows/Makefile.src
@@ -1,5 +1,5 @@
 # Makefile for `wget' utility for MSVC
-# Copyright (C) 1995, 1996, 1997 Free Software Foundation, Inc.
+# Copyright (C) 1995, 1996, 1997, 2004 Free Software Foundation, Inc.
 
 # This program is free software; you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
diff --git a/windows/Makefile.src.bor b/windows/Makefile.src.bor
index 3457a10..e01c087 100644
--- a/windows/Makefile.src.bor
+++ b/windows/Makefile.src.bor
@@ -1,6 +1,7 @@
 ## Compiler, linker, and lib stuff
 ## Makefile for use with Borland C++ for Win32 executable.
 
+RM=-del
 CC=bcc32
 LINK=ilink32
 
@@ -21,7 +22,7 @@ wget.exe: $(OBJS)
   $(LINK) @&&|
 $(LFLAGS) -Tpe -ap -c +
 $(LIBDIR)\c0x32.obj+
-cmpt.obj
+cmpt.obj+
 connect.obj+
 convert.obj+
 cookies.obj+
@@ -50,8 +51,8 @@ safe-ctype.obj+
 url.obj+
 utils.obj+
 version.obj+
-xmalloc.obj+
-$<,$*
+xmalloc.obj,+
+$<,$*,+
 $(LIBDIR)\import32.lib+
 $(LIBDIR)\cw32.lib
 
@@ -59,6 +60,18 @@ $(LIBDIR)\cw32.lib
 
 |
 
+
+#
+# Dependencies for cleanup
+#
+
+clean:
+	$(RM) *.bak
+	$(RM) *.obj
+	$(RM) *.exe
+	$(RM) *.map
+	$(RM) *.tds
+
 o = .obj
 
 !include "..\windows\wget.dep"
diff --git a/windows/Makefile.top b/windows/Makefile.top
index 69b8df1..619c74b 100644
--- a/windows/Makefile.top
+++ b/windows/Makefile.top
@@ -1,5 +1,5 @@
 # Makefile for `Wget' utility
-# Copyright (C) 1995, 1996, 1997 Free Software Foundation, Inc.
+# Copyright (C) 1995, 1996, 1997, 2004 Free Software Foundation, Inc.
 
 # This program is free software; you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
diff --git a/windows/Makefile.top.bor b/windows/Makefile.top.bor
index 172e801..e264f6c 100644
--- a/windows/Makefile.top.bor
+++ b/windows/Makefile.top.bor
@@ -1,5 +1,5 @@
 # Makefile for `Wget' utility
-# Copyright (C) 1995, 1996, 1997 Free Software Foundation, Inc.
+# Copyright (C) 1995, 1996, 1997, 2004 Free Software Foundation, Inc.
 
 # This program is free software; you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
@@ -25,7 +25,7 @@
 # file, but you are not obligated to do so.  If you do not wish to do
 # so, delete this exception statement from your version.
 
-RM = del
+RM = -del
 CP = copy
 
 # flags passed to recursive makes in subdirectories
@@ -37,39 +37,39 @@ SUBDIRS = src doc #util
 all: Makefile $(SUBDIRS)
 
 $(SUBDIRS): 
-        cd $@
-        $(MAKE)
-        cd ..
+	cd $@
+	$(MAKE)
+	cd ..
 
 # install everything
 install:
-        echo Just do it.
+	echo Just do it.
 
 clean:		clean-recursive		clean-top
 distclean:	distclean-recursive	distclean-top
 realclean:	realclean-recursive	realclean-top
 
 clean-top:
-        $(RM) *.bak
-        $(RM) *.zip
+	$(RM) *.bak
+	$(RM) *.zip
 
 distclean-top: clean-top
-        $(RM) Makefile
-        $(RM) config.h
+	$(RM) Makefile
+	$(RM) config.h
 
 realclean-top: distclean-top
 
 clean-recursive distclean-recursive realclean-recursive:
-        cd src
-        $(MAKE) $(@:-recursive=)
-        cd ..\\doc
-        $(MAKE) $(@:-recursive=)
-        cd ..
+	cd src
+	$(MAKE) $(@:-recursive=)
+	cd ..\\doc
+	$(MAKE) $(@:-recursive=)
+	cd ..
 
 bindist: wget.zip
 
 wget.zip: $(SUBDIRS)
-        $(RM) wget.zip
-        zip -Djl9 wget.zip AUTHORS COPYING INSTALL MACHINES MAILING-LIST NEWS README DOC\\sample.wgetrc
-        zip -Dj9 wget.zip SRC\\WGET.EXE DOC\\WGET.HLP
+	$(RM) wget.zip
+	zip -Djl9 wget.zip AUTHORS COPYING INSTALL MACHINES MAILING-LIST NEWS README doc/sample.wgetrc
+	zip -Dj9 wget.zip src/wget.exe doc/wget.hlp
 
diff --git a/windows/Makefile.watcom b/windows/Makefile.watcom
index f235c64..bbd5a89 100644
--- a/windows/Makefile.watcom
+++ b/windows/Makefile.watcom
@@ -13,7 +13,7 @@ LINK=wlink
 .BEFORE
 	@SET INCLUDE=$(%WATCOM)\h;$(%WATCOM)\h\nt;.
 
-LFLAGS=op q sys nt op st=32767 op vers=1.7 op map op de 'GNU wget 1.7dev' de dw op symf
+LFLAGS=op q sys nt op st=32767 op map op de 'GNU Wget' de dw op symf
 #
 # op q              : quiet !
 # sys nt            : producing an NT app
@@ -25,7 +25,7 @@ LFLAGS=op q sys nt op st=32767 op vers=1.7 op map op de 'GNU wget 1.7dev' de dw
 # op symf           : place debug info in a separate wget.sym
 #
 
-CFLAGS=/zp4 /w4 /fpd /5s /fp5 /bm /mf /bt=nt /DWINDOWS /DHAVE_CONFIG_H
+CFLAGS=/zp4 /w4 /fpd /5s /fp5 /bm /mf /bt=nt /I. /DWINDOWS /DHAVE_CONFIG_H
 #                    ^^^^^^^^ wget will run on NT. We can safely assume a Pentium :-)
 # /zp4= pack structure members with this alignment
 # /d1 = line number debug info
@@ -54,7 +54,8 @@ CFLAGS+= /os /d2
 OBJS = cmpt.obj convert.obj connect.obj cookies.obj ftp.obj ftp-basic.obj  &
       ftp-ls.obj ftp-opie.obj getopt.obj hash.obj host.obj html-parse.obj html-url.obj  &
       http.obj init.obj log.obj main.obj gen-md5.obj gnu-md5.obj netrc.obj progress.obj  &
-      recur.obj res.obj retr.obj safe-ctype.obj url.obj utils.obj version.obj mswindows.obj
+      recur.obj res.obj retr.obj safe-ctype.obj url.obj utils.obj version.obj mswindows.obj &
+      xmalloc.obj
 
 LIBFILES = 
 #
diff --git a/windows/config.h.bor b/windows/config.h.bor
index 2778c10..3ce3276 100644
--- a/windows/config.h.bor
+++ b/windows/config.h.bor
@@ -1,5 +1,6 @@
 /* Configuration header file.
-   Copyright (C) 1995, 1996, 1997, 1998 Free Software Foundation, Inc.
+   Copyright (C) 1995, 1996, 1997, 1998, 2004
+   Free Software Foundation, Inc.
 
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
@@ -29,9 +30,9 @@
 #ifndef CONFIG_H
 #define CONFIG_H
 
-#define HAVE_SNPRINTF
+#define HAVE_SNPRINTF 1
 
-#define HAVE_MEMMOVE
+#define HAVE_MEMMOVE 1
 #define ftruncate chsize
 #define inline __inline
 
@@ -78,7 +79,7 @@
 #define USE_DIGEST 1
 
 /* Define if you want the debug output support compiled in.  */
-#define ENABLE_DEBUG
+#define ENABLE_DEBUG 1
 
 /* Define if you have sys/time.h header. */
 #undef HAVE_SYS_TIME_H
@@ -183,7 +184,7 @@
 #define HAVE_BUILTIN_MD5 1
 
 /* Define if you have the isatty function.  */
-#define HAVE_ISATTY
+#define HAVE_ISATTY 1
 
 /* Define to the length of short. */
 #define SIZEOF_SHORT 2
@@ -198,6 +199,6 @@
 #define socklen_t int
 
 /* Define if you have uint32_t.  */
-#define HAVE_UINT32_T */
+/* #define HAVE_UINT32_T */
 
 #endif /* CONFIG_H */
diff --git a/windows/config.h.mingw b/windows/config.h.mingw
index 1db10b2..d15e462 100644
--- a/windows/config.h.mingw
+++ b/windows/config.h.mingw
@@ -1,6 +1,5 @@
-/* src/config.h.  Generated by configure.  */
 /* Configuration header file.
-   Copyright (C) 1995, 1996, 1997, 1998, 2000, 2001
+   Copyright (C) 1995, 1996, 1997, 1998, 2000, 2001, 2004
    Free Software Foundation, Inc.
 
 This file is part of GNU Wget.
@@ -38,6 +37,9 @@ so, delete this exception statement from your version.  */
 /* For MinGW GCC/Win32, alloca is defined in the <malloc.h> header file. */
 #include <malloc.h>
 
+/* Flag that C files should include <stdarg.h>, not <varargs.h>.  */
+#define WGET_USE_STDARG
+
 /* Define to empty if the keyword does not work.  */
 /* #undef const */
 
@@ -86,6 +88,9 @@ so, delete this exception statement from your version.  */
 /* Define if you want the debug output support compiled in.  */
 #define ENABLE_DEBUG 1
 
+/* Define to 1 if the system has the type `sig_atomic_t'. */
+#define HAVE_SIG_ATOMIC_T 1
+
 /* Define if you have sys/time.h header.  */
 #define HAVE_SYS_TIME_H 1
 
diff --git a/windows/config.h.ms b/windows/config.h.ms
index a31b3e6..56fe81f 100644
--- a/windows/config.h.ms
+++ b/windows/config.h.ms
@@ -1,5 +1,6 @@
 /* Configuration header file.
-   Copyright (C) 1995, 1996, 1997, 1998 Free Software Foundation, Inc.
+   Copyright (C) 1995, 1996, 1997, 1998, 2004
+   Free Software Foundation, Inc.
 
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
@@ -55,7 +56,7 @@
 #define USE_DIGEST 1
 
 /* Define if you want the debug output support compiled in.  */
-#define ENABLE_DEBUG
+#define ENABLE_DEBUG 1
 
 /* Define if you have sys/time.h header. */
 #undef HAVE_SYS_TIME_H
@@ -91,10 +92,13 @@
 #define HAVE_STRERROR 1
 
 /* Define if you have the snprintf function.  */
-#define HAVE_SNPRINTF
+#define HAVE_SNPRINTF 1
 
 /* Define if you have the vsnprintf function.  */
-#define HAVE_VSNPRINTF
+#define HAVE_VSNPRINTF 1
+
+/* Define to 1 if you have the `strpbrk' function. */
+#define HAVE_STRPBRK 1
 
 /* Define if you have the strstr function. */
 #define HAVE_STRSTR 1
@@ -138,8 +142,8 @@
 
 /* None except Digital Mars have sleep/usleep functions */
 #if defined(__DMC__)
-#define HAVE_USLEEP
-#define HAVE_SLEEP
+#define HAVE_USLEEP 1
+#define HAVE_SLEEP 1
 #endif
 
 /* Define if you have the <utime.h> header file.  */
@@ -148,6 +152,12 @@
 /* Define if you have the <sys/utime.h> header file.  */
 #define HAVE_SYS_UTIME_H 1
 
+/* Define to 1 if you have the <limits.h> header file. */
+#define HAVE_LIMITS_H 1
+
+/* Define to 1 if you have the <locale.h> header file. */
+#define HAVE_LOCALE_H 1
+
 /* Define if you have the <sys/select.h> header file.  */
 #undef HAVE_SYS_SELECT_H
 
@@ -155,7 +165,7 @@
 #undef HAVE_PWD_H
 
 /* Define if you have the <signal.h> header file.  */
-#define HAVE_SIGNAL_H
+#define HAVE_SIGNAL_H 1
 
 /* Define to be the name of the operating system.  */
 #define OS_TYPE "Windows"
-- 
cgit v1.0-41-gc330

