diff --git a/src/ChangeLog b/src/ChangeLog
index 077d73c..fb93734 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,8 @@
+2005-07-05  Hrvoje Niksic  <hniksic@xemacs.org>
+
+	* http.c (gethttp): Don't print the request write error message
+	twice.
+
 2005-07-04  Hrvoje Niksic  <hniksic@xemacs.org>
 
 	* openssl.c (openssl_errstr): Instead of always using a large
diff --git a/src/http.c b/src/http.c
index 17dde39..3428f86 100644
--- a/src/http.c
+++ b/src/http.c
@@ -1486,8 +1486,6 @@ gethttp (struct url *u, struct http_stat *hs, int *dt, struct url *proxy)
 	  request_free (connreq);
 	  if (write_error < 0)
 	    {
-	      logprintf (LOG_VERBOSE, _("Failed writing to proxy: %s.\n"),
-			 fd_errstr (sock));
 	      CLOSE_INVALIDATE (sock);
 	      return WRITEFAILED;
 	    }
@@ -1556,8 +1554,6 @@ gethttp (struct url *u, struct http_stat *hs, int *dt, struct url *proxy)
 
   if (write_error < 0)
     {
-      logprintf (LOG_VERBOSE, _("Failed writing HTTP request: %s.\n"),
-		 fd_errstr (sock));
       CLOSE_INVALIDATE (sock);
       request_free (req);
       return WRITEFAILED;
-- 
cgit v1.0-41-gc330

