diff --git a/ChangeLog b/ChangeLog
index 7691445..1500669 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,8 @@
+2011-12-11  Giuseppe Scrivano  <gscrivano@gnu.org>
+
+	* util/trunc.c (main): Call `close' on the fd and check for errors.
+	Reported by: <dga@cs.cmu.edu>.
+
 2011-10-23  Giuseppe Scrivano  <gscrivano@gnu.org>
 
 	* bootstrap.conf (gnulib_modules): Include module `vsnprintf'.
diff --git a/util/trunc.c b/util/trunc.c
index 55cb19d..a5f1dcb 100644
--- a/util/trunc.c
+++ b/util/trunc.c
@@ -128,5 +128,11 @@ main (int argc, char *argv[])
       exit (EXIT_FAILURE);
     }
 
+  if (close (fd) < 0)
+    {
+      perror (PROGRAM_NAME ": close");
+      exit (EXIT_FAILURE);
+    }
+
   return 0;
 }
-- 
cgit v1.0-41-gc330

