diff --git a/src/ChangeLog b/src/ChangeLog
index d5b9f71..904d85a 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,10 @@
 2005-06-25  Hrvoje Niksic  <hniksic@xemacs.org>
 
+	* utils.c (SPRINTF_WGINT): The correct format is %I64d, not just
+	%I64.
+
+2005-06-25  Hrvoje Niksic  <hniksic@xemacs.org>
+
 	* http.c (http_loop): Don't warn about wildcards in HTTP URLs if
 	globbing isn't requested in the first place.
 
diff --git a/src/utils.c b/src/utils.c
index 5eb0d88..d0d7fbc 100644
--- a/src/utils.c
+++ b/src/utils.c
@@ -1345,7 +1345,7 @@ numdigit (wgint number)
 #elif SIZEOF_LONG_LONG >= SIZEOF_WGINT
 # define SPRINTF_WGINT(buf, n) sprintf (buf, "%lld", (long long) (n))
 #elif defined(WINDOWS)
-# define SPRINTF_WGINT(buf, n) sprintf (buf, "%I64", (__int64) (n))
+# define SPRINTF_WGINT(buf, n) sprintf (buf, "%I64d", (__int64) (n))
 #else
 # define SPRINTF_WGINT(buf, n) sprintf (buf, "%j", (intmax_t) (n))
 #endif
-- 
cgit v1.0-41-gc330

