diff --git a/src/ChangeLog b/src/ChangeLog
index 9f52e8e..df91013 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,9 @@
 2012-12-09  Giuseppe Scrivano  <gscrivano@gnu.org>
 
+	* main.c (main): Keep looking for "config" until there are
+	arguments to parse.
+	Reported by: Adrien Dumont <adr.dum@hotmail.fr>
+
 	* test.c: Include "wget.h".
 
 2012-12-08  Michael Stapelberg <michael@stapelberg.de> (tiny change)
diff --git a/src/main.c b/src/main.c
index 4c2eb45..992f60a 100644
--- a/src/main.c
+++ b/src/main.c
@@ -1034,7 +1034,6 @@ main (int argc, char **argv)
                                 short_options, long_options, &longindex)) != -1)
     {
       int confval;
-      bool userrc_ret = true;
       struct cmdline_option *config_opt;
 
       /* There is no short option for "--config". */
@@ -1044,16 +1043,17 @@ main (int argc, char **argv)
           config_opt = &option_data[confval & ~BOOLEAN_NEG_MARKER];
           if (strcmp (config_opt->long_name, "config") == 0)
             {
+              bool userrc_ret = true;
               userrc_ret &= run_wgetrc (optarg);
               use_userconfig = true;
+              if (userrc_ret)
+                break;
+              else
+                {
+                  fprintf (stderr, _("Exiting due to error in %s\n"), optarg);
+                  exit (2);
+                }
             }
-          if (!userrc_ret)
-            {
-              fprintf (stderr, _("Exiting due to error in %s\n"), optarg);
-              exit (2);
-            }
-          else
-            break;
         }
     }
 
-- 
cgit v1.0-41-gc330

