diff --git a/src/ChangeLog b/src/ChangeLog
index aff9445..989106d 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,6 +1,12 @@
 2005-06-15  Hrvoje Niksic  <hniksic@xemacs.org>
 
-	* ftp-basic.c (ftp_pwd): Handle malformed PWD response.
+	* http.c (gethttp): Also report the system error when the POST
+	data file is missing.
+
+2005-06-15  Benno Schulenberg  <benno@nietvergeten.nl>
+
+	* ftp.c, http.c, connect.c, cookies.c, html-url.c, init.c, res.c:
+	Gettext-ize messages that were previously missed.
 
 2005-06-15  Hrvoje Niksic  <hniksic@xemacs.org>
 
diff --git a/src/connect.c b/src/connect.c
index d6626f9..24dd1fa 100644
--- a/src/connect.c
+++ b/src/connect.c
@@ -344,7 +344,7 @@ connect_to_ip (const ip_address *ip, int port, const char *print)
     if (sock >= 0)
       fd_close (sock);
     if (print)
-      logprintf (LOG_VERBOSE, "failed: %s.\n", strerror (errno));
+      logprintf (LOG_VERBOSE, _("failed: %s.\n"), strerror (errno));
     errno = save_errno;
     return -1;
   }
diff --git a/src/cookies.c b/src/cookies.c
index 929ab17..fad20e7 100644
--- a/src/cookies.c
+++ b/src/cookies.c
@@ -862,7 +862,7 @@ cookie_handle_set_cookie (struct cookie_jar *jar,
       if (!check_domain_match (cookie->domain, host))
 	{
 	  logprintf (LOG_NOTQUIET,
-		     "Cookie coming from %s attempted to set domain to %s\n",
+		     _("Cookie coming from %s attempted to set domain to %s\n"),
 		     escnonprint (host), escnonprint (cookie->domain));
 	  xfree (cookie->domain);
 	  goto copy_domain;
@@ -1317,7 +1317,7 @@ cookie_jar_load (struct cookie_jar *jar, const char *file)
   FILE *fp = fopen (file, "r");
   if (!fp)
     {
-      logprintf (LOG_NOTQUIET, "Cannot open cookies file `%s': %s\n",
+      logprintf (LOG_NOTQUIET, _("Cannot open cookies file `%s': %s\n"),
 		 file, strerror (errno));
       return;
     }
diff --git a/src/ftp.c b/src/ftp.c
index 9609f90..31ac37a 100644
--- a/src/ftp.c
+++ b/src/ftp.c
@@ -1111,7 +1111,7 @@ ftp_loop_internal (struct url *u, struct fileinfo *f, ccon *con)
   if (opt.noclobber && file_exists_p (con->target))
     {
       logprintf (LOG_VERBOSE,
-		 _("File `%s' already there, not retrieving.\n"), con->target);
+		 _("File `%s' already there; not retrieving.\n"), con->target);
       /* If the file is there, we suppose it's retrieved OK.  */
       return RETROK;
     }
diff --git a/src/html-url.c b/src/html-url.c
index fcf1e81..853226d 100644
--- a/src/html-url.c
+++ b/src/html-url.c
@@ -702,7 +702,7 @@ get_urls_file (const char *file)
       url = url_parse (url_text, &up_error_code);
       if (!url)
 	{
-	  logprintf (LOG_NOTQUIET, "%s: Invalid URL %s: %s\n",
+	  logprintf (LOG_NOTQUIET, _("%s: Invalid URL %s: %s\n"),
 		     file, url_text, url_error (up_error_code));
 	  xfree (url_text);
 	  continue;
diff --git a/src/http.c b/src/http.c
index 30e4dda..3170d77 100644
--- a/src/http.c
+++ b/src/http.c
@@ -1365,8 +1365,8 @@ gethttp (struct url *u, struct http_stat *hs, int *dt, struct url *proxy)
 	  post_data_size = file_size (opt.post_file_name);
 	  if (post_data_size == -1)
 	    {
-	      logprintf (LOG_NOTQUIET, "POST data file missing: %s\n",
-			 opt.post_file_name);
+	      logprintf (LOG_NOTQUIET, _("POST data file missing: %s (%s)\n"),
+			 opt.post_file_name, strerror (errno));
 	      post_data_size = 0;
 	    }
 	}
@@ -2040,7 +2040,7 @@ http_loop (struct url *u, char **newloc, char **local_file, const char *referer,
       /* If opt.noclobber is turned on and file already exists, do not
 	 retrieve the file */
       logprintf (LOG_VERBOSE, _("\
-File `%s' already there, will not retrieve.\n"), *hstat.local_file);
+File `%s' already there; not retrieving.\n\n"), *hstat.local_file);
       /* If the file is there, we suppose it's retrieved OK.  */
       *dt |= RETROKF;
 
@@ -2440,7 +2440,7 @@ The sizes do not match (local %s) -- retrieving.\n"),
 	  else if (!opt.kill_longer) /* meaning we got more than expected */
 	    {
 	      logprintf (LOG_VERBOSE,
-			 _("%s (%s) - `%s' saved [%s/%s])\n\n"),
+			 _("%s (%s) - `%s' saved [%s/%s]\n\n"),
 			 tms, tmrate, locf,
 			 number_to_static_string (hstat.len),
 			 number_to_static_string (hstat.contlen));
diff --git a/src/res.c b/src/res.c
index 0be8e7c..48d6fa2 100644
--- a/src/res.c
+++ b/src/res.c
@@ -391,7 +391,7 @@ res_parse_from_file (const char *filename)
   struct file_memory *fm = read_file (filename);
   if (!fm)
     {
-      logprintf (LOG_NOTQUIET, "Cannot open %s: %s",
+      logprintf (LOG_NOTQUIET, _("Cannot open %s: %s"),
 		 filename, strerror (errno));
       return NULL;
     }
-- 
cgit v1.0-41-gc330

