diff --git a/src/url.h b/src/url.h
index 69db155..0748e21 100644
--- a/src/url.h
+++ b/src/url.h
@@ -85,7 +85,7 @@ struct url
 char *url_escape (const char *);
 
 struct url *url_parse (const char *, int *, struct iri *iri);
-const char *url_error (const char *, int);
+char *url_error (const char *, int);
 char *url_full_path (const struct url *);
 void url_set_dir (struct url *, const char *);
 void url_set_file (struct url *, const char *);
-- 
cgit v1.0-41-gc330

