diff --git a/src/ChangeLog b/src/ChangeLog
index 8acab92..9aab3a4 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,8 @@
+2008-09-02  Gisle Vanem  <gvanem@broadpark.no>
+
+	* mswindows.h: Must ensure <stdio.h> is included before
+	we redefine ?vsnprintf().
+
 2008-08-08  Steven Schubiger  <stsc@members.fsf.org>
 
 	* main.c, utils.h: Removed some dead conditional DEBUG_MALLOC code.
diff --git a/src/mswindows.h b/src/mswindows.h
index 54821a7..7168727 100644
--- a/src/mswindows.h
+++ b/src/mswindows.h
@@ -78,6 +78,8 @@ as that of the covered work.  */
 # define strncasecmp strnicmp
 #endif
 
+#include <stdio.h>
+
 /* The same for snprintf() and vsnprintf().  */
 #define snprintf _snprintf
 #define vsnprintf _vsnprintf
-- 
cgit v1.0-41-gc330

