diff --git a/src/ChangeLog b/src/ChangeLog
index 5b19436..632cc70 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,5 +1,9 @@
 2001-11-27  Hrvoje Niksic  <hniksic@arsdigita.com>
 
+	* main.c (print_help): Remove stray HAVE_RANDOM code.
+
+2001-11-27  Hrvoje Niksic  <hniksic@arsdigita.com>
+
 	* ftp.c (getftp): Improve output after sending PASV.  Don't
 	attempt to "look up" the IP address we already know; call
 	connect_to_one directly.
diff --git a/src/main.c b/src/main.c
index eaf952e..f552ba6 100644
--- a/src/main.c
+++ b/src/main.c
@@ -166,10 +166,6 @@ Download:\n\
   -Q,  --quota=NUMBER           set retrieval quota to NUMBER.\n\
        --limit-rate=RATE        limit download rate to RATE.\n\
 \n"), stdout);
-#ifdef HAVE_RANDOM
-  fputs (_("\
-\n"), stdout);
-#endif
   fputs (_("\
 Directories:\n\
   -nd  --no-directories            don\'t create directories.\n\
-- 
cgit v1.0-41-gc330

