diff --git a/src/ChangeLog b/src/ChangeLog
index 4a26e4a..2f3de8a 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,8 @@
+2009-09-24  Micah Cowan  <micah@cowan.name>
+
+	* sysdep.h: Use proper prototypes for snprintf, vsnprintf (thanks
+	to Steven Schweda).
+
 2009-09-24  Steven Schweda  <sms@antinode.info>
 
 	* main.c (print_version): Remove VMS-conditional build info
diff --git a/src/sysdep.h b/src/sysdep.h
index f36b026..e0dcdee 100644
--- a/src/sysdep.h
+++ b/src/sysdep.h
@@ -216,10 +216,11 @@ void *memrchr (const void *, int, size_t);
 /* These are defined in snprintf.c.  It would be nice to have an
    snprintf.h, though.  */
 #ifndef HAVE_SNPRINTF
-int snprintf ();
+int snprintf (char *str, size_t count, const char *fmt, ...);
 #endif
 #ifndef HAVE_VSNPRINTF
-int vsnprintf ();
+#include <stdarg.h>
+int vsnprintf (char *str, size_t count, const char *fmt, va_list arg);
 #endif
 
 /* Some systems (Linux libc5, "NCR MP-RAS 3.0", and others) don't
-- 
cgit v1.0-41-gc330

