diff --git a/src/ChangeLog b/src/ChangeLog
index cba2113..07152a5 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,8 @@
+2012-11-24  Giuseppe Scrivano  <gscrivano@gnu.org>
+
+	* warc.c (warc_write_block_from_file): Use `number_to_string' to
+	convert the content-length to a string.
+
 2012-11-15  Giuseppe Scrivano  <gscrivano@gnu.org>
 
 	* retr.c (write_data): Fix comment.
diff --git a/src/warc.c b/src/warc.c
index de99bf7..99e7016 100644
--- a/src/warc.c
+++ b/src/warc.c
@@ -245,15 +245,10 @@ static bool
 warc_write_block_from_file (FILE *data_in)
 {
   /* Add the Content-Length header. */
-  char *content_length;
+  char content_length[22];
   fseeko (data_in, 0L, SEEK_END);
-  if (! asprintf (&content_length, "%ld", ftello (data_in)))
-    {
-      warc_write_ok = false;
-      return false;
-    }
+  number_to_string (content_length, ftello (data_in));
   warc_write_header ("Content-Length", content_length);
-  free (content_length);
 
   /* End of the WARC header section. */
   warc_write_string ("\r\n");
-- 
cgit v1.0-41-gc330

