@@ -1,5 +1,3 @@
-$Id$
-
 ==============================================================================
 
 SQUID Internet Object Cache  http://www.squid-cache.org
@@ -1,8 +1,6 @@
 
 Squid QUICKSTART
 
-$Id$
-
 This document is intended only for people who want to get Squid running
 quickly It is not a substitute for the real documentation.  Squid has
 many features, but only a few of them are useful at the beginning.  Use
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 /* Extended regular expression matching and search library,
  * version 0.12.
  * (Implements POSIX draft P10003.2/D11.2, except for
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_REGEXP_LIBRARY_H
 #define SQUID_REGEXP_LIBRARY_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_COMPAT_CPU_H
 #define SQUID_COMPAT_CPU_H
 
@@ -12,8 +12,6 @@
  *
  *    06-Oct-2007 : Various fixes to allow the build on MinGW
  *
- * Squid CVS $Id$
- *
  *  Original License and code follows.
  */
 #include "squid.h"
@@ -13,8 +13,6 @@
  *    16-Aug-2007 : Altered configure checks
  *                  Un-hacked slightly to use system gethostbyname()
  *
- * Squid CVS $Id$
- *
  *  Original License and code follows.
  */
 
@@ -14,8 +14,6 @@
  *                      - use xinet_ntop instead of inet_ntop
  *                      - use SQUIDHOSTNAMELEN instead of MAXHOSTNAMELEN
  *
- * Squid CVS $Id$
- *
  *  Original License and code follows.
  */
 #include "squid.h"
@@ -16,8 +16,6 @@
  *
  *    04-Nov-2010: drop SPRINTF casting macro
  *
- * Squid CVS $Id$
- *
  *  Original License and code follows.
  */
 
@@ -13,8 +13,6 @@
  *
  *    28-Oct-2007: drop some dead code. now tested working without.
  *
- * Squid CVS $Id$
- *
  *  Original License and code follows.
  */
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_COMPAT_SHM_H
 #define SQUID_COMPAT_SHM_H
 
@@ -12,8 +12,6 @@
  * 			- added squid_ prefix for uniqueness
  * 			  so we can use it where OS copy is broken.
  *
- * Squid VCS $Id$
- *
  *  Original License and code follows.
  */
 
@@ -23,8 +23,6 @@
  * 			- xtables_strtoui renamed to xstrtoui
  * 			- xtables_strtoul renamed to xstrtoul
  *
- * Squid VCS $Id$
- *
  *  Original License and code follows.
  */
 
@@ -1,8 +1,3 @@
-dnl
-dnl  $Id$
-dnl
-dnl
-dnl
 AC_INIT([Squid Web Proxy],[3.HEAD-BZR],[http://www.squid-cache.org/bugs/],[squid])
 AC_PREREQ(2.61)
 AC_CONFIG_HEADERS([include/autoconf.h])
@@ -1,7 +1,4 @@
 ## Process this file with automake to produce Makefile.in
-#
-# $Id$
-#
 
 errordir = $(datadir)/errors
 
@@ -1,5 +1,4 @@
 #!@PERL@
-# $Id$
 
 #if you define this, debugging output will be printed to STDERR.
 #$debug=1;
@@ -1,23 +1,23 @@
-/* $Id$
-* Copyright (C) 2002 Rodrigo Campos
-*
-* This program is free software; you can redistribute it and/or modify
-* it under the terms of the GNU General Public License as published by
-* the Free Software Foundation; either version 2 of the License, or
-* (at your option) any later version.
-*
-* This program is distributed in the hope that it will be useful,
-* but WITHOUT ANY WARRANTY; without even the implied warranty of
-* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-* GNU General Public License for more details.
-*
-* You should have received a copy of the GNU General Public License
-* along with this program; if not, write to the Free Software
-* Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
-*
-* Author: Rodrigo Campos (rodrigo@geekbunker.org)
-*
-*/
+/*
+ * Copyright (C) 2002 Rodrigo Campos
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ *
+ * Author: Rodrigo Campos (rodrigo@geekbunker.org)
+ *
+ */
 #include "squid.h"
 #include "helpers/defines.h"
 #include "rfc1738.h"
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef _SQUID_GETFULLHOSTNAME_H
 #define _SQUID_GETFULLHOSTNAME_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_HASH_H
 #define SQUID_HASH_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef _MEM_METER_H_
 #define _MEM_METER_H_
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_PARSE_H
 #define SQUID_PARSE_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_RADIX_H
 #define	SQUID_RADIX_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef _SQUID_INCLUDE_RFC1123_H
 #define _SQUID_INCLUDE_RFC1123_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_SNMP_API_UTIL_H
 #define SQUID_SNMP_API_UTIL_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_SNMP_CLIENT_H
 #define SQUID_SNMP_CLIENT_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_SNMP_DEBUG_H
 #define SQUID_SNMP_DEBUG_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_SNMP_IMPL_H
 #define SQUID_SNMP_IMPL_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_SNMP_UTIL_H
 #define SQUID_SNMP_UTIL_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_SPLAY_H
 #define SQUID_SPLAY_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef _INC_XUSLEEP_H
 #define _INC_XUSLEEP_H
 
@@ -1,7 +1,4 @@
 ## Process this file with automake to produce Makefile.in
-#
-# $Id$
-#
 
 AUTOMAKE_OPTIONS = dist-bzip2 subdir-objects 1.5
 DIST_SUBDIRS = include src test
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 /*
  * COMPILE WITH:
  *      gcc -Wall md5-test.c -I../include md5.o
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 /* Stub function for programs not implementing statMemoryAccounted */
 #include "squid.h"
 #include "util.h"
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #include "squid.h"
 #include "uudecode.h"
 
@@ -8,8 +8,6 @@
 # Check the Squid-1.1.x cache directory for stale objects - i.e. those
 # which exist on disk but aren't listed in cached's log file.
 
-# $Id$
-
 require "getopts.pl";
 &Getopts("c:drt:vh");
 # -c		: the full path to squid.conf
@@ -1,6 +1,5 @@
 #!/usr/bin/perl -w
 
-# $Id$
 # Convert hexadecimal cache file numbers (from swap log) into full pathnames.  
 # Duane Wessels 6/30/97
 
@@ -6,8 +6,6 @@
 # - i.e. those for which there is a "Content-length:" HTTP header,
 #   and this does not match the size of the cached object
 
-# $Id$
-
 require "getopts.pl";
 require "stat.pl";
 &Getopts("cd");
@@ -9,8 +9,6 @@
 # cf.data.pre and split it into separare files, one per option
 #
 # Henrik Nordstrom <henrik@henriknordstrom.net>
-#
-# $Id$
 
 #
 # The template file is reasonably simple to parse. There's a number of
@@ -1,7 +1,5 @@
 #!/usr/local/bin/perl
 
-# $Id$
-
 select(STDERR); $|=1;
 select(STDOUT); $|=1;
 
@@ -9,8 +9,6 @@
 # cf.data.pre and generate a set of HTML pages to use as documentation.
 #
 # Adrian Chadd <adrian@squid-cache.org>
-#
-# $Id$
 
 #
 # The template file is reasonably simple to parse. There's a number of
@@ -1,8 +1,3 @@
-
-/*
- * $Id$
- */
-
 #ifndef SQUID_COMMCALLS_H
 #define SQUID_COMMCALLS_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #include "squid.h"
 #include "base/TextException.h"
 #include "SquidTime.h"
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_FADING_COUNTER_H
 #define SQUID_FADING_COUNTER_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_HTTP_CONTROL_MSG_H
 #define SQUID_HTTP_CONTROL_MSG_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_MEMSTORE_H
 #define SQUID_MEMSTORE_H
 
@@ -1,8 +1,3 @@
-
-/*
- * $Id$
- */
-
 #include "squid.h"
 #include "acl/SslError.h"
 #include "acl/SslErrorData.h"
@@ -1,8 +1,3 @@
-
-/*
- * $Id$
- */
-
 #ifndef SQUID_ACLSSL_ERROR_H
 #define SQUID_ACLSSL_ERROR_H
 #include "acl/Strategy.h"
@@ -1,8 +1,3 @@
-
-/*
- * $Id$
- */
-
 #ifndef SQUID_ACLSSL_ERRORDATA_H
 #define SQUID_ACLSSL_ERRORDATA_H
 #include "acl/Acl.h"
@@ -1,8 +1,3 @@
-
-/*
- * $Id$
- */
-
 #ifndef SQUID__ADAPTATION__MESSAGE_H
 #define SQUID__ADAPTATION__MESSAGE_H
 
@@ -1,5 +1,4 @@
 /*
- * $Id$
  * DEBUG: section 93    eCAP Interface
  */
 #include "squid.h"
@@ -1,5 +1,4 @@
 /*
- * $Id$
  * DEBUG: section 93    eCAP Interface
  */
 
@@ -1,5 +1,4 @@
 /*
- * $Id$
  * DEBUG: section 93    eCAP Interface
  */
 
@@ -1,5 +1,4 @@
 /*
- * $Id$
  * DEBUG: section 93    eCAP Interface
  */
 
@@ -1,5 +1,4 @@
 /*
- * $Id$
  * DEBUG: section 93    eCAP Interface
  */
 
@@ -1,5 +1,4 @@
 /*
- * $Id$
  * DEBUG: section 93    eCAP Interface
  */
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #include "squid.h"
 #include "AsyncCall.h"
 #include "base/AsyncCall.h"
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_ASYNCCALL_H
 #define SQUID_ASYNCCALL_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_ASYNC_JOB_H
 #define SQUID_ASYNC_JOB_H
 
@@ -1,8 +1,3 @@
-
-/*
- * $Id$
- */
-
 #ifndef SQUID_ASYNCJOBCALLS_H
 #define SQUID_ASYNCJOBCALLS_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_CBC_POINTER_H
 #define SQUID_CBC_POINTER_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_BASE_TIDYPOINTER_H
 #define SQUID_BASE_TIDYPOINTER_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #include "squid.h"
 #include "dlink.h"
 
@@ -10,9 +10,6 @@ is very welcome though to resolve this issue.
 
 Adrian Chadd <adrian@creative.net.au>
 
-$Id$
-
-
 COSS is a Cyclic Object storage system originally designed by
 Eric Stern <estern@logisense.com>. The idea has been extended
 and worked into the current framework by myself.
@@ -1,7 +1,4 @@
 /*
- * $Id$
- * vim: set et :
- *
  * DEBUG: section 47    Store COSS Directory Routines
  * AUTHOR: Eric Stern
  *
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_IDENT_H
 #define SQUID_IDENT_H
 
@@ -1,9 +1,6 @@
 -- SQUID-MIB  { iso org(3) dod(6) internet(1) private(4) enterprises(1) 3495 }
 
 SQUID-MIB DEFINITIONS ::= BEGIN
---
--- $Id$
---
 
 IMPORTS
 	enterprises, Unsigned32, TimeTicks, Gauge32, Counter32,
@@ -1,6 +1,4 @@
 #******************************************************************************
-# $Id$
-#
 # File:		mk-strs.pl
 #
 # Author:	Max Okumoto <okumoto@ucsd.edu>
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #ifndef SQUID_PCONN_H
 #define SQUID_PCONN_H
 
@@ -1,7 +1,4 @@
 #  Makefile for storage modules in the Squid Object Cache server
-#
-#  $Id$
-#
 
 include $(top_srcdir)/src/Common.am
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #include "squid.h"
 #include "ssl/Config.h"
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_SSL_CONFIG_H
 #define SQUID_SSL_CONFIG_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #include "squid.h"
 #include "ssl/certificate_db.h"
 #if HAVE_ERRNO_H
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_SSL_CERTIFICATE_DB_H
 #define SQUID_SSL_CERTIFICATE_DB_H
 
@@ -1,6 +1,3 @@
-/*
- * $Id$
- */
 #include "squid.h"
 #include "Store.h"
 #include "StoreEntryStream.h"
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_SSL_CONTEXT_STORAGE_H
 #define SQUID_SSL_CONTEXT_STORAGE_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #include "squid.h"
 #include "ssl/gadgets.h"
 #include "ssl/crtd_message.h"
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_SSL_CRTD_MESSAGE_H
 #define SQUID_SSL_CRTD_MESSAGE_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #include "squid.h"
 #include "ssl/gadgets.h"
 #if HAVE_OPENSSL_X509V3_H
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #ifndef SQUID_SSL_HELPER_H
 #define SQUID_SSL_HELPER_H
 
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 #include "squid.h"
 #include "helpers/defines.h"
 #include "ssl/gadgets.h"
@@ -1,7 +1,5 @@
 #include "squid.h"
 
-/* $Id$
-
 /*
  * On some systems, FD_SETSIZE is set to something lower than the
  * actual number of files which can be opened.  IRIX is one case,
@@ -1,7 +1,3 @@
-/*
- * $Id$
- */
-
 // XXX: This file is made of large pieces of src/tools.cc
 // with only a few minor modifications. TODO: redesign or delete.
 
@@ -70,7 +70,6 @@ configuration file.
 modes of operation
 ==================
 
-$Id$
 Usage:  purge   [-a] [-c cf] [-d l] [-(f|F) fn | -(e|E) re] [-p h[:p]]
                 [-P #] [-s] [-v] [-C dir [-H]] [-n]
 
@@ -1,8 +1,5 @@
 #include "squid.h"
 
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    conffile.cc
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    conffile.hh
@@ -1,8 +1,5 @@
 #include "squid.h"
 
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    convert.cc
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    convert.hh
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    copyout.cc
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    copyout.hh
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    purge.cc
@@ -149,7 +146,6 @@ static bool verbose  = false;
 static bool envelope = false;
 static bool no_fork  = false;
 static const char* programname = 0;
-static const char* RCS_ID = "$Id$";
 
 // ----------------------------------------------------------------------
 
@@ -574,9 +570,9 @@ void
 helpMe( void )
 // purpuse: write help message and exit
 {
-    printf( "\n%s\nUsage:\t%s\t[-a] [-c cf] [-d l] [-(f|F) fn | -(e|E) re] "
+    printf( "\nUsage:\t%s\t[-a] [-c cf] [-d l] [-(f|F) fn | -(e|E) re] "
             "[-p h[:p]]\n\t\t[-P #] [-s] [-v] [-C dir [-H]] [-n]\n\n",
-            ::RCS_ID, ::programname );
+            ::programname );
     printf(
         " -a\tdisplay a little rotating thingy to indicate that I am alive (tty only).\n"
         " -c c\tsquid.conf location, default \"%s\".\n"
@@ -757,8 +753,8 @@ parseCommandline( int argc, char* argv[], REList*& head,
 
     // show results
     if ( showme ) {
-        printf( "#\n# Currently active values for %s:\n# %s\n",
-                ::programname, ::RCS_ID );
+        printf( "#\n# Currently active values for %s:\n",
+                ::programname);
         printf( "# Debug level       : " );
         if ( ::debugFlag ) printf( "%#6.4x", ::debugFlag );
         else printf( "production level" ); // printf omits 0x prefix for 0!
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:   Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 // File:     signal.cc
 // Date:     Sat Feb 28 1998
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:   Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 // File:     signal.hh
 // Date:     Sat Feb 28 1998
@@ -1,8 +1,4 @@
 #include "squid.h"
-
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    socket.hh
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    socket.hh
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    squid-tlv.cc
@@ -1,6 +1,3 @@
-//
-// $Id$
-//
 // Author:  Jens-S. V?ckler <voeckler@rvs.uni-hannover.de>
 //
 // File:    squid-tlv.hh