diff --git a/src/incremen.c b/src/incremen.c
index e4645fa..ec8a86d 100644
--- a/src/incremen.c
+++ b/src/incremen.c
@@ -454,7 +454,7 @@ read_directory_file (void)
 	incremental_version = 0;
 
       if (incremental_version > TAR_INCREMENTAL_VERSION)
-	ERROR((1, 0, _("Unsupported incremental format version: %s"),
+	ERROR((1, 0, _("Unsupported incremental format version: %d"),
 	       incremental_version));
       
       t = u = (errno = 0, strtoumax (buf, &ebuf, 10));
-- 
cgit v1.0-41-gc330

