diff --git a/src/create.c b/src/create.c
index 4cd5980..605251a 100644
--- a/src/create.c
+++ b/src/create.c
@@ -1317,7 +1317,7 @@ check_links (void)
     {
       if (lp->nlink)
 	{
-	  WARN ((0, 0, _("Missing links to '%s'.\n"), lp->name));
+	  WARN ((0, 0, _("Missing links to %s.\n"), quote (lp->name)));
 	}
     }
 }
diff --git a/src/extract.c b/src/extract.c
index f348b82..dcb341d 100644
--- a/src/extract.c
+++ b/src/extract.c
@@ -475,7 +475,7 @@ prepare_to_extract (char const *file_name)
     case KEEP_NEWER_FILES:
       if (file_newer_p (file_name, &current_stat_info))
 	{
-	  WARN ((0, 0, _("Current `%s' is newer"), file_name));
+	  WARN ((0, 0, _("Current %s is newer"), quote (file_name)));
 	  return 0;
 	}
       break;
diff --git a/src/xheader.c b/src/xheader.c
index 64b2d7a..0893bb3 100644
--- a/src/xheader.c
+++ b/src/xheader.c
@@ -159,7 +159,7 @@ xheader_set_keyword_equal (char *kw, char *eq)
   if (strcmp (kw, "delete") == 0)
     {
       if (xheader_protected_pattern_p (p))
-	USAGE_ERROR ((0, 0, _("Pattern %s cannot be used"), p));
+	USAGE_ERROR ((0, 0, _("Pattern %s cannot be used"), quote (p)));
       xheader_list_append (&keyword_pattern_list, p, NULL);
     }
   else if (strcmp (kw, "exthdr.name") == 0)
-- 
cgit v1.0-41-gc330

