diff --git a/src/xheader.c b/src/xheader.c
index 9b291a2..aed2eb0 100644
--- a/src/xheader.c
+++ b/src/xheader.c
@@ -573,8 +573,8 @@ decx (void *data, char const *keyword, char const *value, size_t size)
   if (t)
     t->decoder (st, keyword, value, size);
   else
-    ERROR((0, 0, _("Ignoring unknown extended header keyword `%s'"),
-	   keyword));
+    WARN((0, 0, _("Ignoring unknown extended header keyword `%s'"),
+	 keyword));
 }
 
 void
-- 
cgit v1.0-41-gc330

