diff --git a/src/tar.c b/src/tar.c
index 4a58a74..4b86e75 100644
--- a/src/tar.c
+++ b/src/tar.c
@@ -2480,7 +2480,7 @@ main (int argc, char **argv)
   name_term ();
 
   if (exit_status == TAREXIT_FAILURE)
-    error (0, 0, _("Error exit delayed from previous errors"));
+    error (0, 0, _("Exiting with failure status due to previous errors"));
 
   if (stdlis == stdout)
     close_stdout ();
-- 
cgit v1.0-41-gc330

