diff --git a/ChangeLog b/ChangeLog
index 4efe128..88da272 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,10 @@
+2003-08-29  Paul Eggert  <eggert@twinsun.com>
+
+	* NEWS, bootstrap: Drop en_GB locale; it was more trouble than it
+	was worth (e.g., different users in en_GB disagree about -ize
+	versus -ise).
+	* po/en_GB.po: Remove.
+
 2003-07-28  Sergey Poznyakoff  <gray@Mirddin.farlep.net>
 
 	* TODO: Updated
diff --git a/NEWS b/NEWS
index 5102aa9..f68759c 100644
--- a/NEWS
+++ b/NEWS
@@ -14,11 +14,10 @@ version 1.13.26
   option is now the default; use --no-overwrite-dir if you prefer
   the previous default behavior.
 
-* New message translations en_GB (British English), fi (Finnish), gl
-  (Galician), hr (Croatian), hu (Hungarian), ms (Malaysian), nb
-  (Norwegian), ro (Romanian), sk (Slovak), zh_CN (Chinese simplified),
-  zh_TW (Chinese traditional).  The code 'no' for Norwegian has been
-  withdrawn; use 'nb' instead.
+* New message translations fi (Finnish), gl (Galician), hr (Croatian),
+  hu (Hungarian), ms (Malaysian), nb (Norwegian), ro (Romanian), sk
+  (Slovak), zh_CN (Chinese simplified), zh_TW (Chinese traditional).
+  The code 'no' for Norwegian has been withdrawn; use 'nb' instead.
 
 * The semantics of -o option is changed. When extracting, it
 does the same as --no-same-owner GNU tar option. This is compatible
diff --git a/bootstrap b/bootstrap
index 34becb2..86ac961 100755
--- a/bootstrap
+++ b/bootstrap
@@ -139,7 +139,7 @@ done
 
 echo "$0: getting translations into po..."
 (cd po &&
- rm -f dummy `ls | sed -n '/\.gmo$/p; /^en_GB\.po$/d; /\.po/p'` &&
+ rm -f dummy `ls | sed -n '/\.gmo$/p; /\.po/p'` &&
  wget -nv -nd -r -l 1 -A .po -C off \
    http://www.iro.umontreal.ca/contrib/po/maint/tar/ &&
  ls *.po | sed 's/\.po$//' >LINGUAS
diff --git a/po/en_GB.po b/po/en_GB.po
deleted file mode 100644
index 32aa968..0000000
--- a/po/en_GB.po
+++ /dev/null
@@ -1,41 +0,0 @@
-# British English messages for GNU tar
-# Copyright (C) 2003 Free Software Foundation, Inc.
-# Paul Eggert <eggert@twinsun.com>, 2003.
-#
-msgid ""
-msgstr ""
-"Project-Id-Version: GNU tar 1.13.25\n"
-"Report-Msgid-Bugs-To: bug-tar@gnu.org\n"
-"POT-Creation-Date: 2003-07-04 00:11-0700\n"
-"PO-Revision-Date: 2003-07-04 00:11-0700\n"
-"Last-Translator: Paul Eggert <eggert@twinsun.com>\n"
-"Language-Team: English <en@translate.freefriends.org>\n"
-"MIME-Version: 1.0\n"
-"Content-Type: text/plain; charset=US-ASCII\n"
-"Content-Transfer-Encoding: 8bit\n"
-
-#: lib/getopt.c:838 lib/getopt.c:841
-#, c-format
-msgid "%s: unrecognized option `--%s'\n"
-msgstr "%s: unrecognised option `--%s'\n"
-
-#: lib/getopt.c:849 lib/getopt.c:852
-#, c-format
-msgid "%s: unrecognized option `%c%s'\n"
-msgstr "%s: unrecognised option `%c%s'\n"
-
-# POSIX requires the word "illegal" in the POSIX locale, but the error
-# is not really against the law.  This is not the POSIX locale, so fix
-# the wording.
-#: lib/getopt.c:899 lib/getopt.c:902
-#, c-format
-msgid "%s: illegal option -- %c\n"
-msgstr "%s: unrecognised option `-%c'\n"
-
-# POSIX requires the word "illegal" in the POSIX locale, but the error
-# is not really against the law.  This is not the POSIX locale, so fix
-# the wording.
-#: lib/getopt.c:908 lib/getopt.c:911
-#, fuzzy, c-format
-msgid "%s: invalid option -- %c\n"
-msgstr "%s: unrecognised option `-%c'\n"
-- 
cgit v1.0-41-gc330

