diff --git a/src/compare.c b/src/compare.c
index ea41cb3..7c514b9 100644
--- a/src/compare.c
+++ b/src/compare.c
@@ -362,15 +362,6 @@ static void
 diff_dumpdir (struct tar_stat_info *dir)
 {
   const char *dumpdir_buffer;
-  struct stat stat_data;
-
-  if (deref_stat (dir->file_name, &stat_data) != 0)
-    {
-      if (errno == ENOENT)
-	stat_warn (dir->file_name);
-      else
-	stat_error (dir->file_name);
-    }
 
   if (dir->fd == 0)
     {
-- 
cgit v1.0-41-gc330

