diff --git a/configure.ac b/configure.ac
index cdeb3bf..f6a48f9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -29,6 +29,7 @@ gl_USE_SYSTEM_EXTENSIONS
 AC_PROG_CC
 AC_EXEEXT
 AC_PROG_RANLIB
+AC_PROG_YACC
 AC_SYS_LARGEFILE
 AC_ISC_POSIX
 AC_C_INLINE
@@ -99,54 +100,8 @@ jm_AC_TYPE_UINTMAX_T
 
 
 # gnulib modules
+tar_GNULIB
 
-AM_FUNC_GETLINE
-
-gl_BACKUPFILE
-gl_DIRNAME
-gl_ERROR
-gl_EXCLUDE
-gl_FILEBLOCKS
-gl_FUNC_ALLOCA
-gl_FUNC_FNMATCH_GNU
-gl_FUNC_MEMSET
-gl_FUNC_MKTIME
-gl_FUNC_RMDIR
-gl_FUNC_STRTOL
-gl_FUNC_STRTOUL
-gl_FUNC_STRTOULL
-gl_FUNC_STRTOIMAX
-gl_FUNC_STRTOUMAX
-gl_GETDATE
-gl_GETOPT
-gl_GETTIME
-gl_HASH
-gl_HUMAN
-gl_MODECHANGE
-gl_OBSTACK
-gl_PATHMAX
-gl_QUOTE
-gl_QUOTEARG
-gl_SAFE_READ
-gl_SAFE_WRITE
-gl_SAVE_CWD
-gl_SAVEDIR
-gl_STRCASE
-gl_TIME_R
-gl_TIMESPEC
-gl_XALLOC
-gl_XGETCWD
-gl_XSTRTOL
-gl_FUNC_STPCPY
-
-jm_FUNC_FTRUNCATE
-jm_FUNC_GLIBC_UNLOCKED_IO
-jm_FUNC_CHOWN
-jm_FUNC_LCHOWN
-jm_FUNC_MALLOC
-jm_FUNC_REALLOC
-jm_FUNC_UTIME
-jm_XSTRTOUMAX
 
 AC_CHECK_MEMBERS([struct stat.st_spare1, struct stat.st_atim.tv_nsec, struct stat.st_atimespec.tv_nsec, struct stat.st_atimensec], , ,
                  [
-- 
cgit v1.0-41-gc330

