diff --git a/src/tar.c b/src/tar.c
index eacdaaf..6f21cb5 100644
--- a/src/tar.c
+++ b/src/tar.c
@@ -1759,8 +1759,11 @@ decode_options (int argc, char **argv)
     {
       if (multi_volume_option)
 	USAGE_ERROR ((0, 0, _("Cannot use multi-volume compressed archives")));
-      if (subcommand_option == UPDATE_SUBCOMMAND)
+      if (subcommand_option == UPDATE_SUBCOMMAND
+	  || subcommand_option == APPEND_SUBCOMMAND)
 	USAGE_ERROR ((0, 0, _("Cannot update compressed archives")));
+      if (subcommand_option == CAT_SUBCOMMAND)
+	USAGE_ERROR ((0, 0, _("Cannot concatenate compressed archives")));
     }
 
   /* It is no harm to use --pax-option on non-pax archives in archive
-- 
cgit v1.0-41-gc330

