diff --git a/src/buffer.c b/src/buffer.c
index 949505c..1d164b4 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -1210,12 +1210,7 @@ flush_read (void)
 	  archive_read_error ();
 
       if (status == 0)
-	{
-	  if (left % BLOCKSIZE != 0)
-	    ERROR ((0, 0, _("%d garbage bytes ignored at end of archive"),
-		    (int) ((record_size - left) % BLOCKSIZE)));
-	  break;
-	}
+	break;
 
       if (! read_full_records_option)
 	FATAL_ERROR ((0, 0, _("Unaligned block (%lu bytes) in archive"),
-- 
cgit v1.0-41-gc330

