diff --git a/src/incremen.c b/src/incremen.c
index 4bed23c..b33140d 100644
--- a/src/incremen.c
+++ b/src/incremen.c
@@ -188,10 +188,7 @@ get_directory_contents (char *path, dev_t device)
 
     if (! dirp)
       {
-	if (ignore_failed_read_option)
-	  savedir_warn (path);
-	else
-	  savedir_error (path);
+        savedir_error (path);
       }
     errno = 0;
 
@@ -229,10 +226,7 @@ get_directory_contents (char *path, dev_t device)
 
 	      if (deref_stat (dereference_option, name_buffer, &stat_data))
 		{
-		  if (ignore_failed_read_option)
-		    stat_warn (name_buffer);
-		  else
-		    stat_error (name_buffer);
+		  stat_diag (name_buffer);
 		  continue;
 		}
 
diff --git a/src/names.c b/src/names.c
index a19afd7..5cb1aa5 100644
--- a/src/names.c
+++ b/src/names.c
@@ -833,10 +833,7 @@ collect_and_sort_names (void)
 
       if (deref_stat (dereference_option, name->name, &statbuf) != 0)
 	{
-	  if (ignore_failed_read_option)
-	    stat_warn (name->name);
-	  else
-	    stat_error (name->name);
+	  stat_diag (name->name);
 	  continue;
 	}
       if (S_ISDIR (statbuf.st_mode))
-- 
cgit v1.0-41-gc330

