diff --git a/NEWS b/NEWS
index 6226d07..2fa8f51 100644
--- a/NEWS
+++ b/NEWS
@@ -1,9 +1,16 @@
-GNU tar NEWS - User visible changes. 2011-11-26
+GNU tar NEWS - User visible changes. 2012-01-18
 Please send GNU tar bug reports to <bug-tar@gnu.org>
 
 
 version 1.26.90 (Git)
 
+* Quoting
+
+In the default C locale, diagnostics and output of 'tar' have been
+adjusted to quote 'like this' (with apostrophes) instead of `like this'
+(with an accent grave character and an apostrophe).  This tracks
+recent changes to the GNU coding standards.
+
 * New features
 
 ** --owner and --group names and numbers
@@ -74,14 +81,14 @@ are met, tar retries extraction using an alternative decompressor:
  2. The compression program was not explicitly requested in the command
  line by the use of such options as -z, -j, etc.
 
-For example, if `compress' is not available, tar will try `gzip'.
+For example, if 'compress' is not available, tar will try 'gzip'.
 
 
 version 1.24 - Sergey Poznyakoff, 2010-10-24
 
 * The --full-time option.
 
-New command line option `--full-time' instructs tar to output file
+New command line option '--full-time' instructs tar to output file
 time stamps to the full resolution.
 
 * Bugfixes.
@@ -134,15 +141,15 @@ invoked as in the example below:
 
 ** --remove-files
 
-`Tar --remove-files' failed to remove a directory which contained
+'tar --remove-files' failed to remove a directory which contained
 symlinks to another files within that directory.
 
 ** --test-label behavior
 
-In case of a mismatch, `tar --test-label LABEL' exits with code 1,
+In case of a mismatch, 'tar --test-label LABEL' exits with code 1,
 not 2 as it did in previous versions.
 
-The `--verbose' option used with `--test-label' provides additional
+The '--verbose' option used with '--test-label' provides additional
 diagnostics.
 
 Several volume labels may be specified in a command line, e.g.:
@@ -154,7 +161,7 @@ matches the actual volume label.
 
 ** --label used with --update
 
-The `--label' option can be used with `--update' to prevent accidental
+The '--label' option can be used with '--update' to prevent accidental
 update of an archive:
 
   tar -rf archive --label 'My volume' .
@@ -182,28 +189,28 @@ to regular files and pipes).
 When a read-only operation (e.g. --list or --extract) is requested
 on a regular file, tar attemtps to speed up accesses by using lseek.
 
-* New command line option `--warning'
+* New command line option '--warning'
 
-The `--warning' command line option allows to suppress or enable
-particular warning messages during `tar' run.  It takes a single
-argument (a `keyword'), identifying the class of warning messages
-to affect.  If the argument is prefixed with `no-', such warning
+The '--warning' command line option allows to suppress or enable
+particular warning messages during 'tar' run.  It takes a single
+argument (a 'keyword'), identifying the class of warning messages
+to affect.  If the argument is prefixed with 'no-', such warning
 messages are suppressed.  For example,
 
   tar --warning=no-alone-zero-block -x -f archive
 
-suppresses the output of `A lone zero block' diagnostics, which is
-normally issued if `archive' ends with a single block of zeros.
+suppresses the output of "A lone zero block" diagnostics, which is
+normally issued if 'archive' ends with a single block of zeros.
 
 See Tar Manual, section 3.9 "Controlling Warning Messages", for a
 detailed discussion.
 
-* New command line option `--level'
+* New command line option '--level'
 
-The `--level=N' option sets the incremental dump level N.  It
+The '--level=N' option sets the incremental dump level N.  It
 is valid when used in conjunction with the -c and --listed-incremental
 options.  So far the only meaningful value for N is 0.  The
-`--level=0' option forces creating the level 0 dump, by truncating
+'--level=0' option forces creating the level 0 dump, by truncating
 the snapshot file if it exists.
 
 * Files removed during incremental dumps
@@ -232,7 +239,7 @@ can be overridden by the
   --pax-opion='exthdr.mtime=STRING'
 
 command line option.  The STRING is either number of seconds since
-the Epoch or a `Time reference' (see below).
+the Epoch or a "Time reference" (see below).
 
 Modification times in ustar header blocks of global extended
 headers are set to the time when tar was invoked.
@@ -242,7 +249,7 @@ This can be overridden by the
   --pax-opion='globexthdr.mtime=STRING'
 
 command line option.  The STRING is either number of seconds since
-the Epoch or a `Time reference' (see below).
+the Epoch or a "Time reference" (see below).
 
 * Time references in --pax-option argument.
 
@@ -250,7 +257,7 @@ Any value from the --pax-option argument that is enclosed in a pair
 of curly braces represents a time reference.  The string between the
 braces is understood either as a textual time representation, as described in
 chapter 7, "Date input formats", of the Tar manual, or as a name of
-an existing file, starting with `/' or `.'.  In the latter
+an existing file, starting with '/' or '.'.  In the latter
 case, it is replaced with the modification time of that file.
 
 * Environment of --to-command script.
@@ -285,7 +292,7 @@ Tar uses xz for compression if one of the following conditions is met:
 
  1. The option --xz or -J (see below) is used.
  2. The xz binary is set as compressor using --use-compress-program option.
- 3. The file name of the archive being created ends in `.xz' and
+ 3. The file name of the archive being created ends in '.xz' and
  auto-compress option (-a) is used.
 
 Xz is used for decompression if one of the following conditions is met:
@@ -345,17 +352,17 @@ control type of archive members affected by them.  The flags are:
    Apply transformation to hard link targets.
 
 Corresponding upper-case letters negate the meaning, so that
-`H' means ``do not apply transformation to hard link targets.''
+'H' means "do not apply transformation to hard link targets".
 
-The scope flags are listed in the third part of an `s' expression,
+The scope flags are listed in the third part of an 's' expression,
 e.g.:
 
    tar --transform 's|^|/usr/local/|S'
 
-Default is `rsh', which means that transformations are applied to
+Default is 'rsh', which means that transformations are applied to
 both regular archive members and to the targets of symbolic and hard
 links.  If several transform expressions are used, the default flags
-can be changed using `flags=' statement before the expressions, e.g.:
+can be changed using 'flags=' statement before the expressions, e.g.:
 
    tar --transform 'flags=S;s|^|/usr/local/|S'
 
@@ -391,29 +398,29 @@ they refer to, instead of creating usual hard link members (type '1').
 This action allows to specify an action to be executed upon hitting a
 checkpoint.  Recognized actions are: dot, echo (the default),
 echo=string, ttyout=string, exec=cmdline, and sleep=value.  Any number
-of `--checkpoint-action' options can be specified, the actions will be
+of '--checkpoint-action' options can be specified, the actions will be
 executed in order of their appearance in the command line.  See
 chapter 3.8 "Checkpoints" for a complete description.
 
 * New options --no-check-device, --check-device.
 
-The `--no-check-device' option disables comparing device numbers during
+The '--no-check-device' option disables comparing device numbers during
 preparatory stage of an incremental dump.  This allows to avoid
 creating full dumps if the device numbers change (e.g. when using an
 LVM snapshot).
 
-The `--check-device' option enables comparing device numbers.  This is
+The '--check-device' option enables comparing device numbers.  This is
 the default.  This option is provided to undo the effect of the previous
-`--no-check-device' option, e.g. if it was set in TAR_OPTIONS
+'--no-check-device' option, e.g. if it was set in TAR_OPTIONS
 environment variable.
 
 * The --transform option.
 
-Any number of `--transform' options can be given in the command line.
+Any number of '--transform' options can be given in the command line.
 The specified transformations will be applied in turn.
 
-The argument to `--transform' option can be a list of replace
-expressions, separated by a semicolon (as in `sed').
+The argument to '--transform' option can be a list of replace
+expressions, separated by a semicolon (as in 'sed').
 
 Filename transformations are applied to symbolic link targets
 during both creation and extraction.  Tar 1.19 used them only
@@ -499,7 +506,7 @@ option affects hard and soft link targets and the --strip-components
 option affects hard link targets as well.
 
 * End-of-volume script can send the new volume name to tar by writing
-  it to the file descriptor stored in the environment variable `TAR_FD'.
+  it to the file descriptor stored in the environment variable TAR_FD.
 
 
 version 1.16.1 - Sergey Poznyakoff, 2006-12-09
@@ -569,7 +576,7 @@ following command line options:
      --wildcards              use wildcards
      --anchored               patterns match file name start
      --ignore-case            ignore case
-     --wildcards-match-slash  wildcards match `/'
+     --wildcards-match-slash  wildcards match '/'
 
 Each of these options has a '--no-' counterpart that disables its
 effect (e.g. --no-wildcards).
@@ -609,7 +616,7 @@ or archive.   It generalizes --show-stored-names option, introduced in
 1.15.90.  In particular, when creating an archive in verbose mode, it lists
 member names as stored in the archive, i.e., with any eventual prefixes
 removed and file name transformations applied.  The option is useful,
-for example, while comparing `tar cv' and `tar tv' outputs.
+for example, while comparing 'tar cv' and 'tar tv' outputs.
 
 ** New incremental snapshot file format keeps information about file names
 as well as that about directories.
@@ -634,7 +641,7 @@ version 1.15.90 - Sergey Poznyakoff, 2006-02-19
 * New features
 
 ** Any number of -T (--files-from) options may be used in the command line.
-The file specified with -T may include any valid `tar' options,
+The file specified with -T may include any valid 'tar' options,
 including another -T option.
 Compatibility note: older versions of tar would only recognize -C
 as an option name within the file list file.  Now any file whose name
@@ -645,7 +652,7 @@ dash, use the --add-file option.
 automatically.  It is no longer necessary to give the --null option.
 
 ** New option --no-unquote disables the unquoting of input file names.
-This is useful for processing output from `find dir -print0'.
+This is useful for processing output from 'find dir -print0'.
 An orthogonal option --unquote is provided as well.
 
 ** New option --test-label tests the archive volume label.
@@ -659,7 +666,7 @@ tar prints the label name if present and exits with code 0.
 ** New option --show-stored-names.  When creating an archive in verbose mode,
 it lists member names as stored in the archive, i.e., with any eventual
 prefixes removed.  The option is useful, for example, while comparing
-`tar cv' and `tar tv' outputs.
+'tar cv' and 'tar tv' outputs.
 
 ** New option --to-command pipes the contents of archive members to the
 specified command.
@@ -732,7 +739,7 @@ version 1.15 - Sergey Poznyakoff, 2004-12-20
 
 * Compressed archives are recognised automatically, it is no longer
 necessary to specify -Z, -z, or -j options to read them.  Thus, you can
-now run `tar tf archive.tar.gz'.
+now run 'tar tf archive.tar.gz'.
 
 * When restoring incremental dumps, --one-file-system option
 prevents directory hierarchies residing on different devices
@@ -752,7 +759,7 @@ seeks.
 * Restore script starts restoring only if it is given --all (-a) option,
 or some patterns.  This is to prevent accidental restores.
 
-* `tar --verify' prints a warning if during archive creation some of
+* 'tar --verify' prints a warning if during archive creation some of
 the file names had their prefixes stripped off.
 
 * New option --exclude-caches instructs tar to exclude cache directories
@@ -762,11 +769,11 @@ containing a standardized tag file, as specified at:
 	http://www.brynosaurus.com/cachedir/spec.html
 
 * New configure option --with-rmt allows to specify full path name to
-the `rmt' utility.  This supersedes DEFAULT_RMT_COMMAND variable
+the 'rmt' utility.  This supersedes DEFAULT_RMT_COMMAND variable
 introduced in version 1.14
 
 * New configure variable DEFAULT_RMT_DIR allows to specify the directory
-where to install `rmt' utility.  This is necessary since modifying
+where to install 'rmt' utility.  This is necessary since modifying
 --libexecdir as was suggested for version 1.14 produced a side effect: it
 also modified installation prefix for backup scripts (if
 --enable-backup-scripts was given).
@@ -820,12 +827,12 @@ version 1.14 - Sergey Poznyakoff, 2004-05-11
   the subcommands --delete, --diff, --extract or --list when a list of
   files is given either on the command line or via -T option.  This
   option instructs tar to process only the Nth occurrence of each named
-  file.  N defaults to 1, so `tar -x -f archive --occurrence filename'
-  extracts the first occurrence of `filename' from `archive'
+  file.  N defaults to 1, so 'tar -x -f archive --occurrence filename'
+  extracts the first occurrence of 'filename' from 'archive'
   and terminates without scanning to the end of the archive.
 
 * New option --pax-option allows to control the handling of POSIX
-  keywords in `pax' extended headers.  It is equivalent to `pax'
+  keywords in 'pax' extended headers.  It is equivalent to 'pax'
   -o option.
 
 * --incremental and --listed-incremental options work correctly on
@@ -931,7 +938,7 @@ version 1.13.19 - Paul Eggert, 2001-01-13
 
 version 1.13.18 - Paul Eggert, 2000-10-29
 
-* Some security problems have been fixed.  `tar -x' now modifies only
+* Some security problems have been fixed.  'tar -x' now modifies only
   files under the working directory, unless you also specify an unsafe
   option like --absolute-names or --overwrite.
 
@@ -957,12 +964,12 @@ version 1.13.18 - Paul Eggert, 2000-10-29
 * This version of tar works best with GNU gzip test version 1.3 or later.
   Please see <ftp://alpha.gnu.org/gnu/gzip/>.
 
-* `tar --delete -f -' now works again.
+* 'tar --delete -f -' now works again.
 
 
 version 1.13.17 - Paul Eggert, 2000-01-07.
 
-* `tar --delete -f -' is no longer allowed; it was too buggy.
+* 'tar --delete -f -' is no longer allowed; it was too buggy.
 * Diagnostic messages have been made more regular and consistent.
 
 
@@ -979,12 +986,12 @@ version 1.13.16 - Paul Eggert, 1999-12-13.
   The new --overwrite option enables the old default behavior.
 
   For regular files, tar implements this change by using the O_EXCL
-  option of `open' to ensure that it creates the file; if this fails, it
+  option of 'open' to ensure that it creates the file; if this fails, it
   removes the file and tries again.  This is similar to the behavior of
   the --unlink-first option, but it is faster in the common case of
   extracting a new directory.
 
-* By default, tar now ignores file names containing a component of `..'
+* By default, tar now ignores file names containing a component of '..'
   when extracting, and warns about such file names when creating an archive.
   To enable the old behavior, use the -P or --absolute-names option.
 
@@ -1026,20 +1033,20 @@ version 1.13.14 - Paul Eggert, 1999-11-07.
 version 1.13.13 - Paul Eggert, 1999-10-11.
 
 * Invalid headers in tar files now elicit errors, not just warnings.
-* `tar --version' output conforms to the latest GNU coding standards.
-* If you specify an invalid date, `tar' now substitutes (time_t) -1.
-* `configure --with-dmalloc' is no longer available.
+* 'tar --version' output conforms to the latest GNU coding standards.
+* If you specify an invalid date, 'tar' now substitutes (time_t) -1.
+* 'configure --with-dmalloc' is no longer available.
 
 
 version 1.13.12 - Paul Eggert, 1999-09-24.
 
-* `tar' now supports hard links to symbolic links.
+* 'tar' now supports hard links to symbolic links.
 
 * New options --no-same-owner, --no-same-permissions.
 
 * --total now also outputs a human-readable size, and a throughput value.
 
-* `tar' now uses two's-complement base-256 when outputting header
+* 'tar' now uses two's-complement base-256 when outputting header
   values that are out of the range of the standard unsigned base-8
   format.  This affects archive members with negative or huge time
   stamps or uids, and archive members 8 GB or larger.  The new tar
@@ -1048,12 +1055,12 @@ version 1.13.12 - Paul Eggert, 1999-09-24.
   behavior, which uses unportable representations for negative values,
   and which rejects large files.
 
-* On 32-bit hosts, `tar' now assumes that an incoming time stamp T in
+* On 32-bit hosts, 'tar' now assumes that an incoming time stamp T in
   the range 2**31 <= T < 2**32 represents the negative time (T -
   2**32).  This behavior is nonstandard and is not portable to 64-bit
-  time_t hosts, so `tar' issues a warning.
+  time_t hosts, so 'tar' issues a warning.
 
-* `tar' no longer gives up extracting immediately upon discovering
+* 'tar' no longer gives up extracting immediately upon discovering
   that an archive contains garbage at the end.  It attempts to extract
   as many files as possible from the good data before the garbage.
 
@@ -1070,40 +1077,40 @@ version 1.13.11 - Paul Eggert, 1999-08-23.
 * -T /dev/null now matches nothing; previously, it matched anything
   if no explicit operands were given.
 
-* The `--' option now works the same as with other GNU utilities;
+* The '--' option now works the same as with other GNU utilities;
   it causes later operands to be interpreted as file names, not options,
-  even if they begin with `-'.
+  even if they begin with '-'.
 
 * For the --newer and --after-date options, the table of time zone
-  abbreviations like `EST' has been updated to match current practice.
+  abbreviations like 'EST' has been updated to match current practice.
   Also, local time abbreviations are now recognized, even if they are
   not in tar's hardwired table.  Remember, though, that you should use
-  numeric UTC offsets like `-0500' instead of abbreviations like
-  `EST', as abbreviations are not standardized and are ambiguous.
+  numeric UTC offsets like '-0500' instead of abbreviations like
+  'EST', as abbreviations are not standardized and are ambiguous.
 
 
 version 1.13.10 - Paul Eggert, 1999-08-20.
 
-* `tar' now uses signed base-64 when outputting header values that are
+* 'tar' now uses signed base-64 when outputting header values that are
   out of the range of the standard unsigned base-8 format.  [This
   change was superseded in 1.13.12, described above.]
 
 
 version 1.13.9 - Paul Eggert, 1999-08-18.
 
-* `tar' now writes two zero blocks at end-of-archive instead of just one.
-  POSIX.1 requires this, and some other `tar' implementations check for it.
+* 'tar' now writes two zero blocks at end-of-archive instead of just one.
+  POSIX.1 requires this, and some other 'tar' implementations check for it.
 
-* `tar' no longer silently accepts a block containing nonzero checksum bytes
+* 'tar' no longer silently accepts a block containing nonzero checksum bytes
   as a zero block.
 
-* `tar' now reads buggy tar files that have a null byte at the start of a
+* 'tar' now reads buggy tar files that have a null byte at the start of a
   numeric header field.
 
 
 version 1.13.8 - Paul Eggert, 1999-08-16.
 
-* For compatibility with traditional `tar', intermediate directories
+* For compatibility with traditional 'tar', intermediate directories
   created automatically by root are no longer given the uid and gid of
   the original file or directory.
 
@@ -1112,7 +1119,7 @@ version 1.13.7 - Paul Eggert, 1999-08-14.
 
 * --listed-incremental and --newer are now incompatible options.
 
-* When creating an archive, leading `./' is no longer stripped,
+* When creating an archive, leading './' is no longer stripped,
   to match traditional tar's behavior (and simplify the documentation).
 
 * --diff without --absolute-names no longer falls back on absolute names.
@@ -1203,7 +1210,7 @@ Creation
 * Implement --numeric-owner for ignoring symbolic names at create time.
 * New --owner, --group --mode options, still preliminary.
 * Recognize creating an archive on /dev/null, so Amanda works faster.
-* Object to the creation of an empty archive (like in `tar cf FILE').
+* Object to the creation of an empty archive (like in 'tar cf FILE').
 * Barely start implementing --posix and POSIXLY_CORRECT.
 
 Extraction
@@ -1282,7 +1289,7 @@ Version 1.10.13 - 1992-01.
 
 * Now uses GNU standard configure, generated by Autoconf.
 
-* Long options now use `--'; use of `+' is deprecated and support
+* Long options now use '--'; use of '+' is deprecated and support
 for it will eventually be removed.
 
 * New option --null causes filenames read by -T to be
@@ -1311,8 +1318,8 @@ the various tape drives will get used in sequence and then wrap
 around to the beginning.
 
 * Remote archive names no longer have to be in /dev: any file with a
-`:' is interpreted as remote.  If new option --force-local is given,
-then even archive files with a `:' are considered local.
+':' is interpreted as remote.  If new option --force-local is given,
+then even archive files with a ':' are considered local.
 
 * New option --atime-preserve restores (if possible) atimes to
 their original values after dumping the file.
diff --git a/README b/README
index bf65e82..957e1dc 100644
--- a/README
+++ b/README
@@ -4,31 +4,31 @@ See the end of file for copying conditions.
 * Introduction
 
 Please glance through *all* sections of this
-`README' file before starting configuration.  Also make sure you read files
-`ABOUT-NLS' and `INSTALL' if you are not familiar with them already.
+'README' file before starting configuration.  Also make sure you read files
+'ABOUT-NLS' and 'INSTALL' if you are not familiar with them already.
 
-If you got the `tar' distribution in `shar' format, time stamps ought to be
-properly restored; do not ignore such complaints at `unshar' time.
+If you got the 'tar' distribution in 'shar' format, time stamps ought to be
+properly restored; do not ignore such complaints at 'unshar' time.
 
-GNU `tar' saves many files together into a single tape or disk
+GNU 'tar' saves many files together into a single tape or disk
 archive, and can restore individual files from the archive.  It includes
 multivolume support, the ability to archive sparse files, automatic archive
 compression/decompression, remote archives and special features that allow
-`tar' to be used for incremental and full backups.  This distribution
-also includes `rmt', the remote tape server.  The `mt' tape drive control
-program is in the GNU `cpio' distribution.
+'tar' to be used for incremental and full backups.  This distribution
+also includes 'rmt', the remote tape server.  The 'mt' tape drive control
+program is in the GNU 'cpio' distribution.
 
-GNU `tar' is derived from John Gilmore's public domain `tar'.
+GNU 'tar' is derived from John Gilmore's public domain 'tar'.
 
-See file `ABOUT-NLS' for how to customize this program to your language.
-See file `COPYING' for copying conditions.
-See file `INSTALL' for compilation and installation instructions.
-See file `PORTS' for various ports of GNU tar to non-Unix systems.
-See file `NEWS' for a list of major changes in the current release.
-See file `THANKS' for a list of contributors.
+See file 'ABOUT-NLS' for how to customize this program to your language.
+See file 'COPYING' for copying conditions.
+See file 'INSTALL' for compilation and installation instructions.
+See file 'PORTS' for various ports of GNU tar to non-Unix systems.
+See file 'NEWS' for a list of major changes in the current release.
+See file 'THANKS' for a list of contributors.
 
-Besides those configure options documented in files `INSTALL' and
-`ABOUT-NLS', an extra option may be accepted after `./configure':
+Besides those configure options documented in files 'INSTALL' and
+'ABOUT-NLS', an extra option may be accepted after './configure':
 
 * Install
 
@@ -40,11 +40,11 @@ values are GNU, V7, OLDGNU, USTAR and POSIX.
 
 ** Selecting the default archive device
 
-The default archive device is now `stdin' on read and `stdout' on write.
-The installer can still override this by presetting `DEFAULT_ARCHIVE'
-in the environment before configuring (the behavior of `-[0-7]' or
-`-[0-7]lmh' options in `tar' are then derived automatically).  Similarly,
-`DEFAULT_BLOCKING' can be preset to something else than 20.
+The default archive device is now 'stdin' on read and 'stdout' on write.
+The installer can still override this by presetting 'DEFAULT_ARCHIVE'
+in the environment before configuring (the behavior of '-[0-7]' or
+'-[0-7]lmh' options in 'tar' are then derived automatically).  Similarly,
+'DEFAULT_BLOCKING' can be preset to something else than 20.
 
 ** Selecting full pathname of the "rmt" binary.
 
@@ -76,13 +76,13 @@ directory.
 Use option --enable-backup-scripts to compile and install these
 scripts.
 
-** `--disable-largefile' omits support for large files, even if the
+** '--disable-largefile' omits support for large files, even if the
 operating system supports large files.  Typically, large files are
 those larger than 2 GB on a 32-bit host.
 
 * Installation hints
 
-Here are a few hints which might help installing `tar' on some systems.
+Here are a few hints which might help installing 'tar' on some systems.
 
 ** gzip and bzip2.
 
@@ -103,7 +103,7 @@ then you have encountered a gzip incompatibility that should be fixed
 in gzip test version 1.3, which as of this writing is available at
 <ftp://alpha.gnu.org/gnu/gzip/>.  You can work around the
 incompatibility by using a shell command like
- `gzip -d <file.tar.gz | tar -xzf -'.
+ 'gzip -d <file.tar.gz | tar -xzf -'.
 
 ** Solaris issues.
 
@@ -118,13 +118,13 @@ understand these headers.
 
 ** Static linking.
 
-Some platform will, by default, prepare a smaller `tar' executable
-which depends on shared libraries.  Since GNU `tar' may be used for
+Some platform will, by default, prepare a smaller 'tar' executable
+which depends on shared libraries.  Since GNU 'tar' may be used for
 system-level backups and disaster recovery, installers might prefer to
-force static linking, making a bigger `tar' executable maybe, but able to
+force static linking, making a bigger 'tar' executable maybe, but able to
 work standalone, in situations where shared libraries are not available.
 The way to achieve static linking varies between systems.  Set LDFLAGS
-to a value from the table below, before configuration (see `INSTALL').
+to a value from the table below, before configuration (see 'INSTALL').
 
 	Platform	Compiler	LDFLAGS
 
@@ -137,18 +137,18 @@ to a value from the table below, before configuration (see `INSTALL').
 	Solaris		(vendor)	-Bstatic
 	SunOS		(vendor)	-Bstatic
 
-** Failed tests `ignfail.sh' or `incremen.sh'.
+** Failed tests 'ignfail.sh' or 'incremen.sh'.
 
 In an NFS environment, lack of synchronization between machine clocks
 might create difficulties to any tool comparing dates and file time stamps,
-like `tar' in incremental dumps.  This has been a recurrent problem with
+like 'tar' in incremental dumps.  This has been a recurrent problem with
 GNU Make for the last few years.  We would like a general solution.
 
 ** BSD compatibility matters.
 
-Set LIBS to `-lbsd' before configuration (see `INSTALL') if the linker
-complains about `bsd_ioctl' (Slackware).  Also set CPPFLAGS to
-`-I/usr/include/bsd' if <sgtty.h> is not found (Slackware).
+Set LIBS to '-lbsd' before configuration (see 'INSTALL') if the linker
+complains about 'bsd_ioctl' (Slackware).  Also set CPPFLAGS to
+'-I/usr/include/bsd' if <sgtty.h> is not found (Slackware).
 
 ** OPENStep 4.2 swap files
 
@@ -159,43 +159,43 @@ the simplest workaround is to avoid tarring this file.
 
 * Special topics
 
-Here are a few special matters about GNU `tar', not related to build
+Here are a few special matters about GNU 'tar', not related to build
 matters.  See previous section for such.
 
 ** File attributes.
 
-About *security*, it is probable that future releases of `tar' will have
+About *security*, it is probable that future releases of 'tar' will have
 some behavior changed.  There are many pending suggestions to choose from.
-Today, extracting an archive not being `root', `tar' will restore suid/sgid
-bits on files but owned by the extracting user.  `root' automatically gets
-a lot of special privileges, `-p' might later become required to get them.
+Today, extracting an archive not being 'root', 'tar' will restore suid/sgid
+bits on files but owned by the extracting user.  'root' automatically gets
+a lot of special privileges, '-p' might later become required to get them.
 
-GNU `tar' does not properly restore symlink attributes.  Various systems
+GNU 'tar' does not properly restore symlink attributes.  Various systems
 implement flavors of symbolic links showing different behavior and
 properties.  We did not successfully sorted all these out yet.  Currently,
-the `lchown' call will be used if available, but that's all.
+the 'lchown' call will be used if available, but that's all.
 
 ** POSIX compliance.
 
-GNU `tar' is able to create archive in the following formats:
+GNU 'tar' is able to create archive in the following formats:
 
   *** The format of UNIX version 7
   *** POSIX.1-1988 format, also known as "ustar format"
   *** POSIX.1-2001 format, also known as "pax format"
   *** Old GNU format (described below)
 
-In addition to those, GNU `tar' is also able to read archives
-produced by `star' archiver.
+In addition to those, GNU 'tar' is also able to read archives
+produced by 'star' archiver.
 
-A so called `Old GNU' format is based on an early draft of the
-POSIX 1003.1 `ustar' standard which is different from the final
+A so called 'Old GNU' format is based on an early draft of the
+POSIX 1003.1 'ustar' standard which is different from the final
 standard. It defines its extensions (such as incremental backups
 and handling of the long file names) in a way incompatible with
 any existing tar archive format, therefore the use of old GNU
 format is strongly discouraged.
 
 Please read the file NEWS for more information about POSIX compliance
-and new `tar' features.
+and new 'tar' features.
 
 * What's next?
 
@@ -216,14 +216,14 @@ to the report address; rather take special arrangement with the maintainer.
 
 Your feedback will help us to make a better and more portable package.
 Consider documentation errors as bugs, and report them as such.  If you
-develop anything pertaining to `tar' or have suggestions, let us know
+develop anything pertaining to 'tar' or have suggestions, let us know
 and share your findings by writing to <bug-tar@gnu.org>.
 
 
 * Copying
 
 	Copyright (C) 1990, 1991, 1992, 1994, 1997, 1998, 1999, 2000,
-	2001, 2003, 2004, 2007 Free Software Foundation, Inc.
+	2001, 2003, 2004, 2007, 2012 Free Software Foundation, Inc.
 
 	This file is part of GNU tar.
 
diff --git a/README-hacking b/README-hacking
index 35e4a30..2230535 100644
--- a/README-hacking
+++ b/README-hacking
@@ -36,15 +36,15 @@ Normally you will have to run bootstrap only once. However, if you
 intend to hack on GNU tar, you might need to run it again later. In
 this case, you will probably want to save some time and bandwidth by
 avoiding downloading the same files again. If so, create in GNU tar
-root directory the file named `.bootstrap' with the following
+root directory the file named '.bootstrap' with the following
 contents:
 
    --gnulib-srcdir=$HOME/gnulib
 
-Replace `$HOME/gnulib' with the actual directory where the Gnulib
+Replace '$HOME/gnulib' with the actual directory where the Gnulib
 sources reside.
 
-For more information about `bootstrap', run `bootstrap --help'.
+For more information about 'bootstrap', run 'bootstrap --help'.
 
 
 * Copyright information
diff --git a/configure.ac b/configure.ac
index b4dc871..f2763bd 100644
--- a/configure.ac
+++ b/configure.ac
@@ -157,7 +157,7 @@ if test -z "$DEFAULT_ARCHIVE"; then
   DEFAULT_ARCHIVE=-
 else
   if test -z "`ls $DEFAULT_ARCHIVE 2>/dev/null`"; then
-    AC_MSG_WARN(DEFAULT_ARCHIVE \`$DEFAULT_ARCHIVE' not found on this system)
+    AC_MSG_WARN(DEFAULT_ARCHIVE '$DEFAULT_ARCHIVE' not found on this system)
   fi
   # FIXME: Look for DEFTAPE in <sys/mtio.h>.
   # FIXME: Let DEVICE_PREFIX be configured from the environment.
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 205a1ad..e61699d 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -34,7 +34,7 @@ tar_TEXINFOS = \
 EXTRA_DIST = gendocs_template mastermenu.el texify.sed untabify.el
 
 # The rendering level is anyone of PUBLISH, DISTRIB or PROOF.
-# Just call `make RENDITION=PROOF [target]' if you want PROOF rendition.
+# Just call 'make RENDITION=PROOF [target]' if you want PROOF rendition.
 RENDITION = DISTRIB
 
 MAKEINFOFLAGS=-D$(RENDITION)
diff --git a/doc/tar.texi b/doc/tar.texi
index 1a28cb2..a52b322 100644
--- a/doc/tar.texi
+++ b/doc/tar.texi
@@ -13,9 +13,9 @@
 @c Maintenance notes:
 @c  1. Pay attention to @FIXME{}s and @UNREVISED{}s
 @c  2. Before creating final variant:
-@c    2.1. Run `make check-options' to make sure all options are properly
+@c    2.1. Run 'make check-options' to make sure all options are properly
 @c         documented;
-@c    2.2. Run `make master-menu' (see comment before the master menu).
+@c    2.2. Run 'make master-menu' (see comment before the master menu).
 
 @include rendition.texi
 @include value.texi
@@ -1434,7 +1434,7 @@ example:
 @smallexample
 @group
 $ @kbd{tar --create --verbose --file archive /etc/mail}
-tar: Removing leading `/' from member names
+tar: Removing leading '/' from member names
 /etc/mail/
 /etc/mail/sendmail.cf
 /etc/mail/aliases
@@ -4165,17 +4165,17 @@ Disable all warning messages.
 @item symlink-cast
 @samp{Attempting extraction of symbolic links as hard links}
 @kwindex unknown-cast
-@cindex @samp{Unknown file type `%c', extracted as normal file}, warning message
+@cindex @samp{Unknown file type '%c', extracted as normal file}, warning message
 @item unknown-cast
-@samp{%s: Unknown file type `%c', extracted as normal file}
+@samp{%s: Unknown file type '%c', extracted as normal file}
 @kwindex ignore-newer
 @cindex @samp{Current %s is newer or same age}, warning message
 @item ignore-newer
 @samp{Current %s is newer or same age}
 @kwindex unknown-keyword
-@cindex @samp{Ignoring unknown extended header keyword `%s'}, warning message
+@cindex @samp{Ignoring unknown extended header keyword '%s'}, warning message
 @item unknown-keyword
-@samp{Ignoring unknown extended header keyword `%s'}
+@samp{Ignoring unknown extended header keyword '%s'}
 @kwindex decompress-program
 @item decompress-program
 Controls verbose description of failures occurring when trying to run
@@ -4952,7 +4952,7 @@ For example:
 
 @smallexample
 $ @kbd{tar -c -f archive.tar -v --mtime=yesterday .}
-tar: Option --mtime: Treating date `yesterday' as 2006-06-20
+tar: Option --mtime: Treating date 'yesterday' as 2006-06-20
 13:06:29.152478
 @dots{}
 @end smallexample
@@ -6945,7 +6945,7 @@ When @command{tar} is invoked with @option{--create} (@option{-c}),
 @group
 $ @kbd{tar cf a.tar}
 tar: Cowardly refusing to create an empty archive
-Try `tar --help' or `tar --usage' for more information.
+Try 'tar --help' or 'tar --usage' for more information.
 @end group
 @end smallexample
 
@@ -7836,7 +7836,7 @@ $ @kbd{tar tf arch.tar --quoting-style=escape}
 Control characters, single quote and backslash are printed using
 backslash notation.  All names are quoted using left and right
 quotation marks, appropriate to the current locale.  If it does not
-define quotation marks, use @samp{`} as left and @samp{'} as right
+define quotation marks, use @samp{'} as left and as right
 quotation marks.  Any occurrences of the right quotation mark in a
 name are escaped with @samp{\}, for example:
 
@@ -7845,13 +7845,13 @@ For example:
 @smallexample
 @group
 $ @kbd{tar tf arch.tar --quoting-style=locale}
-`./'
-`./a space'
-`./a\'single\'quote'
-`./a"double"quote'
-`./a\\backslash'
-`./a\ttab'
-`./a\nnewline'
+'./'
+'./a space'
+'./a\'single\'quote'
+'./a"double"quote'
+'./a\\backslash'
+'./a\ttab'
+'./a\nnewline'
 @end group
 @end smallexample
 
@@ -8298,7 +8298,7 @@ ensure he is using the right date.  For example:
 @smallexample
 @group
 $ @kbd{tar -c -f archive.tar --after-date='10 days ago' .}
-tar: Option --after-date: Treating date `10 days ago' as 2006-06-11
+tar: Option --after-date: Treating date '10 days ago' as 2006-06-11
 13:19:37.232434
 @end group
 @end smallexample
@@ -9478,7 +9478,7 @@ produces the following diagnostics:
 
 @smallexample
 $ tar -c -f ../archive.tar -l jeden
-tar: Missing links to `jeden'.
+tar: Missing links to 'jeden'.
 @end smallexample
 
 Although creating special records for hard links helps keep a faithful
@@ -9490,7 +9490,7 @@ archive created in previous examples produces, in the absense of file
 
 @smallexample
 $ tar xf archive.tar ./one
-tar: ./one: Cannot hard link to `./jeden': No such file or directory
+tar: ./one: Cannot hard link to './jeden': No such file or directory
 tar: Error exit delayed from previous errors
 @end smallexample
 
@@ -10030,8 +10030,8 @@ run mode is enabled by @option{-n} command line argument:
 @group
 $ @kbd{xsparse -n /home/gray/GNUSparseFile.6058/sparsefile}
 Reading v.1.0 sparse map
-Expanding file `/home/gray/GNUSparseFile.6058/sparsefile' to
-`/home/gray/sparsefile'
+Expanding file '/home/gray/GNUSparseFile.6058/sparsefile' to
+'/home/gray/sparsefile'
 Finished dry run
 @end group
 @end smallexample
@@ -10052,8 +10052,8 @@ similar to that from the dry run mode, use @option{-v} option:
 @group
 $ @kbd{xsparse -v /home/gray/GNUSparseFile.6058/sparsefile}
 Reading v.1.0 sparse map
-Expanding file `/home/gray/GNUSparseFile.6058/sparsefile' to
-`/home/gray/sparsefile'
+Expanding file '/home/gray/GNUSparseFile.6058/sparsefile' to
+'/home/gray/sparsefile'
 Done
 @end group
 @end smallexample
@@ -10074,8 +10074,8 @@ Found variable GNU.sparse.minor = 0
 Found variable GNU.sparse.name = sparsefile
 Found variable GNU.sparse.realsize = 217481216
 Reading v.1.0 sparse map
-Expanding file `/home/gray/GNUSparseFile.6058/sparsefile' to
-`/home/gray/sparsefile'
+Expanding file '/home/gray/GNUSparseFile.6058/sparsefile' to
+'/home/gray/sparsefile'
 Done
 @end group
 @end smallexample
@@ -10177,7 +10177,7 @@ Found variable GNU.sparse.size = 217481216
 Found variable GNU.sparse.numblocks = 208
 Found variable GNU.sparse.name = sparsefile
 Found variable GNU.sparse.map = 0,2048,1050624,2048,@dots{}
-Expanding file `GNUSparseFile.28124/sparsefile' to `sparsefile'
+Expanding file 'GNUSparseFile.28124/sparsefile' to 'sparsefile'
 Done
 @end group
 @end smallexample
@@ -11299,7 +11299,7 @@ is@footnote{If you run @GNUTAR{} under a different locale, the
 translation to the locale's language will be used.}:
 
 @smallexample
-Prepare volume #@var{n} for `@var{archive}' and hit return:
+Prepare volume #@var{n} for '@var{archive}' and hit return:
 @end smallexample
 
 @noindent
@@ -11668,7 +11668,7 @@ you will get:
 @smallexample
 @group
 $ @kbd{tar -rf archive --label 'My volume' .}
-tar: Archive not labeled to match `My volume'
+tar: Archive not labeled to match 'My volume'
 @end group
 @end smallexample
 
@@ -11982,7 +11982,7 @@ Conversely, with the @option{--keep-old-files} (@option{-k}) and
 @option{--skip-old-files} options, @command{tar} refuses to replace
 existing files when extracting.  The difference between the two
 options is that the former treats existing files as errors whereas the
-latter just silently ignores them.  
+latter just silently ignores them.
 
 Finally, with the @option{--no-overwrite-dir} option, @command{tar}
 refuses to replace the permissions or ownership of already-existing
diff --git a/scripts/backup.in b/scripts/backup.in
index 37bfed3..6bbd47c 100644
--- a/scripts/backup.in
+++ b/scripts/backup.in
@@ -149,7 +149,7 @@ message 20 "BACKUP_DIRS=$BACKUP_DIRS"
 message 20 "BACKUP_FILES=$BACKUP_FILES"
 
 # The buch of commands below is run in a subshell for which all output is
-# piped through `tee' to the logfile.  Doing this, instead of having
+# piped through 'tee' to the logfile.  Doing this, instead of having
 # multiple pipelines all over the place, is cleaner and allows access to
 # the exit value from various commands more easily.
 (
@@ -187,7 +187,7 @@ message 20 "BACKUP_FILES=$BACKUP_FILES"
                 "--label='`print_level` backup of ${fs} on ${remotehost} at ${NOW}'" \
 		-C ${fs} .
 
-     # `rsh' doesn't exit with the exit status of the remote command.  What
+     # 'rsh' doesn't exit with the exit status of the remote command.  What
      # stupid lossage.  TODO: think of a reliable workaround.
      if [ $? -ne 0 ] ; then
 	 echo "Backup of ${1} failed." 1>&2
diff --git a/scripts/backup.sh.in b/scripts/backup.sh.in
index adcdfe2..cd05626 100644
--- a/scripts/backup.sh.in
+++ b/scripts/backup.sh.in
@@ -48,7 +48,7 @@ MT_REWIND=mt_rewind
 MT_OFFLINE=mt_offline
 MT_STATUS=mt_status
 
-# Insure `mail' is in PATH.
+# Insure 'mail' is in PATH.
 PATH="/usr/ucb:${PATH}"
 export PATH
 # Put startdate in the subject line of mailed report, since if it happens
@@ -211,8 +211,8 @@ init_backup() {
 	TAR_PART1="${TAR_PART1} --info-script='${DUMP_REMIND_SCRIPT}'"
     fi
     # Set logfile name
-    # Logfile name should be in the form  ``log-1993-03-18-level-0''
-    # They go in the directory `@sysconfdir@/log'.
+    # Logfile name should be in the form 'log-1993-03-18-level-0'
+    # They go in the directory '@sysconfdir@/log'.
     # i.e. year-month-date.  This format is useful for sorting by name, since
     # logfiles are intentionally kept online for future reference.
     LOGFILE="${LOGPATH}/log-`now`-level-${DUMP_LEVEL}"
@@ -305,7 +305,7 @@ backup_host() {
     if [ "z${localhost}" != "z$rhost" ] ; then
 	$RSH "$rhost" ${TAR_PART1} -f "${localhost}:${TAPE_FILE}" $@
     else
-	# Using `sh -c exec' causes nested quoting and shell substitution
+	# Using 'sh -c exec' causes nested quoting and shell substitution
         # to be handled here in the same way rsh handles it.
         CMD="exec ${TAR_PART1} -f \"${TAPE_FILE}\" $@"
         message 10 "CMD: $CMD"
diff --git a/scripts/dump-remind.in b/scripts/dump-remind.in
index a3f013f..f66271a 100644
--- a/scripts/dump-remind.in
+++ b/scripts/dump-remind.in
@@ -9,7 +9,7 @@
 # be put in the tape drive.
 #
 
-# Include location of `sendmail' and GNU finger.
+# Include location of 'sendmail' and GNU finger.
 PATH="/usr/lib:/usr/local/gnubin:${PATH}"
 export PATH
 
@@ -25,8 +25,8 @@ MT_OFFLINE
 # which users are logged into consoles (and thus in the office and capable
 # of changing tapes).
 #
-# Certain users (like `root') aren't real users, and shouldn't be notified.
-# Neither should `zippy', `elvis', etc. (on the GNU machines) since they're
+# Certain users (like 'root') aren't real users, and shouldn't be notified.
+# Neither should 'zippy', 'elvis', etc. (on the GNU machines) since they're
 # just test accounts.
 recipients="`
     finger .clients 2> /dev/null \
diff --git a/scripts/xsparse.c b/scripts/xsparse.c
index 14d5658..8f42c37 100644
--- a/scripts/xsparse.c
+++ b/scripts/xsparse.c
@@ -164,7 +164,7 @@ get_var (FILE *fp, char **name, char **value)
   p += 11;
   q = strchr (p, '=');
   if (!q)
-    die (1, "malformed header: expected `=' not found");
+    die (1, "malformed header: expected '=' not found");
   *q++ = 0;
   q[strlen (q) - 1] = 0;
   *name = p;
@@ -194,7 +194,7 @@ read_xheader (char *name)
 	printf ("Found variable GNU.sparse.%s = %s\n", kw, val);
 
       if (expect && strcmp (kw, expect))
-	die (1, "bad keyword sequence: expected `%s' but found `%s'",
+	die (1, "bad keyword sequence: expected '%s' but found '%s'",
 	     expect, kw);
       expect = NULL;
       if (strcmp (kw, "name") == 0)
@@ -235,13 +235,13 @@ read_xheader (char *name)
 	    {
 	      sparse_map[i].offset = string_to_off (val, &val);
 	      if (*val != ',')
-		die (1, "bad GNU.sparse.map: expected `,' but found `%c'",
+		die (1, "bad GNU.sparse.map: expected ',' but found '%c'",
 		     *val);
 	      sparse_map[i].numbytes = string_to_off (val+1, &val);
 	      if (*val != ',')
 		{
 		  if (!(*val == 0 && i == sparse_map_size-1))
-		    die (1, "bad GNU.sparse.map: expected `,' but found `%c'",
+		    die (1, "bad GNU.sparse.map: expected ',' but found '%c'",
 			 *val);
 		}
 	      else
@@ -252,7 +252,7 @@ read_xheader (char *name)
 	}
     }
   if (expect)
-    die (1, "bad keyword sequence: expected `%s' not found", expect);
+    die (1, "bad keyword sequence: expected '%s' not found", expect);
   if (version_major == 0 && sparse_map_size == 0)
     die (1, "size of the sparse map unknown");
   if (i != sparse_map_size)
@@ -449,7 +449,7 @@ main (int argc, char **argv)
     die (1, "cannot open file %s (%d)", outname, errno);
 
   if (verbose)
-    printf ("Expanding file `%s' to `%s'\n", inname, outname);
+    printf ("Expanding file '%s' to '%s'\n", inname, outname);
 
   if (dry_run)
     {
diff --git a/src/buffer.c b/src/buffer.c
index e52b1b1..4a980b5 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -103,7 +103,7 @@ bool write_archive_to_stdout;
 
 /* Multi-volume tracking support */
 
-/* When creating a multi-volume archive, each `bufmap' represents
+/* When creating a multi-volume archive, each 'bufmap' represents
    a member stored (perhaps partly) in the current record buffer.
    After flushing the record to the output media, all bufmaps that
    represent fully written members are removed from the list, then
@@ -337,23 +337,23 @@ const char *
 first_decompress_program (int *pstate)
 {
   struct zip_program const *zp;
-  
+
   if (use_compress_program_option)
     return use_compress_program_option;
 
   if (archive_compression_type == ct_none)
     return NULL;
 
-  *pstate = 0; 
+  *pstate = 0;
   zp = find_zip_program (archive_compression_type, pstate);
   return zp ? zp->program : NULL;
 }
-    
+
 const char *
 next_decompress_program (int *pstate)
 {
   struct zip_program const *zp;
-  
+
   if (use_compress_program_option)
     return NULL;
   zp = find_zip_program (archive_compression_type, pstate);
diff --git a/src/checkpoint.c b/src/checkpoint.c
index a2896ab..f2f17a2 100644
--- a/src/checkpoint.c
+++ b/src/checkpoint.c
@@ -215,16 +215,16 @@ run_checkpoint_actions (bool do_write)
 	    if (!str)
 	      {
 		if (do_write)
-		  /* TRANSLATORS: This is a ``checkpoint of write operation'',
- 		     *not* ``Writing a checkpoint''.
-		     E.g. in Spanish ``Punto de comprobaci@'on de escritura'',
-		     *not* ``Escribiendo un punto de comprobaci@'on'' */
+		  /* TRANSLATORS: This is a "checkpoint of write operation",
+		     *not* "Writing a checkpoint".
+		     E.g. in Spanish "Punto de comprobaci@'on de escritura",
+		     *not* "Escribiendo un punto de comprobaci@'on" */
 		  str = gettext ("Write checkpoint %u");
 		else
-		  /* TRANSLATORS: This is a ``checkpoint of read operation'',
-	             *not* ``Reading a checkpoint''.
-		     E.g. in Spanish ``Punto de comprobaci@'on de lectura'',
-		     *not* ``Leyendo un punto de comprobaci@'on'' */
+		  /* TRANSLATORS: This is a "checkpoint of read operation",
+	             *not* "Reading a checkpoint".
+		     E.g. in Spanish "Punto de comprobaci@'on de lectura",
+		     *not* "Leyendo un punto de comprobaci@'on" */
 		  str = gettext ("Read checkpoint %u");
 	      }
 	    tmp = expand_checkpoint_string (str, do_write, checkpoint);
diff --git a/src/compare.c b/src/compare.c
index 639e935..2fc72b7 100644
--- a/src/compare.c
+++ b/src/compare.c
@@ -477,7 +477,7 @@ diff_archive (void)
   switch (current_header->header.typeflag)
     {
     default:
-      ERROR ((0, 0, _("%s: Unknown file type `%c', diffed as normal file"),
+      ERROR ((0, 0, _("%s: Unknown file type '%c', diffed as normal file"),
 	      quotearg_colon (current_stat_info.file_name),
 	      current_header->header.typeflag));
       /* Fall through.  */
diff --git a/src/create.c b/src/create.c
index 2e0bfc3..745796c 100644
--- a/src/create.c
+++ b/src/create.c
@@ -785,9 +785,9 @@ start_header (struct tar_stat_info *st)
 	. . . . . . . . .   9 = Omron UNIOS-B 4.3BSD 1.60Beta
 
 	     . = works
-	     # = ``impossible file type''
+	     # = "impossible file type"
 
-     The following mask for old archive removes the `#'s in column 4
+     The following mask for old archive removes the '#'s in column 4
      above, thus making GNU tar both a universal donor and a universal
      acceptor for Paul's test.  */
 
diff --git a/src/extract.c b/src/extract.c
index bd5e7bf..0e499e5 100644
--- a/src/extract.c
+++ b/src/extract.c
@@ -1386,7 +1386,7 @@ prepare_to_extract (char const *file_name, int typeflag, tar_extractor_t *fun)
     default:
       WARNOPT (WARN_UNKNOWN_CAST,
 	       (0, 0,
-		_("%s: Unknown file type `%c', extracted as normal file"),
+		_("%s: Unknown file type '%c', extracted as normal file"),
 		quotearg_colon (file_name), typeflag));
       *fun = extract_file;
     }
diff --git a/src/incremen.c b/src/incremen.c
index 69cbd59..04b946f 100644
--- a/src/incremen.c
+++ b/src/incremen.c
@@ -848,7 +848,7 @@ store_rename (struct directory *dir, struct obstack *stk)
 	 are ignored when hit by this function next time.
 	 If the chain forms a cycle, prev points to the entry DIR is renamed
 	 from. In this case it still retains DIRF_RENAMED flag, which will be
-	 cleared in the `else' branch below */
+	 cleared in the 'else' branch below */
       for (prev = dir; prev && prev->orig != dir; prev = prev->orig)
 	DIR_CLEAR_FLAG (prev, DIRF_RENAMED);
 
diff --git a/src/misc.c b/src/misc.c
index 3add371..126161b 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -105,7 +105,7 @@ quote_copy_string (const char *string)
    completes the unquoting anyway.
 
    This is used for reading the saved directory file in incremental
-   dumps.  It is used for decoding old `N' records (demangling names).
+   dumps.  It is used for decoding old 'N' records (demangling names).
    But also, it is used for decoding file arguments, would they come
    from the shell or a -T file, and for decoding the --exclude
    argument.  */
@@ -962,7 +962,7 @@ page_aligned_alloc (void **ptr, size_t size)
 
 struct namebuf
 {
-  char *buffer;		/* directory, `/', and directory member */
+  char *buffer;		/* directory, '/', and directory member */
   size_t buffer_size;	/* allocated size of name_buffer */
   size_t dir_length;	/* length of directory part in buffer */
 };
diff --git a/src/names.c b/src/names.c
index ba4d509..f533c08 100644
--- a/src/names.c
+++ b/src/names.c
@@ -669,9 +669,9 @@ label_notfound (void)
 
 /* Sort *singly* linked LIST of names, of given LENGTH, using COMPARE
    to order names.  Return the sorted list.  Note that after calling
-   this function, the `prev' links in list elements are messed up.
+   this function, the 'prev' links in list elements are messed up.
 
-   Apart from the type `struct name' and the definition of SUCCESSOR,
+   Apart from the type 'struct name' and the definition of SUCCESSOR,
    this is a generic list-sorting function, but it's too painful to
    make it both generic and portable
    in C.  */
@@ -894,7 +894,7 @@ name_compare (void const *entry1, void const *entry2)
 }
 
 
-/* Rebase `name' member of CHILD and all its siblings to
+/* Rebase 'name' member of CHILD and all its siblings to
    the new PARENT. */
 static void
 rebase_child_list (struct name *child, struct name *parent)
diff --git a/src/sparse.c b/src/sparse.c
index 8e5ad28..f7a9fe7 100644
--- a/src/sparse.c
+++ b/src/sparse.c
@@ -591,18 +591,18 @@ sparse_diff_file (int fd, struct tar_stat_info *st)
 /* Old GNU Format. The sparse file information is stored in the
    oldgnu_header in the following manner:
 
-   The header is marked with type 'S'. Its `size' field contains
+   The header is marked with type 'S'. Its 'size' field contains
    the cumulative size of all non-empty blocks of the file. The
-   actual file size is stored in `realsize' member of oldgnu_header.
+   actual file size is stored in 'realsize' member of oldgnu_header.
 
-   The map of the file is stored in a list of `struct sparse'.
+   The map of the file is stored in a list of 'struct sparse'.
    Each struct contains offset to the block of data and its
    size (both as octal numbers). The first file header contains
    at most 4 such structs (SPARSES_IN_OLDGNU_HEADER). If the map
-   contains more structs, then the field `isextended' of the main
-   header is set to 1 (binary) and the `struct sparse_header'
+   contains more structs, then the field 'isextended' of the main
+   header is set to 1 (binary) and the 'struct sparse_header'
    header follows, containing at most 21 following structs
-   (SPARSES_IN_SPARSE_HEADER). If more structs follow, `isextended'
+   (SPARSES_IN_SPARSE_HEADER). If more structs follow, 'isextended'
    field of the extended header is set and next  next extension header
    follows, etc... */
 
diff --git a/src/tar.c b/src/tar.c
index 9cea7a5..4f68534 100644
--- a/src/tar.c
+++ b/src/tar.c
@@ -78,7 +78,7 @@ static void
 request_stdin (const char *option)
 {
   if (stdin_used_by)
-    USAGE_ERROR ((0, 0, _("Options `-%s' and `-%s' both want standard input"),
+    USAGE_ERROR ((0, 0, _("Options '-%s' and '-%s' both want standard input"),
 		  stdin_used_by, option));
 
   stdin_used_by = option;
@@ -247,7 +247,7 @@ tar_set_quoting_style (char *arg)
 	return;
       }
   FATAL_ERROR ((0, 0,
-		_("Unknown quoting style `%s'. Try `%s --quoting-style=help' to get a list."), arg, program_invocation_short_name));
+		_("Unknown quoting style '%s'. Try '%s --quoting-style=help' to get a list."), arg, program_invocation_short_name));
 }
 
 
@@ -347,7 +347,7 @@ enum
 const char *argp_program_version = "tar (" PACKAGE_NAME ") " VERSION;
 const char *argp_program_bug_address = "<" PACKAGE_BUGREPORT ">";
 static char const doc[] = N_("\
-GNU `tar' saves many files together into a single tape or disk archive, \
+GNU 'tar' saves many files together into a single tape or disk archive, \
 and can restore individual files from the archive.\n\
 \n\
 Examples:\n\
@@ -355,7 +355,7 @@ Examples:\n\
   tar -tvf archive.tar         # List all files in archive.tar verbosely.\n\
   tar -xf archive.tar          # Extract all files from archive.tar.\n")
 "\v"
-N_("The backup suffix is `~', unless set with --suffix or SIMPLE_BACKUP_SUFFIX.\n\
+N_("The backup suffix is '~', unless set with --suffix or SIMPLE_BACKUP_SUFFIX.\n\
 The version control may be set with --backup or VERSION_CONTROL, values are:\n\n\
   none, off       never make backups\n\
   t, numbered     make numbered backups\n\
@@ -543,7 +543,7 @@ static struct argp_option options[] = {
   {"rsh-command", RSH_COMMAND_OPTION, N_("COMMAND"), 0,
    N_("use remote COMMAND instead of rsh"), GRID+1 },
 #ifdef DEVICE_PREFIX
-  {"-[0-7][lmh]", 0, NULL, OPTION_DOC, /* It is OK, since `name' will never be
+  {"-[0-7][lmh]", 0, NULL, OPTION_DOC, /* It is OK, since 'name' will never be
 					  translated */
    N_("specify drive and density"), GRID+1 },
 #endif
@@ -685,7 +685,7 @@ static struct argp_option options[] = {
   {"recursion", RECURSION_OPTION, 0, 0,
    N_("recurse into directories (default)"), GRID+1 },
   {"absolute-names", 'P', 0, 0,
-   N_("don't strip leading `/'s from file names"), GRID+1 },
+   N_("don't strip leading '/'s from file names"), GRID+1 },
   {"dereference", 'h', 0, 0,
    N_("follow symlinks; archive and dump the files they point to"), GRID+1 },
   {"hard-dereference", HARD_DEREFERENCE_OPTION, 0, 0,
@@ -723,7 +723,7 @@ static struct argp_option options[] = {
   {"anchored", ANCHORED_OPTION, 0, 0,
    N_("patterns match file name start"), GRID+1 },
   {"no-anchored", NO_ANCHORED_OPTION, 0, 0,
-   N_("patterns match after any `/' (default for exclusion)"), GRID+1 },
+   N_("patterns match after any '/' (default for exclusion)"), GRID+1 },
   {"no-ignore-case", NO_IGNORE_CASE_OPTION, 0, 0,
    N_("case sensitive matching (default)"), GRID+1 },
   {"wildcards", WILDCARDS_OPTION, 0, 0,
@@ -731,9 +731,9 @@ static struct argp_option options[] = {
   {"no-wildcards", NO_WILDCARDS_OPTION, 0, 0,
    N_("verbatim string matching"), GRID+1 },
   {"no-wildcards-match-slash", NO_WILDCARDS_MATCH_SLASH_OPTION, 0, 0,
-   N_("wildcards do not match `/'"), GRID+1 },
+   N_("wildcards do not match '/'"), GRID+1 },
   {"wildcards-match-slash", WILDCARDS_MATCH_SLASH_OPTION, 0, 0,
-   N_("wildcards match `/' (default for exclusion)"), GRID+1 },
+   N_("wildcards match '/' (default for exclusion)"), GRID+1 },
 #undef GRID
 
 #define GRID 130
@@ -932,7 +932,7 @@ set_subcommand_option (enum subcommand subcommand)
   if (subcommand_option != UNKNOWN_SUBCOMMAND
       && subcommand_option != subcommand)
     USAGE_ERROR ((0, 0,
-		  _("You may not specify more than one `-Acdtrux' or `--test-label' option")));
+		  _("You may not specify more than one '-Acdtrux' or '--test-label' option")));
 
   subcommand_option = subcommand;
 }
@@ -1062,7 +1062,7 @@ report_textual_dates (struct tar_args *args)
 	{
 	  char const *treated_as = tartime (p->ts, true);
 	  if (strcmp (p->date, treated_as) != 0)
-	    WARN ((0, 0, _("Option %s: Treating date `%s' as %s"),
+	    WARN ((0, 0, _("Option %s: Treating date '%s' as %s"),
 		   p->option, p->date, treated_as));
 	}
       free (p->date);
@@ -2208,7 +2208,7 @@ parse_opt (int key, char *arg, struct argp_state *state)
 	    break;
 
 	  default:
-	    argp_error (state, _("Unknown density: `%c'"), arg[0]);
+	    argp_error (state, _("Unknown density: '%c'"), arg[0]);
 	  }
 	sprintf (cursor, "%d", device);
 
@@ -2225,7 +2225,7 @@ parse_opt (int key, char *arg, struct argp_state *state)
 #else /* not DEVICE_PREFIX */
 
       argp_error (state,
-		  _("Options `-[0-7][lmh]' not supported by *this* tar"));
+		  _("Options '-[0-7][lmh]' not supported by *this* tar"));
 
 #endif /* not DEVICE_PREFIX */
 
@@ -2356,7 +2356,7 @@ decode_options (int argc, char **argv)
 	      if (in < argv + argc)
 		*out++ = *in++;
 	      else
-		USAGE_ERROR ((0, 0, _("Old option `%c' requires an argument."),
+		USAGE_ERROR ((0, 0, _("Old option '%c' requires an argument."),
 			      *letter));
 	    }
 	}
@@ -2457,11 +2457,11 @@ decode_options (int argc, char **argv)
 	archive_name_array[0] = DEFAULT_ARCHIVE;
     }
 
-  /* Allow multiple archives only with `-M'.  */
+  /* Allow multiple archives only with '-M'.  */
 
   if (archive_names > 1 && !multi_volume_option)
     USAGE_ERROR ((0, 0,
-		  _("Multiple archive files require `-M' option")));
+		  _("Multiple archive files require '-M' option")));
 
   if (listed_incremental_option
       && NEWER_OPTION_INITIALIZED (newer_mtime_option))
@@ -2561,7 +2561,7 @@ decode_options (int argc, char **argv)
     USAGE_ERROR ((0, 0, _("--preserve-order is not compatible with "
 			  "--listed-incremental")));
 
-  /* Forbid using -c with no input files whatsoever.  Check that `-f -',
+  /* Forbid using -c with no input files whatsoever.  Check that '-f -',
      explicit or implied, is used correctly.  */
 
   switch (subcommand_option)
@@ -2595,7 +2595,7 @@ decode_options (int argc, char **argv)
 	   archive_name_cursor++)
 	if (!strcmp (*archive_name_cursor, "-"))
 	  USAGE_ERROR ((0, 0,
-			_("Options `-Aru' are incompatible with `-f -'")));
+			_("Options '-Aru' are incompatible with '-f -'")));
 
     default:
       break;
@@ -2684,7 +2684,7 @@ main (int argc, char **argv)
     {
     case UNKNOWN_SUBCOMMAND:
       USAGE_ERROR ((0, 0,
-		    _("You must specify one of the `-Acdtrux' or `--test-label'  options")));
+		    _("You must specify one of the '-Acdtrux' or '--test-label'  options")));
 
     case CAT_SUBCOMMAND:
     case UPDATE_SUBCOMMAND:
diff --git a/src/tar.h b/src/tar.h
index ce9850c..0b090af 100644
--- a/src/tar.h
+++ b/src/tar.h
@@ -80,9 +80,9 @@ struct posix_header
 /* tar Header Block, GNU extensions.  */
 
 /* In GNU tar, SYMTYPE is for to symbolic links, and CONTTYPE is for
-   contiguous files, so maybe disobeying the `reserved' comment in POSIX
+   contiguous files, so maybe disobeying the "reserved" comment in POSIX
    header description.  I suspect these were meant to be used this way, and
-   should not have really been `reserved' in the published standards.  */
+   should not have really been "reserved" in the published standards.  */
 
 /* *BEWARE* *BEWARE* *BEWARE* that the following information is still
    boiling, and may change.  Even if the OLDGNU format description should be
diff --git a/src/transform.c b/src/transform.c
index 77aa0a2..e9b776a 100644
--- a/src/transform.c
+++ b/src/transform.c
@@ -347,7 +347,7 @@ parse_transform_expr (const char *expr)
 	      break;
 
 	    case 'L':
-	      /* Turn the replacement to lowercase until a `\U' or `\E'
+	      /* Turn the replacement to lowercase until a '\U' or '\E'
 		 is found, */
 	      add_case_ctl_segment (tf, ctl_locase);
 	      cur++;
@@ -360,7 +360,7 @@ parse_transform_expr (const char *expr)
 	      break;
 
 	    case 'U':
-	      /* Turn the replacement to uppercase until a `\L' or `\E'
+	      /* Turn the replacement to uppercase until a '\L' or '\E'
 		 is found, */
 	      add_case_ctl_segment (tf, ctl_upcase);
 	      cur++;
@@ -373,7 +373,7 @@ parse_transform_expr (const char *expr)
 	      break;
 
 	    case 'E':
-	      /* Stop case conversion started by `\L' or `\U'. */
+	      /* Stop case conversion started by '\L' or '\U'. */
 	      add_case_ctl_segment (tf, ctl_stop);
 	      cur++;
 	      break;
diff --git a/src/xheader.c b/src/xheader.c
index 2284e97..0946493 100644
--- a/src/xheader.c
+++ b/src/xheader.c
@@ -86,10 +86,10 @@ struct keyword_list
 /* List of keyword patterns set by delete= option */
 static struct keyword_list *keyword_pattern_list;
 
-/* List of keyword/value pairs set by `keyword=value' option */
+/* List of keyword/value pairs set by 'keyword=value' option */
 static struct keyword_list *keyword_global_override_list;
 
-/* List of keyword/value pairs set by `keyword:=value' option */
+/* List of keyword/value pairs set by 'keyword:=value' option */
 static struct keyword_list *keyword_override_list;
 
 /* List of keyword/value pairs decoded from the last 'g' type header */
@@ -618,7 +618,7 @@ decx (void *data, char const *keyword, char const *value, size_t size)
     t->decoder (st, keyword, value, size);
   else
     WARNOPT (WARN_UNKNOWN_KEYWORD,
-	     (0, 0, _("Ignoring unknown extended header keyword `%s'"),
+	     (0, 0, _("Ignoring unknown extended header keyword '%s'"),
 	      keyword));
 }
 
diff --git a/tests/backup01.at b/tests/backup01.at
index cf0a907..239444b 100644
--- a/tests/backup01.at
+++ b/tests/backup01.at
@@ -43,7 +43,7 @@ tar xfv test.tar --backup --warning=no-timestamp
 [0],
 [dir1/
 dir1/file1
-Renaming `dir1/file1' to `dir1/file1~'
+Renaming 'dir1/file1' to 'dir1/file1~'
 dir2/
 ])
 
diff --git a/tests/incr01.at b/tests/incr01.at
index fad8d3e..8915683 100644
--- a/tests/incr01.at
+++ b/tests/incr01.at
@@ -46,7 +46,7 @@ tar xvfg archive.1 /dev/null --warning=no-timestamp
 directory/bar
 separator
 directory/
-tar: Deleting `directory/bar'
+tar: Deleting 'directory/bar'
 ],
 [],[],[],[gnu, oldgnu, posix])
 
diff --git a/tests/incr02.at b/tests/incr02.at
index 38b78a5..6d83b10 100644
--- a/tests/incr02.at
+++ b/tests/incr02.at
@@ -50,9 +50,9 @@ genfile --stat=mtime dir/subdir1 > ts
 # dir/subdir2/
 # dir/subdir1/foofile
 #
-# When restoring from this directory structure, `dir/subdir2/' used to
+# When restoring from this directory structure, 'dir/subdir2/' used to
 # trigger apply_nonancestor_delayed_set_stat() which restored stats for
-# `subdir1' prior to restoring `dir/subdir1/foofile'. Then, restoring the
+# 'subdir1' prior to restoring 'dir/subdir1/foofile'. Then, restoring the
 # latter clobbered the directory timestamp.
 
 tar -cf archive -g db dir
diff --git a/tests/incr04.at b/tests/incr04.at
index 5658ac9..dda0e0d 100644
--- a/tests/incr04.at
+++ b/tests/incr04.at
@@ -58,7 +58,7 @@ a/c/
 ],
 [tar: a: Directory is new
 tar: a/b: Directory is new
-tar: a/c: Directory has been renamed from `a/b'
+tar: a/c: Directory has been renamed from 'a/b'
 ],[],[],[gnu, oldgnu, posix])
 
 AT_CLEANUP
diff --git a/tests/incremental.at b/tests/incremental.at
index 5873d83..49b6d7b 100644
--- a/tests/incremental.at
+++ b/tests/incremental.at
@@ -35,7 +35,7 @@ ls -l structure/file >/dev/null
 # If the time of an initial backup and the creation time of a file contained
 # in that backup are the same, the file will be backed up again when an
 # incremental backup is done, because the incremental backup backs up
-# files created `on or after' the initial backup time.  Without the sleep
+# files created "on or after" the initial backup time.  Without the sleep
 # command, behaviour of tar becomes variable, depending whether the system
 # clock ticked over to the next second between creating the file and
 # backing it up.
diff --git a/tests/label04.at b/tests/label04.at
index fc3aa26..6cb81b0 100644
--- a/tests/label04.at
+++ b/tests/label04.at
@@ -43,7 +43,7 @@ tar -rf archive --label='New volume' file
 ],
 [# Create volume
 # Update: wrong label
-tar: Volume `New volume' does not match `My volume'
+tar: Volume 'New volume' does not match 'My volume'
 tar: Error is not recoverable: exiting now
 # Update: right label
 ],[],[],[gnu,oldgnu,posix])
diff --git a/tests/label05.at b/tests/label05.at
index 5b968a1..d24cbcb 100644
--- a/tests/label05.at
+++ b/tests/label05.at
@@ -40,7 +40,7 @@ tar -rf archive file
 ],
 [# Create volume
 # Update: wrong label
-tar: Archive not labeled to match `My volume'
+tar: Archive not labeled to match 'My volume'
 tar: Error is not recoverable: exiting now
 # Update: right label
 ],[],[],[gnu,oldgnu,posix])
diff --git a/tests/link03.at b/tests/link03.at
index 30272bf..6018844 100644
--- a/tests/link03.at
+++ b/tests/link03.at
@@ -48,7 +48,7 @@ testing archive.2
 file1
 file1
 ],
-[tar: Missing links to `file1'.
+[tar: Missing links to 'file1'.
 ])
 
 AT_CLEANUP
diff --git a/tests/listed02.at b/tests/listed02.at
index 94a339c..c292e4a 100644
--- a/tests/listed02.at
+++ b/tests/listed02.at
@@ -126,8 +126,8 @@ tart/c0/cq2
 tart/c1/ca1
 tart/c1/ca2
 Extracting incremental archive
-tar: Deleting `tart/a1'
-tar: Deleting `tart/b1'
+tar: Deleting 'tart/a1'
+tar: Deleting 'tart/b1'
 tart/
 tart/b2
 tart/c0/
@@ -149,7 +149,7 @@ tar: tart/c0: Directory is new
 tar: tart/c1: Directory is new
 tar: tart: Directory is new
 Creating incremental archive
-tar: tart/c2: Directory has been renamed from `tart/c1'
+tar: tart/c2: Directory has been renamed from 'tart/c1'
 Extracting main archive
 ],
 [],[],[gnu, oldgnu])
diff --git a/tests/multiv03.at b/tests/multiv03.at
index e013c19..eb58437 100644
--- a/tests/multiv03.at
+++ b/tests/multiv03.at
@@ -36,7 +36,7 @@ BFILE=`awk 'BEGIN { for (i = 0; i < 101; i++) printf "b"; exit; }'`
 
 cat > ../experr <<EOF
 tar: $BFILE: file name too long to be stored in a GNU multivolume header, truncated
-tar: \`$BFILE' is possibly continued on this volume: header contains truncated name
+tar: '$BFILE' is possibly continued on this volume: header contains truncated name
 EOF
 
 cat > ../expout <<EOF
diff --git a/tests/multiv05.at b/tests/multiv05.at
index 9d963c7..0bb2872 100644
--- a/tests/multiv05.at
+++ b/tests/multiv05.at
@@ -58,7 +58,7 @@ m4_foreach([file],
   [Diffing file
 ])dnl
 ,
-[tar: `trzy' is not continued on this volume
+[tar: 'trzy' is not continued on this volume
 ],[],[], [gnu])
 
 AT_CLEANUP
diff --git a/tests/multiv08.at b/tests/multiv08.at
index c7bfb9d..364235e 100644
--- a/tests/multiv08.at
+++ b/tests/multiv08.at
@@ -17,7 +17,7 @@
 # Description: Tar 1.23 would in some cases silently fail to create
 # a continuation header in multivolume archives.
 #
-# In this testcase, the file `a' is 18.5 blocks long and the file `b'
+# In this testcase, the file 'a' is 18.5 blocks long and the file 'b'
 # is 19.5 blocks long.
 
 AT_SETUP([multivolume header creation])
diff --git a/tests/rename01.at b/tests/rename01.at
index e62e89d..ce8d030 100644
--- a/tests/rename01.at
+++ b/tests/rename01.at
@@ -81,7 +81,7 @@ End directory listing 2
 ],
 [tar: foo: Directory is new
 tar: foo/bar: Directory is new
-tar: foo/baz: Directory has been renamed from `foo/bar'
+tar: foo/baz: Directory has been renamed from 'foo/bar'
 ],
 [],[],[gnu, oldgnu, posix])
 
diff --git a/tests/rename02.at b/tests/rename02.at
index 4c51f26..835139e 100644
--- a/tests/rename02.at
+++ b/tests/rename02.at
@@ -96,7 +96,7 @@ End directory listing 2
 [tar: foo: Directory is new
 tar: foo/bar: Directory is new
 tar: foo/bar/baz: Directory is new
-tar: foo/baz: Directory has been renamed from `foo/bar/baz'
+tar: foo/baz: Directory has been renamed from 'foo/bar/baz'
 ],
 [],[],[gnu, oldgnu, posix])
 
diff --git a/tests/rename03.at b/tests/rename03.at
index d6d9a27..5b07c0c 100644
--- a/tests/rename03.at
+++ b/tests/rename03.at
@@ -118,9 +118,9 @@ tar: foo/b: Directory is new
 tar: foo/c: Directory is new
 tar: foo: Directory is new
 Second dump
-tar: foo/a: Directory has been renamed from `foo/c'
-tar: foo/b: Directory has been renamed from `foo/a'
-tar: foo/c: Directory has been renamed from `foo/b'
+tar: foo/a: Directory has been renamed from 'foo/c'
+tar: foo/b: Directory has been renamed from 'foo/a'
+tar: foo/c: Directory has been renamed from 'foo/b'
 ],
 [],[],[gnu, oldgnu, posix])
 
diff --git a/tests/sparse04.at b/tests/sparse04.at
index 9520803..c839931 100644
--- a/tests/sparse04.at
+++ b/tests/sparse04.at
@@ -23,7 +23,7 @@ AT_KEYWORDS([sparse sparse04])
 
 # Description: Tar versions from 1.15.92 to 1.25 would incorrectly
 # store sparse file names longer than 100 characters in pax mode.
-# Namely, the `path' keyword of the produced PAX header would contain the
+# Namely, the 'path' keyword of the produced PAX header would contain the
 # crafted name of the header itself, instead of that of the file.
 # Reported by: Kamil Dudka <kdudka@redhat.com>
 # References: <201011250026.44908.kdudka@redhat.com>,
diff --git a/tests/star/README b/tests/star/README
index 2b96aa6..21faf99 100644
--- a/tests/star/README
+++ b/tests/star/README
@@ -3,7 +3,7 @@ star "test archives". The archives themselves can be obtained
 from ftp://ftp.berlios.de/pub/star/testscripts.
 
 These tests are disabled by default. There are two ways to run
-them.  The simplest is by `make check-full' command.  It requires wget
+them.  The simplest is by 'make check-full' command.  It requires wget
 and GNU md5sum to be installed. If you don't have these utilities,
 download the test archives manually and set the environment variable
 STAR_TESTSCRIPTS to point to the directory where they reside, for
@@ -13,9 +13,9 @@ example:
 or
 	make TESTS_ENVIRONMENT='STAR_TESTSCRIPTS=testdir' check
 
-The file `quicktest.sh' is a separate test. It is never executed
-within `make check' command, you will need to run it manually.
-Please, carefully read section `quicktest.sh' below before running
+The file 'quicktest.sh' is a separate test. It is never executed
+within 'make check' command, you will need to run it manually.
+Please, carefully read section 'quicktest.sh' below before running
 it.
 
 Following is a short description of the tests:
@@ -52,12 +52,12 @@ containing very large files (in this case -- 10 GB).
 
 This is a test for compliance to POSIX.1-1990 tar specification. It
 requires two files: ustar-all-quicktest.tar and quicktest.filelist,
-(they usually reside in star/testscripts directory), and `tartest' program
+(they usually reside in star/testscripts directory), and 'tartest' program
 (also part of star distribution). The test must be run only with root
 privileges, so it is a good idea to verify the contents of
 ustar-all-quicktest.tar before running it.
 
-If `tartest' is not in your PATH, use TARTEST variable to specify its
+If 'tartest' is not in your PATH, use TARTEST variable to specify its
 full file name. You may pass both STAR_TESTSCRIPTS and TARTEST
 variables as command line arguments to quicktest.sh, as shown in
 the example below:
diff --git a/tests/update01.at b/tests/update01.at
index 374b971..129061b 100644
--- a/tests/update01.at
+++ b/tests/update01.at
@@ -20,7 +20,7 @@
 
 # Description: If dir is a directory and arc is a tar archive which
 # contains that directory, and dir contains some modifications added
-# after adding it to the archive, then `tar -u dir' would add dir/ to
+# after adding it to the archive, then 'tar -u dir' would add dir/ to
 # the archive.
 # Last-Affected-Version: 1.22.90
 # References: <4AD4E703.80500@teclabs.eu>
diff --git a/tests/volume.at b/tests/volume.at
index 87a910b..ec46b7d 100644
--- a/tests/volume.at
+++ b/tests/volume.at
@@ -43,13 +43,13 @@ test $? = 2
 [0],
 [],
 [-----
-tar: Volume `label' does not match `lab'
+tar: Volume 'label' does not match 'lab'
 tar: Error is not recoverable: exiting now
 -----
-tar: Volume `label' does not match `bel'
+tar: Volume 'label' does not match 'bel'
 tar: Error is not recoverable: exiting now
 -----
-tar: Volume `label' does not match `babel'
+tar: Volume 'label' does not match 'babel'
 tar: Error is not recoverable: exiting now
 ],
 [],[],[gnu, oldgnu])
-- 
cgit v1.0-41-gc330

