diff --git a/src/misc.c b/src/misc.c
index 0ade200..148d766 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -660,6 +660,30 @@ mknod_error (char const *name)
 }
 
 void
+exec_error (char const *name)
+{
+  call_arg_error ("exec", name);
+}
+
+void
+fork_error (char const *name)
+{
+  call_arg_error ("fork", name);
+}
+
+void
+dup2_error (char const *name)
+{
+  call_arg_error ("dup2", name);
+}
+
+void
+pipe_error (char const *name)
+{
+  call_arg_error ("pipe", name);
+}
+
+void
 open_error (char const *name)
 {
   call_arg_error ("open", name);
-- 
cgit v1.0-41-gc330

