diff --git a/THANKS b/THANKS
index c8052a4..79aa0ce 100644
--- a/THANKS
+++ b/THANKS
@@ -402,6 +402,7 @@ Phil Proudman		phil@proudman51.freeserve.co.uk
 Philippe Defert		defert@cern.ch
 Piercarlo Grandi	piercarl@sabi.demon.co.uk
 Pierce Cantrell		cantrell@ee.tamu.edu
+Piotr Rotter 		piotr.rotter@active24.pl
 R. Kent Dybvig		dyb@cadence.bloomington.in.us
 R. Scott Butler		butler@prism.es.dupont.com
 Rainer Orth		ro@TechFak.Uni-Bielefeld.DE
diff --git a/src/common.h b/src/common.h
index 21df8c1..ccf27c3 100644
--- a/src/common.h
+++ b/src/common.h
@@ -596,6 +596,7 @@ char *zap_slashes (char *name);
 char *normalize_filename (const char *name);
 void replace_prefix (char **pname, const char *samp, size_t slen,
 		     const char *repl, size_t rlen);
+char *tar_savedir (const char *name, int must_exist);
 
 typedef struct namebuf *namebuf_t;
 namebuf_t namebuf_create (const char *dir);
diff --git a/src/incremen.c b/src/incremen.c
index 557df30..8aeefbe 100644
--- a/src/incremen.c
+++ b/src/incremen.c
@@ -1582,7 +1582,7 @@ try_purge_directory (char const *directory_name)
   if (!is_dumpdir (&current_stat_info))
     return false;
 
-  current_dir = savedir (directory_name);
+  current_dir = tar_savedir (directory_name, 0);
 
   if (!current_dir)
     /* The directory doesn't exist now.  It'll be created.  In any
diff --git a/src/misc.c b/src/misc.c
index 1b78b69..d68d806 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -631,7 +631,7 @@ remove_any_file (const char *file_name, enum remove_option option)
 
 	case RECURSIVE_REMOVE_OPTION:
 	  {
-	    char *directory = savedir (file_name);
+	    char *directory = tar_savedir (file_name, 0);
 	    char const *entry;
 	    size_t entrylen;
 
@@ -1144,3 +1144,31 @@ namebuf_name (namebuf_t buf, const char *name)
   strcpy (buf->buffer + buf->dir_length, name);
   return buf->buffer;
 }
+
+/* Return the filenames in directory NAME, relative to the chdir_fd.
+   If the directory does not exist, report error if MUST_EXIST is
+   true.
+
+   Return NULL on errors.
+*/
+char *
+tar_savedir (const char *name, int must_exist)
+{
+  char *ret = NULL;
+  DIR *dir = NULL;
+  int fd = openat (chdir_fd, name, open_read_flags | O_DIRECTORY);
+  if (fd < 0)
+    {
+      if (!must_exist && errno == ENOENT)
+	return NULL;
+      open_error (name);
+    }
+  else if (! ((dir = fdopendir (fd))
+	      && (ret = streamsavedir (dir))))
+    savedir_error (name);
+
+  if (dir ? closedir (dir) != 0 : 0 <= fd && close (fd) != 0)
+    savedir_error (name);
+
+  return ret;
+}
diff --git a/src/update.c b/src/update.c
index 5e9ab98..72141b2 100644
--- a/src/update.c
+++ b/src/update.c
@@ -145,16 +145,8 @@ update_archive (void)
 		  {
 		    if (S_ISDIR (s.st_mode))
 		      {
-			char *p, *dirp;
-			DIR *stream = NULL;
-			int fd = openat (chdir_fd, name->name,
-					 open_read_flags | O_DIRECTORY);
-			if (fd < 0)
-			  open_error (name->name);
-			else if (! ((stream = fdopendir (fd))
-				    && (dirp = streamsavedir (stream))))
-			  savedir_error (name->name);
-			else
+			char *p, *dirp = tar_savedir (name->name, 1);
+			if (dirp)
 			  {
 			    namebuf_t nbuf = namebuf_create (name->name);
 
@@ -167,11 +159,6 @@ update_archive (void)
 
 			    remname (name);
 			  }
-
-			if (stream
-			    ? closedir (stream) != 0
-			    : 0 <= fd && close (fd) != 0)
-			  savedir_error (name->name);
 		      }
 		    else if (tar_timespec_cmp (get_stat_mtime (&s),
 					       current_stat_info.mtime)
diff --git a/tests/Makefile.am b/tests/Makefile.am
index b6bceea..cfab3e0 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -111,6 +111,7 @@ TESTSUITE_AT = \
  incr04.at\
  incr05.at\
  incr06.at\
+ incr07.at\
  indexfile.at\
  ignfail.at\
  label01.at\
diff --git a/tests/incr07.at b/tests/incr07.at
new file mode 100644
index 0000000..3d5fc0e
--- /dev/null
+++ b/tests/incr07.at
@@ -0,0 +1,106 @@
+# Process this file with autom4te to create testsuite. -*- Autotest -*-
+# Test suite for GNU tar.
+# Copyright 2009, 2013 Free Software Foundation, Inc.
+#
+# GNU tar is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# GNU tar is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+AT_SETUP([incremental restores with -C])
+AT_KEYWORDS([incremental extract incr07])
+
+# Tar 1.26 had problems extracting from incremental restores when given
+# the -C option.  The code in incremen.c:try_purge_directory and
+# misc.c:remove_any_file was using savedir(), which ignored eventual changes
+# in the current working directory and caused the malfunctioning.
+#
+# The problem was reported by Piotr Rotter on 2013-03-22.
+#
+# This testcase is based on scripts provided by Piotr Rotter and Nathan
+# Stratton Treadway.
+#
+# References: <514C8F56.90900@active24.pl>,
+#             http://lists.gnu.org/archive/html/bug-tar/2013-03/msg00036.html, 
+#             <20130326181922.GZ3732@shire.ontko.com>,
+#             http://lists.gnu.org/archive/html/bug-tar/2013-03/msg00042.html,
+#             <20130327051828.GA3732@shire.ontko.com>,
+#             http://lists.gnu.org/archive/html/bug-tar/2013-03/msg00043.html,
+#             <20130327054957.GB3732@shire.ontko.com>,
+#             http://lists.gnu.org/archive/html/bug-tar/2013-03/msg00044.html
+
+AT_TAR_CHECK([
+mkdir A
+echo 'a' > A/a
+echo 'a' > A/b
+
+decho C0
+tar -g test.snar -vcf test.0.tar A
+
+echo 'a' > A/c
+decho C1
+tar -g test.snar -vcf test.1.tar A
+
+rm -f A/a
+decho C2
+tar -g test.snar -vcf test.2.tar A
+
+mkdir ext
+rm -rf A
+
+decho E0
+tar -g test.snar -vxf test.0.tar -C ext/
+decho E1
+tar -g test.snar -vxf test.1.tar -C ext/
+
+mkdir ext/A/B
+touch ext/A/B/file
+
+decho E2
+tar -g test.snar -vxf test.2.tar -C ext/
+
+echo FIN
+test -d A && echo >&2 "toplevel A exists"
+exit 0
+],
+[0],
+[C0
+A/
+A/a
+A/b
+C1
+A/
+A/c
+C2
+A/
+E0
+A/
+A/a
+A/b
+E1
+A/
+A/c
+E2
+A/
+tar: Deleting 'A/a'
+tar: Deleting 'A/B'
+FIN
+],
+[C0
+tar: A: Directory is new
+C1
+C2
+E0
+E1
+E2
+],[],[],[gnu, oldgnu, posix])
+
+AT_CLEANUP
diff --git a/tests/testsuite.at b/tests/testsuite.at
index 36868c0..8aeebfe 100644
--- a/tests/testsuite.at
+++ b/tests/testsuite.at
@@ -296,6 +296,7 @@ m4_include([incr03.at])
 m4_include([incr04.at])
 m4_include([incr05.at])
 m4_include([incr06.at])
+m4_include([incr07.at])
 
 AT_BANNER([Files removed while archiving])
 m4_include([filerem01.at])
-- 
cgit v1.0-41-gc330

