diff --git a/src/buffer.c b/src/buffer.c
index e77f47c..d60eaea 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -1080,7 +1080,7 @@ write_error (ssize_t status)
 static void
 read_error (void)
 {
-  WARN ((0, errno, _("Read error on %s"), *archive_name_cursor));
+  ERROR ((0, errno, _("Read error on %s"), *archive_name_cursor));
 
   if (record_start_block == 0)
     FATAL_ERROR ((0, 0, _("At beginning of tape, quitting now")));
-- 
cgit v1.0-41-gc330

