diff --git a/src/common.h b/src/common.h
index f9b97af..2cca38c 100644
--- a/src/common.h
+++ b/src/common.h
@@ -630,9 +630,10 @@ void sys_save_archive_dev_ino (void);
 void sys_drain_input_pipe (void);
 void sys_wait_for_child (pid_t);
 void sys_spawn_shell (void);
-void sys_compare_uid_gid (struct stat *a, struct stat *b);
+bool sys_compare_uid (struct stat *a, struct stat *b);
+bool sys_compare_gid (struct stat *a, struct stat *b);
 bool sys_file_is_archive (struct tar_stat_info *p);
-void sys_compare_links (struct stat *link_data, struct stat *stat_data);
+bool sys_compare_links (struct stat *link_data, struct stat *stat_data);
 int sys_truncate (int fd);
 void sys_reset_uid_gid (void);
 pid_t sys_child_open_for_compress (void);
@@ -641,9 +642,10 @@ ssize_t sys_write_archive_buffer (void);
 bool sys_get_archive_stat (void);
 
 /* Module compare.c */
-void report_difference (const char *message, ...);
+void report_difference (struct tar_stat_info *st, const char *message, ...);
 
 /* Module sparse.c */
 bool sparse_file_p (struct tar_stat_info *stat);
 enum dump_status sparse_dump_file (int fd, struct tar_stat_info *stat);
 enum dump_status sparse_extract_file (int fd, struct tar_stat_info *stat, off_t *size);
+bool sparse_diff_file (int fd, struct tar_stat_info *stat);
-- 
cgit v1.0-41-gc330

