diff --git a/src/common.h b/src/common.h
index 9933aa6..b158a37 100644
--- a/src/common.h
+++ b/src/common.h
@@ -445,7 +445,11 @@ union block * write_extended (bool global, struct tar_stat_info *st,
 union block *start_private_header (const char *name, size_t size);
 void write_eot (void);
 void check_links (void);
-
+void exclusion_tag_warning (const char *dirname, const char *tagname,
+			    const char *message);
+enum exclusion_tag_type check_exclusion_tags (char *dirname,
+					      const char **tag_file_name);
+     
 #define GID_TO_CHARS(val, where) gid_to_chars (val, where, sizeof (where))
 #define MAJOR_TO_CHARS(val, where) major_to_chars (val, where, sizeof (where))
 #define MINOR_TO_CHARS(val, where) minor_to_chars (val, where, sizeof (where))
-- 
cgit v1.0-41-gc330

